{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "56851891",
   "metadata": {
    "id": "vECLNfPmhyDb"
   },
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5189f358",
   "metadata": {
    "id": "bsBOtyuTAJsC",
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: numpy in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (1.24.3)\n",
      "Requirement already satisfied: pandas in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (1.5.3)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from pandas) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from pandas) (2022.7)\n",
      "Requirement already satisfied: numpy>=1.21.0 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from pandas) (1.24.3)\n",
      "Requirement already satisfied: six>=1.5 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from python-dateutil>=2.8.1->pandas) (1.16.0)\n",
      "Requirement already satisfied: scipy in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (1.11.4)\n",
      "Requirement already satisfied: numpy<1.28.0,>=1.21.6 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from scipy) (1.24.3)\n",
      "Requirement already satisfied: hmmlearn in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (0.3.0)\n",
      "Requirement already satisfied: numpy>=1.10 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from hmmlearn) (1.24.3)\n",
      "Requirement already satisfied: scikit-learn!=0.22.0,>=0.16 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from hmmlearn) (1.3.0)\n",
      "Requirement already satisfied: scipy>=0.19 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from hmmlearn) (1.11.4)\n",
      "Requirement already satisfied: joblib>=1.1.1 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from scikit-learn!=0.22.0,>=0.16->hmmlearn) (1.2.0)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from scikit-learn!=0.22.0,>=0.16->hmmlearn) (2.2.0)\n",
      "Requirement already satisfied: statsmodels in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (0.14.0)\n",
      "Requirement already satisfied: numpy>=1.18 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from statsmodels) (1.24.3)\n",
      "Requirement already satisfied: scipy!=1.9.2,>=1.4 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from statsmodels) (1.11.4)\n",
      "Requirement already satisfied: pandas>=1.0 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from statsmodels) (1.5.3)\n",
      "Requirement already satisfied: patsy>=0.5.2 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from statsmodels) (0.5.3)\n",
      "Requirement already satisfied: packaging>=21.3 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from statsmodels) (23.0)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from pandas>=1.0->statsmodels) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from pandas>=1.0->statsmodels) (2022.7)\n",
      "Requirement already satisfied: six in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from patsy>=0.5.2->statsmodels) (1.16.0)\n",
      "Requirement already satisfied: mpl_scatter_density in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (0.7)\n",
      "Requirement already satisfied: numpy in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from mpl_scatter_density) (1.24.3)\n",
      "Requirement already satisfied: matplotlib>=3.0 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from mpl_scatter_density) (3.7.1)\n",
      "Requirement already satisfied: fast-histogram>=0.3 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from mpl_scatter_density) (0.12)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from matplotlib>=3.0->mpl_scatter_density) (1.0.5)\n",
      "Requirement already satisfied: cycler>=0.10 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from matplotlib>=3.0->mpl_scatter_density) (0.11.0)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from matplotlib>=3.0->mpl_scatter_density) (4.25.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from matplotlib>=3.0->mpl_scatter_density) (1.4.4)\n",
      "Requirement already satisfied: packaging>=20.0 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from matplotlib>=3.0->mpl_scatter_density) (23.0)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from matplotlib>=3.0->mpl_scatter_density) (9.4.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from matplotlib>=3.0->mpl_scatter_density) (3.0.9)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from matplotlib>=3.0->mpl_scatter_density) (2.8.2)\n",
      "Requirement already satisfied: six>=1.5 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from python-dateutil>=2.7->matplotlib>=3.0->mpl_scatter_density) (1.16.0)\n",
      "Requirement already satisfied: tqdm in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (4.65.0)\n",
      "Requirement already satisfied: colorama in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (0.4.6)\n",
      "Requirement already satisfied: pyliftover in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (0.4)\n"
     ]
    }
   ],
   "source": [
    "!pip install numpy\n",
    "!pip install pandas\n",
    "!pip install scipy\n",
    "!pip install hmmlearn\n",
    "!pip install statsmodels\n",
    "!pip install mpl_scatter_density\n",
    "!pip install tqdm\n",
    "!pip install colorama\n",
    "!pip install pyliftover"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "09f9e786",
   "metadata": {
    "executionInfo": {
     "elapsed": 8,
     "status": "ok",
     "timestamp": 1703779362738,
     "user": {
      "displayName": "Oscar Thomas Aguilar",
      "userId": "15151607230073393698"
     },
     "user_tz": 300
    },
    "id": "BZfHQLuPhyDl"
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.patches import Rectangle\n",
    "import seaborn as sns\n",
    "from hmmlearn import hmm\n",
    "import gzip\n",
    "import os\n",
    "from tqdm import tqdm\n",
    "import statsmodels.api as sm\n",
    "import scipy.stats as stats\n",
    "from scipy.stats import fisher_exact\n",
    "import math\n",
    "import joblib\n",
    "import glob\n",
    "from pyliftover import LiftOver\n",
    "# from multiprocessing import Pool, cpu_count\n",
    "import requests\n",
    "import warnings\n",
    "\n",
    "# Suppress all warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a34819db",
   "metadata": {
    "executionInfo": {
     "elapsed": 173,
     "status": "ok",
     "timestamp": 1703790405518,
     "user": {
      "displayName": "Oscar Thomas Aguilar",
      "userId": "15151607230073393698"
     },
     "user_tz": 300
    },
    "id": "9IWHWPQ_hyDm"
   },
   "outputs": [],
   "source": [
    "# Specify project directories in Sherlock\n",
    "data_path = '/oak/stanford/groups/mrivas/projects/wgs-constraint-llm/data/'\n",
    "results_path = '/oak/stanford/groups/mrivas/projects/wgs-constraint-llm/osthoag/wgs-constraint-llm/results/'\n",
    "\n",
    "# Specify the file paths\n",
    "annotation_file_path = data_path + 'gencode.v44.basic.annotation.gtf.gz'\n",
    "variants_file_path = data_path + 'rgc_me_variant_frequencies_20231004.vcf.gz'\n",
    "coverage_file_path = data_path + 'gnomad.exomes.v4.0.coverage.summary.tsv.bgz'\n",
    "constraint_file_path = data_path + 'gnomad.v4.0.constraint_metrics.tsv'\n",
    "domain_file_path = data_path + 'uniprotdomains.gz'\n",
    "gene_domain_file_path = data_path + 'chr2_gene_domain_combinations.tsv.gz'\n",
    "alpha_missense_file_path = data_path + 'AlphaMissense_hg38.tsv.gz'\n",
    "bipex_variants_file_path = data_path + 'BipEx_variant_results.tsv.bgz'\n",
    "aou_variants_file_path = data_path + 'aou_variants.tsv.gz'\n",
    "genome_coverage_file_path = data_path + 'gnomad.genomes.r3.0.1.coverage.summary.tsv.bgz'\n",
    "ukb_variants_file_path = data_path + 'ukb24310_c19.qced_nonrel.vmiss.gz'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "104f205c",
   "metadata": {
    "id": "vECLNfPmhyDb"
   },
   "source": [
    "## Define helper methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3740e7dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to calculate the proportion of each row that overlaps with an array of positions\n",
    "def calculate_overlap(row, positions, start_col='start', end_col='end'):\n",
    "    return np.sum((row[start_col] <= positions) & (positions <= row[end_col])) / (row[end_col] - row[start_col] + 1)\n",
    "\n",
    "def get_sequence(coverage_df, variants_df):\n",
    "    # Get the length of the genetic sequence\n",
    "    sequence_length = max(coverage_df['pos'].max(), variants_df['pos'].max())\n",
    "\n",
    "    # Create boolean mask for exome coverage\n",
    "    coverage_mask = np.zeros(sequence_length + 1, dtype=bool)\n",
    "\n",
    "    # Use NumPy boolean indexing to get mask for positions with over 80% coverage\n",
    "    coverage_mask[coverage_df['pos'].to_numpy()] = 1\n",
    "\n",
    "    # Initialize values to zero for all positions\n",
    "    sequence = np.zeros(sequence_length + 1)\n",
    "\n",
    "    # Set positions to 1 where a variant exists\n",
    "    sequence[variants_df['pos'].to_numpy()] = 1\n",
    "\n",
    "    # Filter for only the protein-coding regions with over 80% exome coverage\n",
    "    observations = np.array(sequence[coverage_mask])\n",
    "    \n",
    "    positions = np.where(coverage_mask)[0]\n",
    "    \n",
    "    return observations, positions\n",
    "\n",
    "def get_HMM_predictions(observations, model, order=2):\n",
    "    # Flatten the higher-order structure\n",
    "    X = np.stack([observations[i:i-order] for i in range(order)], axis=1)\n",
    "\n",
    "    # Convert observations to counts\n",
    "    X_counts = np.column_stack([(X == i).sum(axis=1) for i in range(2)])\n",
    "\n",
    "    # Predict probabilities for each position\n",
    "    probabilities = model.predict_proba(X_counts)\n",
    "\n",
    "    return probabilities\n",
    "\n",
    "def fit_HMM(observations, order=2):\n",
    "    # Flatten the higher-order structure\n",
    "    X = np.stack([observations[i:i-order] for i in range(order)], axis=1)\n",
    "\n",
    "    # Convert observations to counts\n",
    "    X_counts = np.column_stack([(X == i).sum(axis=1) for i in range(2)])\n",
    "\n",
    "    # Create and fit a first-order HMM\n",
    "    model = hmm.MultinomialHMM(n_components=2, random_state=10)\n",
    "    model.fit(X_counts)\n",
    "\n",
    "    return model\n",
    "\n",
    "def ols_regression(predictions_df):\n",
    "    # Add a constant term to the independent variable for the intercept\n",
    "    X = sm.add_constant(predictions_df['prob_0'])\n",
    "\n",
    "    # Fit the linear regression model\n",
    "    model4 = sm.OLS(predictions_df['observation'],X).fit()\n",
    "\n",
    "    model4.summary().tables[1].pvalues_precision = 100  # Adjust the number of significant digits\n",
    "\n",
    "    # Get the summary of the regression\n",
    "    print(model4.summary())\n",
    "\n",
    "    # Extract the F-statistic and its associated p-value\n",
    "    f_statistic = model4.fvalue\n",
    "    p_value_f_statistic = model4.f_pvalue\n",
    "    \n",
    "    return f_statistic, p_value_f_statistic\n",
    "\n",
    "def fishers_exact_test(gene_id, ac_case, an_case, ac_ctrl, an_ctrl):\n",
    "    if any((value < 0) or (math.isnan(value)) for value in [ac_case, an_case - ac_case, ac_ctrl, an_ctrl - ac_ctrl]):\n",
    "        print(f\"Negative values detected: gene={gene_id}, ac_case={ac_case}, an_case={an_case}, ac_ctrl={ac_ctrl}, an_ctrl={an_ctrl}\")\n",
    "        return None, None\n",
    "\n",
    "    contingency_table = [[ac_case, an_case - ac_case], [ac_ctrl, an_ctrl - ac_ctrl]]\n",
    "    odds_ratio, p_value = fisher_exact(contingency_table)\n",
    "    return odds_ratio, p_value\n",
    "\n",
    "def apply_fishers_exact_test(cases_df):\n",
    "    # Calculate total counts for cases and controls\n",
    "    cases_df['total_ac_case'] = cases_df.groupby(['gene_id', 'group'])['ac_case'].transform('sum')\n",
    "    cases_df['total_an_case'] = cases_df.groupby(['gene_id', 'group'])['an_case'].transform('max')\n",
    "    cases_df['total_ac_ctrl'] = cases_df.groupby(['gene_id', 'group'])['ac_ctrl'].transform('sum')\n",
    "    cases_df['total_an_ctrl'] = cases_df.groupby(['gene_id', 'group'])['an_ctrl'].transform('max')\n",
    "    \n",
    "    print(cases_df)\n",
    "\n",
    "    # Remove duplicate rows for unique combinations of 'gene_id' and 'group'\n",
    "    unique_cases_df = cases_df[['gene_id', 'gene_name', 'group', 'total_ac_case', 'total_an_case', 'total_ac_ctrl', 'total_an_ctrl']].drop_duplicates()\n",
    "\n",
    "    # Drop Nan values from gene_id\n",
    "    unique_cases_df = unique_cases_df.dropna(subset=['gene_id'])\n",
    "    print(unique_cases_df.apply(lambda row: fishers_exact_test(row['gene_id'],row['total_ac_case'], row['total_an_case'], row['total_ac_ctrl'], row['total_an_ctrl']), axis=1, result_type='expand'))\n",
    "    # Apply Fisher's exact test for each unique combination\n",
    "    unique_cases_df[['odds_ratio', 'p_value']] = unique_cases_df.apply(lambda row: fishers_exact_test(row['gene_id'],row['total_ac_case'], row['total_an_case'], row['total_ac_ctrl'], row['total_an_ctrl']), axis=1, result_type='expand')\n",
    "\n",
    "    # Sort the DataFrame by 'p_value' in increasing order\n",
    "    sorted_df = unique_cases_df.sort_values(by='p_value')\n",
    "    \n",
    "    return sorted_df\n",
    "\n",
    "# Initialize the liftover object for hg19 to hg38\n",
    "lo = LiftOver('hg19', 'hg38')\n",
    "\n",
    "# Function to liftover a single row\n",
    "def liftover(row):\n",
    "    lifted = lo.convert_coordinate(row['chr'], row['pos'])\n",
    "    if lifted:\n",
    "        # Returns the first lifted coordinate (chromosome, position)\n",
    "        return lifted[0][0], int(lifted[0][1])\n",
    "    else:\n",
    "        # Returns 0 if liftover fails\n",
    "        return row['chr'], 0\n",
    "    \n",
    "from scipy.stats import chi2\n",
    "def fisher_method(p1, p2):\n",
    "    p_values = np.array([p1, p2])\n",
    "    \n",
    "    if np.isnan(p1) and np.isnan(p2):\n",
    "        return None\n",
    "    if np.isnan(p1):\n",
    "        return p2\n",
    "    if np.isnan(p2):\n",
    "        return p1\n",
    "        \n",
    "\n",
    "    # Convert p-values to chi-squared statistics\n",
    "    chi_squared_stats = -2 * np.log(p_values)\n",
    "\n",
    "    # Sum of chi-squared statistics\n",
    "    chi_squared_sum = np.sum(chi_squared_stats)\n",
    "\n",
    "    # Degrees of freedom for the chi-squared distribution\n",
    "    degrees_of_freedom = 2 * len(p_values)\n",
    "\n",
    "    # Combined p-value using chi-squared distribution\n",
    "    p_comb = 1 - chi2.cdf(chi_squared_sum, degrees_of_freedom)\n",
    "\n",
    "    return p_comb\n",
    "\n",
    "def plot_hist_from_predictions(predictions_df):\n",
    "    # Create a figure with two subplots in one row and two columns\n",
    "    fig, axs = plt.subplots(1, 2, figsize=(12, 5))\n",
    "\n",
    "    # Plot the first histogram in the first subplot\n",
    "    axs[0].hist(predictions_df['observation'], edgecolor=\"blue\")\n",
    "    axs[0].set_title('Histogram of Observations')\n",
    "    axs[0].set_xlabel('Has a variant')\n",
    "    axs[0].set_ylabel('Frequency')\n",
    "\n",
    "    # Plot the second histogram in the second subplot\n",
    "    axs[1].hist(predictions_df['prob_0'], edgecolor=\"blue\")\n",
    "    axs[1].set_title('Histogram of Predictions')\n",
    "    axs[1].set_xlabel('Probability of 0')\n",
    "    axs[1].set_ylabel('Frequency')\n",
    "\n",
    "    # Adjust layout to prevent clipping of titles\n",
    "    plt.tight_layout()\n",
    "\n",
    "    # Show the plots\n",
    "    plt.show()\n",
    "    \n",
    "def plot_subsequence_predictions(start_idx, end_idx, name='Target Region'):\n",
    "    fig, axes = plt.subplots(nrows=2, ncols=1, figsize=(10, 5), sharex=True)\n",
    "    plt.suptitle(\"Actual vs Predicted for \" + name)\n",
    "\n",
    "    # Plot the original sequence\n",
    "    axes[0].bar(range(start_idx, end_idx), observations[start_idx:end_idx], color='blue', label='Original Sequence')\n",
    "    axes[0].set_ylabel('Observation')\n",
    "    axes[0].legend()\n",
    "\n",
    "    # Plot the predicted probabilities as stacked barplots\n",
    "    axes[1].bar(range(start_idx, end_idx), probabilities[start_idx:end_idx, 0], color='orange', label='Probability of 0')\n",
    "    axes[1].bar(range(start_idx, end_idx), probabilities[start_idx:end_idx, 1], bottom=probabilities[start_idx:end_idx, 0], color='green', label='Probability of 1')\n",
    "    axes[1].set_xlabel('Position')\n",
    "    axes[1].set_ylabel('Prediction')\n",
    "    axes[1].legend()\n",
    "\n",
    "    plt.tight_layout(rect=[0, 0, 1, 0.96])\n",
    "    plt.show()\n",
    "    \n",
    "def plot_group_QQ(significance_df):\n",
    "    # Get unique groups\n",
    "    unique_groups = significance_df['group'].unique()\n",
    "\n",
    "    # Number of unique groups\n",
    "    num_groups = len(unique_groups)\n",
    "\n",
    "    # Create a figure with subplots\n",
    "    fig, axs = plt.subplots(num_groups, 1, figsize=(8, 4 * num_groups))\n",
    "\n",
    "    # Loop over each group and create a Q-Q plot\n",
    "    for i, group in enumerate(unique_groups):\n",
    "        # Filter the DataFrame for the current group\n",
    "        group_df = significance_df[significance_df['group'] == group]\n",
    "\n",
    "        # Calculate -log10(p-values)\n",
    "        group_df['minus_log_pvalue'] = -np.log10(group_df['p_value'])\n",
    "\n",
    "        # Sort by p-value in ascending order\n",
    "        group_df = group_df.sort_values('p_value', ascending=True)\n",
    "\n",
    "        # Calculate theoretical quantiles\n",
    "        theoretical_quantiles = stats.norm.ppf((np.arange(1, len(group_df) + 1) / (len(group_df) + 1)))\n",
    "\n",
    "        # Transform theoretical quantiles into -log10(p-values)\n",
    "        group_df['theoretical_quantiles'] = -np.log10(stats.norm.cdf(theoretical_quantiles))\n",
    "\n",
    "        # Plotting the Q-Q plot for the group\n",
    "        ax = axs[i] if num_groups > 1 else axs\n",
    "        ax.scatter(group_df['theoretical_quantiles'], group_df['minus_log_pvalue'], label=f'Observed -log10(p-value) for group {group}')\n",
    "\n",
    "        # Define x for the expected line and confidence intervals\n",
    "        x = np.linspace(min(group_df['theoretical_quantiles']), max(group_df['theoretical_quantiles']), 100)\n",
    "        y = x\n",
    "        ax.plot(x, y, color='r', ls='--', label='Expected -log10(p-value)')\n",
    "        ax.fill_between(x, y - 1.96*np.sqrt((1-x)*x/len(group_df)), y + 1.96*np.sqrt((1-x)*x/len(group_df)), color='r', alpha=0.2)\n",
    "\n",
    "        # Labeling significant genes\n",
    "        significant_genes = group_df[group_df['p_value'] < .0001]\n",
    "        for idx, row in significant_genes.iterrows():\n",
    "            ax.text(row['theoretical_quantiles'], row['minus_log_pvalue'], row['gene_name'], fontsize=6)\n",
    "\n",
    "        # Labeling the plot\n",
    "        ax.set_xlabel('Expected -log10(p-value)')\n",
    "        ax.set_ylabel('Observed -log10(p-value)')\n",
    "        ax.set_title(f'Q-Q Plot of Gene P-Values for Group {group}')\n",
    "        ax.legend()\n",
    "\n",
    "    # Adjust layout\n",
    "    plt.tight_layout()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "230a0306",
   "metadata": {
    "id": "yAEgmgmSGY7d"
   },
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "37366f9f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript</th>\n",
       "      <th>transcript_num</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>std_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>65565</td>\n",
       "      <td>65573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENST00000641515</td>\n",
       "      <td>2</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69037</td>\n",
       "      <td>70005</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENST00000641515</td>\n",
       "      <td>2</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>450743</td>\n",
       "      <td>451678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000284733.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>ENST00000426406.4</td>\n",
       "      <td>ENST00000426406</td>\n",
       "      <td>4</td>\n",
       "      <td>OR4F29-201</td>\n",
       "      <td>ENSG00000284733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>685719</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000284662.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>ENST00000332831.5</td>\n",
       "      <td>ENST00000332831</td>\n",
       "      <td>5</td>\n",
       "      <td>OR4F16-201</td>\n",
       "      <td>ENSG00000284662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>524</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>924432</td>\n",
       "      <td>924948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENST00000616016.5</td>\n",
       "      <td>ENST00000616016</td>\n",
       "      <td>5</td>\n",
       "      <td>SAMD11-209</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1998488</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>10470</td>\n",
       "      <td>10763</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000212907.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND4L</td>\n",
       "      <td>ENST00000361335.1</td>\n",
       "      <td>ENST00000361335</td>\n",
       "      <td>1</td>\n",
       "      <td>MT-ND4L-201</td>\n",
       "      <td>ENSG00000212907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1998495</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>10760</td>\n",
       "      <td>12137</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000198886.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND4</td>\n",
       "      <td>ENST00000361381.2</td>\n",
       "      <td>ENST00000361381</td>\n",
       "      <td>2</td>\n",
       "      <td>MT-ND4-201</td>\n",
       "      <td>ENSG00000198886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1998509</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>12337</td>\n",
       "      <td>14145</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000198786.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND5</td>\n",
       "      <td>ENST00000361567.2</td>\n",
       "      <td>ENST00000361567</td>\n",
       "      <td>2</td>\n",
       "      <td>MT-ND5-201</td>\n",
       "      <td>ENSG00000198786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1998515</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>ENST00000361681.2</td>\n",
       "      <td>ENST00000361681</td>\n",
       "      <td>2</td>\n",
       "      <td>MT-ND6-201</td>\n",
       "      <td>ENSG00000198695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1998523</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>ENST00000361789.2</td>\n",
       "      <td>ENST00000361789</td>\n",
       "      <td>2</td>\n",
       "      <td>MT-CYB-201</td>\n",
       "      <td>ENSG00000198727</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>654548 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr   source feature   start     end score strand frame  \\\n",
       "60       chr1   HAVANA     CDS   65565   65573     .      +     0   \n",
       "63       chr1   HAVANA     CDS   69037   70005     .      +     0   \n",
       "236      chr1   HAVANA     CDS  450743  451678     .      -     0   \n",
       "304      chr1   HAVANA     CDS  685719  686654     .      -     0   \n",
       "524      chr1   HAVANA     CDS  924432  924948     .      +     0   \n",
       "...       ...      ...     ...     ...     ...   ...    ...   ...   \n",
       "1998488  chrM  ENSEMBL     CDS   10470   10763     .      +     0   \n",
       "1998495  chrM  ENSEMBL     CDS   10760   12137     .      +     0   \n",
       "1998509  chrM  ENSEMBL     CDS   12337   14145     .      +     0   \n",
       "1998515  chrM  ENSEMBL     CDS   14149   14673     .      -     0   \n",
       "1998523  chrM  ENSEMBL     CDS   14747   15887     .      +     0   \n",
       "\n",
       "                    gene_id       gene_type gene_name      transcript_id  \\\n",
       "60        ENSG00000186092.7  protein_coding     OR4F5  ENST00000641515.2   \n",
       "63        ENSG00000186092.7  protein_coding     OR4F5  ENST00000641515.2   \n",
       "236       ENSG00000284733.2  protein_coding    OR4F29  ENST00000426406.4   \n",
       "304       ENSG00000284662.2  protein_coding    OR4F16  ENST00000332831.5   \n",
       "524      ENSG00000187634.13  protein_coding    SAMD11  ENST00000616016.5   \n",
       "...                     ...             ...       ...                ...   \n",
       "1998488   ENSG00000212907.2  protein_coding   MT-ND4L  ENST00000361335.1   \n",
       "1998495   ENSG00000198886.2  protein_coding    MT-ND4  ENST00000361381.2   \n",
       "1998509   ENSG00000198786.2  protein_coding    MT-ND5  ENST00000361567.2   \n",
       "1998515   ENSG00000198695.2  protein_coding    MT-ND6  ENST00000361681.2   \n",
       "1998523   ENSG00000198727.2  protein_coding    MT-CYB  ENST00000361789.2   \n",
       "\n",
       "              transcript transcript_num transcript_name      std_gene_id  \n",
       "60       ENST00000641515              2       OR4F5-201  ENSG00000186092  \n",
       "63       ENST00000641515              2       OR4F5-201  ENSG00000186092  \n",
       "236      ENST00000426406              4      OR4F29-201  ENSG00000284733  \n",
       "304      ENST00000332831              5      OR4F16-201  ENSG00000284662  \n",
       "524      ENST00000616016              5      SAMD11-209  ENSG00000187634  \n",
       "...                  ...            ...             ...              ...  \n",
       "1998488  ENST00000361335              1     MT-ND4L-201  ENSG00000212907  \n",
       "1998495  ENST00000361381              2      MT-ND4-201  ENSG00000198886  \n",
       "1998509  ENST00000361567              2      MT-ND5-201  ENSG00000198786  \n",
       "1998515  ENST00000361681              2      MT-ND6-201  ENSG00000198695  \n",
       "1998523  ENST00000361789              2      MT-CYB-201  ENSG00000198727  \n",
       "\n",
       "[654548 rows x 16 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the GTF file into a pandas DataFrame\n",
    "gene_df = pd.read_csv(annotation_file_path, sep='\\t', comment='#', header=None, \n",
    "                      names=['chr', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute'], \n",
    "                      dtype={'start': int, 'end': int})\n",
    "\n",
    "# Extract 'gene_id' from attributes\n",
    "gene_df['gene_id'] = gene_df['attribute'].str.extract(r'gene_id \"(.*?)\"')\n",
    "\n",
    "# Extract 'gene_type' from attributes\n",
    "gene_df['gene_type'] = gene_df['attribute'].str.extract(r'gene_type \"(.*?)\"')\n",
    "\n",
    "# Extract 'gene_name' from attributes\n",
    "gene_df['gene_name'] = gene_df['attribute'].str.extract(r'gene_name \"(.*?)\"')\n",
    "\n",
    "# Extract 'transcript_id' from attributes\n",
    "gene_df['transcript_id'] = gene_df['attribute'].str.extract(r'transcript_id \"(.*?)\"')\n",
    "\n",
    "# Extract 'transcript' and 'num' from transcript_id\n",
    "gene_df[['transcript', 'transcript_num']] = gene_df['transcript_id'].str.split('.', expand=True)\n",
    "\n",
    "# Extract 'transcript_name' from attributes\n",
    "gene_df['transcript_name'] = gene_df['attribute'].str.extract(r'transcript_name \"(.*?)\"')\n",
    "\n",
    "# Drop the original attribute column\n",
    "gene_df = gene_df.drop('attribute', axis=1)\n",
    "\n",
    "# Filter rows for protein-coding regions\n",
    "gene_df = gene_df[(gene_df['gene_type'] == 'protein_coding') & (gene_df['feature'] == 'CDS')]\n",
    "\n",
    "gene_df['std_gene_id'] = gene_df['gene_id'].str.split('.').str[0]\n",
    "\n",
    "gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30df8022",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the file into a pandas DataFrame\n",
    "alpha_missense_df = pd.read_csv(alpha_missense_file_path, sep='\\t', header=3)#, nrows=20)\n",
    "\n",
    "# Rename columns to standard labels\n",
    "alpha_missense_df.rename(columns={\"#CHROM\": \"chr\", \"POS\": \"pos\", \"REF\": \"ref\", \"ALT\": 'alt'}, inplace=True)\n",
    "\n",
    "# Extract 'transcript' and 'num' from transcript_id\n",
    "alpha_missense_df[['transcript', 'transcript_num']] = alpha_missense_df['transcript_id'].str.split('.', expand=True)\n",
    "alpha_missense_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e65487d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus</th>\n",
       "      <th>alleles</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>consequence</th>\n",
       "      <th>hgvsc</th>\n",
       "      <th>hgvsp</th>\n",
       "      <th>cadd</th>\n",
       "      <th>mpc</th>\n",
       "      <th>polyphen</th>\n",
       "      <th>group</th>\n",
       "      <th>...</th>\n",
       "      <th>ac_ctrl</th>\n",
       "      <th>an_ctrl</th>\n",
       "      <th>estimate</th>\n",
       "      <th>chi_sq_stat</th>\n",
       "      <th>p_value</th>\n",
       "      <th>in_analysis</th>\n",
       "      <th>in_gnomad_non_neuro</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:925922</td>\n",
       "      <td>[\"G\",\"A\"]</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>other_missense</td>\n",
       "      <td>c.-20G&gt;A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22.30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bipolar Disorder</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>28844</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>chr1</td>\n",
       "      <td>925922</td>\n",
       "      <td>SAMD11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:925922</td>\n",
       "      <td>[\"G\",\"A\"]</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>other_missense</td>\n",
       "      <td>c.-20G&gt;A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22.30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bipolar Disorder (including Schizoaffective)</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>28844</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>chr1</td>\n",
       "      <td>925922</td>\n",
       "      <td>SAMD11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:925922</td>\n",
       "      <td>[\"G\",\"A\"]</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>other_missense</td>\n",
       "      <td>c.-20G&gt;A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22.30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bipolar Disorder 1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>28844</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>chr1</td>\n",
       "      <td>925922</td>\n",
       "      <td>SAMD11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1:925923</td>\n",
       "      <td>[\"G\",\"C\"]</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>other_missense</td>\n",
       "      <td>c.-19G&gt;C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20.20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bipolar Disorder</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>28844</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>chr1</td>\n",
       "      <td>925923</td>\n",
       "      <td>SAMD11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1:925923</td>\n",
       "      <td>[\"G\",\"C\"]</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>other_missense</td>\n",
       "      <td>c.-19G&gt;C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20.20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bipolar Disorder (including Schizoaffective)</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>28844</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>chr1</td>\n",
       "      <td>925923</td>\n",
       "      <td>SAMD11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9867259</th>\n",
       "      <td>chr22:50782243</td>\n",
       "      <td>[\"C\",\"T\"]</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>other_missense</td>\n",
       "      <td>c.52G&gt;A</td>\n",
       "      <td>p.Ala18Thr</td>\n",
       "      <td>18.69</td>\n",
       "      <td>NaN</td>\n",
       "      <td>benign</td>\n",
       "      <td>Bipolar Disorder with Psychosis</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>28844</td>\n",
       "      <td>0.31788</td>\n",
       "      <td>0.43477</td>\n",
       "      <td>0.50966</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50782243</td>\n",
       "      <td>RABL2B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9867260</th>\n",
       "      <td>chr22:50782243</td>\n",
       "      <td>[\"C\",\"T\"]</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>other_missense</td>\n",
       "      <td>c.52G&gt;A</td>\n",
       "      <td>p.Ala18Thr</td>\n",
       "      <td>18.69</td>\n",
       "      <td>NaN</td>\n",
       "      <td>benign</td>\n",
       "      <td>Bipolar Disorder without Psychosis</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>28844</td>\n",
       "      <td>0.38228</td>\n",
       "      <td>0.45574</td>\n",
       "      <td>0.49962</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50782243</td>\n",
       "      <td>RABL2B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9867261</th>\n",
       "      <td>chr22:50782290</td>\n",
       "      <td>[\"G\",\"C\"]</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>other_missense</td>\n",
       "      <td>c.5C&gt;G</td>\n",
       "      <td>p.Ala2Gly</td>\n",
       "      <td>24.10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>possibly_damaging</td>\n",
       "      <td>Bipolar Disorder</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>28844</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50782290</td>\n",
       "      <td>RABL2B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9867262</th>\n",
       "      <td>chr22:50782290</td>\n",
       "      <td>[\"G\",\"C\"]</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>other_missense</td>\n",
       "      <td>c.5C&gt;G</td>\n",
       "      <td>p.Ala2Gly</td>\n",
       "      <td>24.10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>possibly_damaging</td>\n",
       "      <td>Bipolar Disorder (including Schizoaffective)</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>28844</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50782290</td>\n",
       "      <td>RABL2B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9867263</th>\n",
       "      <td>chr22:50782290</td>\n",
       "      <td>[\"G\",\"C\"]</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>other_missense</td>\n",
       "      <td>c.5C&gt;G</td>\n",
       "      <td>p.Ala2Gly</td>\n",
       "      <td>24.10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>possibly_damaging</td>\n",
       "      <td>Bipolar Disorder 1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>28844</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50782290</td>\n",
       "      <td>RABL2B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9867264 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  locus    alleles          gene_id     consequence     hgvsc  \\\n",
       "0           chr1:925922  [\"G\",\"A\"]  ENSG00000187634  other_missense  c.-20G>A   \n",
       "1           chr1:925922  [\"G\",\"A\"]  ENSG00000187634  other_missense  c.-20G>A   \n",
       "2           chr1:925922  [\"G\",\"A\"]  ENSG00000187634  other_missense  c.-20G>A   \n",
       "3           chr1:925923  [\"G\",\"C\"]  ENSG00000187634  other_missense  c.-19G>C   \n",
       "4           chr1:925923  [\"G\",\"C\"]  ENSG00000187634  other_missense  c.-19G>C   \n",
       "...                 ...        ...              ...             ...       ...   \n",
       "9867259  chr22:50782243  [\"C\",\"T\"]  ENSG00000079974  other_missense   c.52G>A   \n",
       "9867260  chr22:50782243  [\"C\",\"T\"]  ENSG00000079974  other_missense   c.52G>A   \n",
       "9867261  chr22:50782290  [\"G\",\"C\"]  ENSG00000079974  other_missense    c.5C>G   \n",
       "9867262  chr22:50782290  [\"G\",\"C\"]  ENSG00000079974  other_missense    c.5C>G   \n",
       "9867263  chr22:50782290  [\"G\",\"C\"]  ENSG00000079974  other_missense    c.5C>G   \n",
       "\n",
       "              hgvsp   cadd  mpc           polyphen  \\\n",
       "0               NaN  22.30  NaN                NaN   \n",
       "1               NaN  22.30  NaN                NaN   \n",
       "2               NaN  22.30  NaN                NaN   \n",
       "3               NaN  20.20  NaN                NaN   \n",
       "4               NaN  20.20  NaN                NaN   \n",
       "...             ...    ...  ...                ...   \n",
       "9867259  p.Ala18Thr  18.69  NaN             benign   \n",
       "9867260  p.Ala18Thr  18.69  NaN             benign   \n",
       "9867261   p.Ala2Gly  24.10  NaN  possibly_damaging   \n",
       "9867262   p.Ala2Gly  24.10  NaN  possibly_damaging   \n",
       "9867263   p.Ala2Gly  24.10  NaN  possibly_damaging   \n",
       "\n",
       "                                                group  ... ac_ctrl  an_ctrl  \\\n",
       "0                                    Bipolar Disorder  ...       0    28844   \n",
       "1        Bipolar Disorder (including Schizoaffective)  ...       0    28844   \n",
       "2                                  Bipolar Disorder 1  ...       0    28844   \n",
       "3                                    Bipolar Disorder  ...       1    28844   \n",
       "4        Bipolar Disorder (including Schizoaffective)  ...       1    28844   \n",
       "...                                               ...  ...     ...      ...   \n",
       "9867259               Bipolar Disorder with Psychosis  ...      15    28844   \n",
       "9867260            Bipolar Disorder without Psychosis  ...      15    28844   \n",
       "9867261                              Bipolar Disorder  ...       0    28844   \n",
       "9867262  Bipolar Disorder (including Schizoaffective)  ...       0    28844   \n",
       "9867263                            Bipolar Disorder 1  ...       0    28844   \n",
       "\n",
       "         estimate  chi_sq_stat  p_value  in_analysis  in_gnomad_non_neuro  \\\n",
       "0             NaN          NaN      NaN         True                False   \n",
       "1             NaN          NaN      NaN         True                False   \n",
       "2             NaN          NaN      NaN         True                False   \n",
       "3             NaN          NaN      NaN         True                False   \n",
       "4             NaN          NaN      NaN         True                False   \n",
       "...           ...          ...      ...          ...                  ...   \n",
       "9867259   0.31788      0.43477  0.50966        False                 True   \n",
       "9867260   0.38228      0.45574  0.49962        False                 True   \n",
       "9867261       NaN          NaN      NaN         True                False   \n",
       "9867262       NaN          NaN      NaN         True                False   \n",
       "9867263       NaN          NaN      NaN         True                False   \n",
       "\n",
       "           chr       pos  gene_name  \n",
       "0         chr1    925922     SAMD11  \n",
       "1         chr1    925922     SAMD11  \n",
       "2         chr1    925922     SAMD11  \n",
       "3         chr1    925923     SAMD11  \n",
       "4         chr1    925923     SAMD11  \n",
       "...        ...       ...        ...  \n",
       "9867259  chr22  50782243     RABL2B  \n",
       "9867260  chr22  50782243     RABL2B  \n",
       "9867261  chr22  50782290     RABL2B  \n",
       "9867262  chr22  50782290     RABL2B  \n",
       "9867263  chr22  50782290     RABL2B  \n",
       "\n",
       "[9867264 rows x 23 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the data into a DataFrame\n",
    "bipex_variant_results_df = pd.read_csv(bipex_variants_file_path, sep='\\t', compression='gzip')\n",
    "\n",
    "# Split locus into chr and pos and convert to integers\n",
    "bipex_variant_results_df[['chr', 'pos']] = bipex_variant_results_df['locus'].str.split(':', expand=True)\n",
    "bipex_variant_results_df['pos'] = bipex_variant_results_df['pos'].astype(int)\n",
    "\n",
    "# Filter out rows\n",
    "bipex_variant_results_df = bipex_variant_results_df[(bipex_variant_results_df['chr'] != \"chrX\") & (bipex_variant_results_df['chr'] != \"chrY\") & (bipex_variant_results_df['chr'] != \"chrMT\")]\n",
    "\n",
    "filter_consequence_list = [\n",
    "    \"synonymous\",\n",
    "    \"non_coding\",\n",
    "    \"NA\"\n",
    "]\n",
    "\n",
    "# Filter rows\n",
    "bipex_variant_results_df = bipex_variant_results_df[~bipex_variant_results_df['consequence'].isin(filter_consequence_list)]\n",
    "bipex_variant_results_df = bipex_variant_results_df[~bipex_variant_results_df['consequence'].isna()]\n",
    "\n",
    "# Add gene names\n",
    "bipex_variant_results_df = pd.merge(bipex_variant_results_df, gene_df[['std_gene_id', 'gene_name']].drop_duplicates(), left_on='gene_id', right_on='std_gene_id', how='left').drop('std_gene_id', axis=1)\n",
    "\n",
    "bipex_variant_results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbf5bd99",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define path to predictions file\n",
    "rgc_predictions_file_path = results_path + 'HMM_rgc_0.9_over20_chr2_predictions_rgc_wgs.tsv.gz'\n",
    "\n",
    "# Load in saved predictions for rgc dataset\n",
    "predictions_df = pd.read_csv(rgc_predictions_file_path, sep='\\t')\n",
    "\n",
    "predictions_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13448208",
   "metadata": {},
   "source": [
    "## Bipolar pLoF Analysis using Fisher's Exact Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a85b8cb0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------------------------------------------------------------------\n",
      "PROCESSING CHROMOSOME 1\n",
      "                  locus     alleles          gene_id consequence      hgvsc  \\\n",
      "93          chr1:930221   [\"C\",\"T\"]  ENSG00000187634         ptv   c.139C>T   \n",
      "94          chr1:930221   [\"C\",\"T\"]  ENSG00000187634         ptv   c.139C>T   \n",
      "95          chr1:930221   [\"C\",\"T\"]  ENSG00000187634         ptv   c.139C>T   \n",
      "323         chr1:939069  [\"T\",\"TC\"]  ENSG00000187634         ptv   c.466dup   \n",
      "324         chr1:939069  [\"T\",\"TC\"]  ENSG00000187634         ptv   c.466dup   \n",
      "...                 ...         ...              ...         ...        ...   \n",
      "1031065  chr1:248918016   [\"C\",\"T\"]  ENSG00000185220         ptv  c.1432C>T   \n",
      "1031066  chr1:248918016   [\"C\",\"T\"]  ENSG00000185220         ptv  c.1432C>T   \n",
      "1031067  chr1:248918016   [\"C\",\"T\"]  ENSG00000185220         ptv  c.1432C>T   \n",
      "1031068  chr1:248918016   [\"C\",\"T\"]  ENSG00000185220         ptv  c.1432C>T   \n",
      "1031069  chr1:248918016   [\"C\",\"T\"]  ENSG00000185220         ptv  c.1432C>T   \n",
      "\n",
      "                      hgvsp  cadd  mpc polyphen  \\\n",
      "93               p.Arg47Ter  37.0  NaN      NaN   \n",
      "94               p.Arg47Ter  37.0  NaN      NaN   \n",
      "95               p.Arg47Ter  37.0  NaN      NaN   \n",
      "323      p.Arg156ProfsTer51  23.0  NaN      NaN   \n",
      "324      p.Arg156ProfsTer51  23.0  NaN      NaN   \n",
      "...                     ...   ...  ...      ...   \n",
      "1031065         p.Arg478Ter  36.0  NaN      NaN   \n",
      "1031066         p.Arg478Ter  36.0  NaN      NaN   \n",
      "1031067         p.Arg478Ter  36.0  NaN      NaN   \n",
      "1031068         p.Arg478Ter  36.0  NaN      NaN   \n",
      "1031069         p.Arg478Ter  36.0  NaN      NaN   \n",
      "\n",
      "                                                group  ... p_value  \\\n",
      "93                                   Bipolar Disorder  ...     NaN   \n",
      "94       Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "95                                 Bipolar Disorder 1  ...     NaN   \n",
      "323                                  Bipolar Disorder  ...     NaN   \n",
      "324      Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "...                                               ...  ...     ...   \n",
      "1031065  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "1031066                            Bipolar Disorder 1  ...     NaN   \n",
      "1031067                            Bipolar Disorder 2  ...     NaN   \n",
      "1031068               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "1031069            Bipolar Disorder without Psychosis  ...     NaN   \n",
      "\n",
      "         in_analysis  in_gnomad_non_neuro   chr        pos  gene_name  \\\n",
      "93              True                False  chr1     930221     SAMD11   \n",
      "94              True                False  chr1     930221     SAMD11   \n",
      "95              True                False  chr1     930221     SAMD11   \n",
      "323             True                 True  chr1     939069     SAMD11   \n",
      "324             True                 True  chr1     939069     SAMD11   \n",
      "...              ...                  ...   ...        ...        ...   \n",
      "1031065         True                 True  chr1  248918016      PGBD2   \n",
      "1031066         True                 True  chr1  248918016      PGBD2   \n",
      "1031067         True                 True  chr1  248918016      PGBD2   \n",
      "1031068         True                 True  chr1  248918016      PGBD2   \n",
      "1031069         True                 True  chr1  248918016      PGBD2   \n",
      "\n",
      "         total_ac_case  total_an_case  total_ac_ctrl  total_an_ctrl  \n",
      "93                  20          27864             12          28844  \n",
      "94                  20          28418             12          28844  \n",
      "95                  14          16476             12          28844  \n",
      "323                 20          27864             12          28844  \n",
      "324                 20          28418             12          28844  \n",
      "...                ...            ...            ...            ...  \n",
      "1031065              9          28420              5          28844  \n",
      "1031066              6          16476              5          28844  \n",
      "1031067              2           6892              5          28844  \n",
      "1031068              2           8810              5          28844  \n",
      "1031069              1           7624              5          28844  \n",
      "\n",
      "[68529 rows x 27 columns]\n",
      "                0         1\n",
      "93       1.725806  0.157328\n",
      "94       1.692138  0.159874\n",
      "95       2.043332  0.068969\n",
      "326      1.395277  0.529741\n",
      "327      2.456994  0.065599\n",
      "...           ...       ...\n",
      "1030605  1.827116  0.298258\n",
      "1030657  2.101202  0.223090\n",
      "1030721  1.674253  0.626695\n",
      "1030722  1.309673  0.669063\n",
      "1030723  0.756631  1.000000\n",
      "\n",
      "[10890 rows x 2 columns]\n",
      "Empty DataFrame\n",
      "Columns: [gene_id, gene_name, group, total_ac_case, total_an_case, total_ac_ctrl, total_an_ctrl, odds_ratio, p_value]\n",
      "Index: []\n",
      "----------------------------------------------------------------------------------------------------\n",
      "PROCESSING CHROMOSOME 2\n",
      "                  locus    alleles          gene_id consequence     hgvsc  \\\n",
      "1031179      chr2:41613  [\"G\",\"A\"]  ENSG00000184731         ptv  c.961C>T   \n",
      "1031180      chr2:41613  [\"G\",\"A\"]  ENSG00000184731         ptv  c.961C>T   \n",
      "1031181      chr2:41613  [\"G\",\"A\"]  ENSG00000184731         ptv  c.961C>T   \n",
      "1031182      chr2:41613  [\"G\",\"A\"]  ENSG00000184731         ptv  c.961C>T   \n",
      "1031183      chr2:41613  [\"G\",\"A\"]  ENSG00000184731         ptv  c.961C>T   \n",
      "...                 ...        ...              ...         ...       ...   \n",
      "1757161  chr2:241897282  [\"G\",\"T\"]  ENSG00000216921         ptv   c.65C>A   \n",
      "1757162  chr2:241897282  [\"G\",\"T\"]  ENSG00000216921         ptv   c.65C>A   \n",
      "1757163  chr2:241897282  [\"G\",\"T\"]  ENSG00000216921         ptv   c.65C>A   \n",
      "1757164  chr2:241897282  [\"G\",\"T\"]  ENSG00000216921         ptv   c.65C>A   \n",
      "1757165  chr2:241897282  [\"G\",\"T\"]  ENSG00000216921         ptv   c.65C>A   \n",
      "\n",
      "               hgvsp  cadd  mpc polyphen  \\\n",
      "1031179  p.Arg321Ter  35.0  NaN      NaN   \n",
      "1031180  p.Arg321Ter  35.0  NaN      NaN   \n",
      "1031181  p.Arg321Ter  35.0  NaN      NaN   \n",
      "1031182  p.Arg321Ter  35.0  NaN      NaN   \n",
      "1031183  p.Arg321Ter  35.0  NaN      NaN   \n",
      "...              ...   ...  ...      ...   \n",
      "1757161   p.Ser22Ter  35.0  NaN      NaN   \n",
      "1757162   p.Ser22Ter  35.0  NaN      NaN   \n",
      "1757163   p.Ser22Ter  35.0  NaN      NaN   \n",
      "1757164   p.Ser22Ter  35.0  NaN      NaN   \n",
      "1757165   p.Ser22Ter  35.0  NaN      NaN   \n",
      "\n",
      "                                                group  ... p_value  \\\n",
      "1031179                              Bipolar Disorder  ...     NaN   \n",
      "1031180  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "1031181                            Bipolar Disorder 1  ...     NaN   \n",
      "1031182                            Bipolar Disorder 2  ...     NaN   \n",
      "1031183               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "...                                               ...  ...     ...   \n",
      "1757161  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "1757162                            Bipolar Disorder 1  ...     NaN   \n",
      "1757163                            Bipolar Disorder 2  ...     NaN   \n",
      "1757164               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "1757165            Bipolar Disorder without Psychosis  ...     NaN   \n",
      "\n",
      "         in_analysis  in_gnomad_non_neuro   chr        pos  gene_name  \\\n",
      "1031179         True                 True  chr2      41613    FAM110C   \n",
      "1031180         True                 True  chr2      41613    FAM110C   \n",
      "1031181         True                 True  chr2      41613    FAM110C   \n",
      "1031182         True                 True  chr2      41613    FAM110C   \n",
      "1031183         True                 True  chr2      41613    FAM110C   \n",
      "...              ...                  ...   ...        ...        ...   \n",
      "1757161         True                False  chr2  241897282    FAM240C   \n",
      "1757162         True                False  chr2  241897282    FAM240C   \n",
      "1757163         True                False  chr2  241897282    FAM240C   \n",
      "1757164         True                False  chr2  241897282    FAM240C   \n",
      "1757165         True                False  chr2  241897282    FAM240C   \n",
      "\n",
      "         total_ac_case  total_an_case  total_ac_ctrl  total_an_ctrl  \n",
      "1031179              6          27866              1          28844  \n",
      "1031180              6          28420              1          28844  \n",
      "1031181              3          16476              1          28844  \n",
      "1031182              2           6892              1          28842  \n",
      "1031183              2           8796              1          28824  \n",
      "...                ...            ...            ...            ...  \n",
      "1757161              1          28392              3          28844  \n",
      "1757162              1          16468              3          28844  \n",
      "1757163              0           6892              3          28844  \n",
      "1757164              1           8806              3          28844  \n",
      "1757165              0           7622              3          28844  \n",
      "\n",
      "[47526 rows x 27 columns]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                0         1\n",
      "1031179  6.211701  0.065825\n",
      "1031180  6.090589  0.068319\n",
      "1031181  5.252777  0.139782\n",
      "1031182  8.371843  0.097239\n",
      "1031183  6.555151  0.138432\n",
      "...           ...       ...\n",
      "1757057  0.338617  0.625034\n",
      "1757058  0.583814  1.000000\n",
      "1757059  0.000000  1.000000\n",
      "1757060  1.091842  1.000000\n",
      "1757061  0.000000  1.000000\n",
      "\n",
      "[6937 rows x 2 columns]\n",
      "Empty DataFrame\n",
      "Columns: [gene_id, gene_name, group, total_ac_case, total_an_case, total_ac_ctrl, total_an_ctrl, odds_ratio, p_value]\n",
      "Index: []\n",
      "----------------------------------------------------------------------------------------------------\n",
      "PROCESSING CHROMOSOME 3\n",
      "                  locus    alleles          gene_id consequence       hgvsc  \\\n",
      "1757800     chr3:366021  [\"G\",\"T\"]  ENSG00000134121         ptv   c.1657G>T   \n",
      "1757801     chr3:366021  [\"G\",\"T\"]  ENSG00000134121         ptv   c.1657G>T   \n",
      "1757802     chr3:366021  [\"G\",\"T\"]  ENSG00000134121         ptv   c.1657G>T   \n",
      "1757803     chr3:366021  [\"G\",\"T\"]  ENSG00000134121         ptv   c.1657G>T   \n",
      "1758022     chr3:382565  [\"A\",\"T\"]  ENSG00000234661         ptv  n.259+2T>A   \n",
      "...                 ...        ...              ...         ...         ...   \n",
      "2318610  chr3:198036028  [\"C\",\"T\"]  ENSG00000185621         ptv   c.1963C>T   \n",
      "2318611  chr3:198036028  [\"C\",\"T\"]  ENSG00000185621         ptv   c.1963C>T   \n",
      "2318612  chr3:198036028  [\"C\",\"T\"]  ENSG00000185621         ptv   c.1963C>T   \n",
      "2318613  chr3:198036028  [\"C\",\"T\"]  ENSG00000185621         ptv   c.1963C>T   \n",
      "2318614  chr3:198036028  [\"C\",\"T\"]  ENSG00000185621         ptv   c.1963C>T   \n",
      "\n",
      "               hgvsp  cadd       mpc polyphen  \\\n",
      "1757800  p.Glu553Ter  41.0       NaN      NaN   \n",
      "1757801  p.Glu553Ter  41.0       NaN      NaN   \n",
      "1757802  p.Glu553Ter  41.0       NaN      NaN   \n",
      "1757803  p.Glu553Ter  41.0       NaN      NaN   \n",
      "1758022          NaN  22.1  0.081987      NaN   \n",
      "...              ...   ...       ...      ...   \n",
      "2318610  p.Arg655Ter  39.0       NaN      NaN   \n",
      "2318611  p.Arg655Ter  39.0       NaN      NaN   \n",
      "2318612  p.Arg655Ter  39.0       NaN      NaN   \n",
      "2318613  p.Arg655Ter  39.0       NaN      NaN   \n",
      "2318614  p.Arg655Ter  39.0       NaN      NaN   \n",
      "\n",
      "                                                group  ... p_value  \\\n",
      "1757800                              Bipolar Disorder  ...     NaN   \n",
      "1757801  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "1757802                            Bipolar Disorder 2  ...     NaN   \n",
      "1757803            Bipolar Disorder without Psychosis  ...     NaN   \n",
      "1758022                              Bipolar Disorder  ...     NaN   \n",
      "...                                               ...  ...     ...   \n",
      "2318610  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "2318611                            Bipolar Disorder 1  ...     NaN   \n",
      "2318612                            Bipolar Disorder 2  ...     NaN   \n",
      "2318613               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "2318614            Bipolar Disorder without Psychosis  ...     NaN   \n",
      "\n",
      "         in_analysis  in_gnomad_non_neuro   chr        pos  gene_name  \\\n",
      "1757800         True                False  chr3     366021       CHL1   \n",
      "1757801         True                False  chr3     366021       CHL1   \n",
      "1757802         True                False  chr3     366021       CHL1   \n",
      "1757803         True                False  chr3     366021       CHL1   \n",
      "1758022         True                False  chr3     382565        NaN   \n",
      "...              ...                  ...   ...        ...        ...   \n",
      "2318610         True                 True  chr3  198036028       LMLN   \n",
      "2318611         True                 True  chr3  198036028       LMLN   \n",
      "2318612         True                 True  chr3  198036028       LMLN   \n",
      "2318613         True                 True  chr3  198036028       LMLN   \n",
      "2318614         True                 True  chr3  198036028       LMLN   \n",
      "\n",
      "         total_ac_case  total_an_case  total_ac_ctrl  total_an_ctrl  \n",
      "1757800              3          27866              5          28844  \n",
      "1757801              3          28420              5          28844  \n",
      "1757802              1           6892              5          28844  \n",
      "1757803              1           7624              5          28844  \n",
      "1758022              0          27864              1          28844  \n",
      "...                ...            ...            ...            ...  \n",
      "2318610             11          28420              9          28844  \n",
      "2318611              7          16476              9          28844  \n",
      "2318612              1           6892              9          28844  \n",
      "2318613              3           8810              9          28844  \n",
      "2318614              2           7624              9          28844  \n",
      "\n",
      "[37873 rows x 27 columns]\n",
      "                0         1\n",
      "1757800  0.621017  0.726868\n",
      "1757801  0.608910  0.726603\n",
      "1757802  0.837005  1.000000\n",
      "1757803  0.756631  1.000000\n",
      "1758022  0.000000  1.000000\n",
      "...           ...       ...\n",
      "2318176  1.240550  0.661366\n",
      "2318177  1.361784  0.605774\n",
      "2318178  1.091367  1.000000\n",
      "2318179  0.840695  1.000000\n",
      "2318319  0.464938  0.698457\n",
      "\n",
      "[5809 rows x 2 columns]\n",
      "Empty DataFrame\n",
      "Columns: [gene_id, gene_name, group, total_ac_case, total_an_case, total_ac_ctrl, total_an_ctrl, odds_ratio, p_value]\n",
      "Index: []\n",
      "----------------------------------------------------------------------------------------------------\n",
      "PROCESSING CHROMOSOME 4\n",
      "                  locus    alleles          gene_id consequence       hgvsc  \\\n",
      "2318652      chr4:59558  [\"T\",\"A\"]  ENSG00000272602         ptv  c.130+2T>A   \n",
      "2318653      chr4:59558  [\"T\",\"A\"]  ENSG00000272602         ptv  c.130+2T>A   \n",
      "2318654      chr4:59558  [\"T\",\"A\"]  ENSG00000272602         ptv  c.130+2T>A   \n",
      "2318655      chr4:59558  [\"T\",\"A\"]  ENSG00000272602         ptv  c.130+2T>A   \n",
      "2318656      chr4:59558  [\"T\",\"A\"]  ENSG00000272602         ptv  c.130+2T>A   \n",
      "...                 ...        ...              ...         ...         ...   \n",
      "2691656  chr4:190025860  [\"G\",\"A\"]  ENSG00000205097         ptv    c.541C>T   \n",
      "2691657  chr4:190025860  [\"G\",\"A\"]  ENSG00000205097         ptv    c.541C>T   \n",
      "2691658  chr4:190025860  [\"G\",\"A\"]  ENSG00000205097         ptv    c.541C>T   \n",
      "2691659  chr4:190025860  [\"G\",\"A\"]  ENSG00000205097         ptv    c.541C>T   \n",
      "2691660  chr4:190025860  [\"G\",\"A\"]  ENSG00000205097         ptv    c.541C>T   \n",
      "\n",
      "               hgvsp  cadd  mpc polyphen  \\\n",
      "2318652          NaN  21.8  NaN      NaN   \n",
      "2318653          NaN  21.8  NaN      NaN   \n",
      "2318654          NaN  21.8  NaN      NaN   \n",
      "2318655          NaN  21.8  NaN      NaN   \n",
      "2318656          NaN  21.8  NaN      NaN   \n",
      "...              ...   ...  ...      ...   \n",
      "2691656  p.Arg181Ter  22.0  NaN      NaN   \n",
      "2691657  p.Arg181Ter  22.0  NaN      NaN   \n",
      "2691658  p.Arg181Ter  22.0  NaN      NaN   \n",
      "2691659  p.Arg181Ter  22.0  NaN      NaN   \n",
      "2691660  p.Arg181Ter  22.0  NaN      NaN   \n",
      "\n",
      "                                                group  ... p_value  \\\n",
      "2318652                              Bipolar Disorder  ...     NaN   \n",
      "2318653  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "2318654                            Bipolar Disorder 2  ...     NaN   \n",
      "2318655               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "2318656            Bipolar Disorder without Psychosis  ...     NaN   \n",
      "...                                               ...  ...     ...   \n",
      "2691656  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "2691657                            Bipolar Disorder 1  ...     NaN   \n",
      "2691658                            Bipolar Disorder 2  ...     NaN   \n",
      "2691659               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "2691660            Bipolar Disorder without Psychosis  ...     NaN   \n",
      "\n",
      "         in_analysis  in_gnomad_non_neuro   chr        pos  gene_name  \\\n",
      "2318652         True                False  chr4      59558     ZNF595   \n",
      "2318653         True                False  chr4      59558     ZNF595   \n",
      "2318654         True                False  chr4      59558     ZNF595   \n",
      "2318655         True                False  chr4      59558     ZNF595   \n",
      "2318656         True                False  chr4      59558     ZNF595   \n",
      "...              ...                  ...   ...        ...        ...   \n",
      "2691656         True                 True  chr4  190025860       FRG2   \n",
      "2691657         True                 True  chr4  190025860       FRG2   \n",
      "2691658         True                 True  chr4  190025860       FRG2   \n",
      "2691659         True                 True  chr4  190025860       FRG2   \n",
      "2691660         True                 True  chr4  190025860       FRG2   \n",
      "\n",
      "         total_ac_case  total_an_case  total_ac_ctrl  total_an_ctrl  \n",
      "2318652              9          27866              5          28844  \n",
      "2318653              9          28420              5          28844  \n",
      "2318654              2           6892              5          28844  \n",
      "2318655              6           8810             10          28844  \n",
      "2318656              1           7624              5          28844  \n",
      "...                ...            ...            ...            ...  \n",
      "2691656              3          28420              3          28844  \n",
      "2691657              1          16476              3          28844  \n",
      "2691658              1           6892              3          28844  \n",
      "2691659              0           8810              3          28844  \n",
      "2691660              2           7624              3          28844  \n",
      "\n",
      "[25446 rows x 27 columns]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                0         1\n",
      "2318652  1.863453  0.294225\n",
      "2318653  1.827116  0.298258\n",
      "2318654  1.674253  0.626695\n",
      "2318655  1.965061  0.232321\n",
      "2318656  0.756631  1.000000\n",
      "...           ...       ...\n",
      "2691555  1.014921  1.000000\n",
      "2691556  0.583531  1.000000\n",
      "2691557  1.395105  0.575595\n",
      "2691558  0.000000  1.000000\n",
      "2691559  2.522610  0.281368\n",
      "\n",
      "[3964 rows x 2 columns]\n",
      "Empty DataFrame\n",
      "Columns: [gene_id, gene_name, group, total_ac_case, total_an_case, total_ac_ctrl, total_an_ctrl, odds_ratio, p_value]\n",
      "Index: []\n",
      "----------------------------------------------------------------------------------------------------\n",
      "PROCESSING CHROMOSOME 5\n",
      "                  locus      alleles          gene_id consequence  \\\n",
      "2691898     chr5:140464    [\"C\",\"T\"]  ENSG00000153404         ptv   \n",
      "2691899     chr5:140464    [\"C\",\"T\"]  ENSG00000153404         ptv   \n",
      "2691900     chr5:140464    [\"C\",\"T\"]  ENSG00000153404         ptv   \n",
      "2691998     chr5:140676  [\"CTG\",\"C\"]  ENSG00000153404         ptv   \n",
      "2691999     chr5:140676  [\"CTG\",\"C\"]  ENSG00000153404         ptv   \n",
      "...                 ...          ...              ...         ...   \n",
      "3125871  chr5:181260671    [\"C\",\"T\"]  ENSG00000183718         ptv   \n",
      "3125872  chr5:181260671    [\"C\",\"T\"]  ENSG00000183718         ptv   \n",
      "3125873  chr5:181260671    [\"C\",\"T\"]  ENSG00000183718         ptv   \n",
      "3125874  chr5:181260671    [\"C\",\"T\"]  ENSG00000183718         ptv   \n",
      "3125875  chr5:181260671    [\"C\",\"T\"]  ENSG00000183718         ptv   \n",
      "\n",
      "                hgvsc                hgvsp   cadd  mpc polyphen  \\\n",
      "2691898      c.157C>T           p.Gln53Ter  28.90  NaN      NaN   \n",
      "2691899      c.157C>T           p.Gln53Ter  28.90  NaN      NaN   \n",
      "2691900      c.157C>T           p.Gln53Ter  28.90  NaN      NaN   \n",
      "2691998  c.373_374del  p.Val125ProfsTer110  15.29  NaN      NaN   \n",
      "2691999  c.373_374del  p.Val125ProfsTer110  15.29  NaN      NaN   \n",
      "...               ...                  ...    ...  ...      ...   \n",
      "3125871      c.143G>A           p.Trp48Ter  35.00  NaN      NaN   \n",
      "3125872      c.143G>A           p.Trp48Ter  35.00  NaN      NaN   \n",
      "3125873      c.143G>A           p.Trp48Ter  35.00  NaN      NaN   \n",
      "3125874      c.143G>A           p.Trp48Ter  35.00  NaN      NaN   \n",
      "3125875      c.143G>A           p.Trp48Ter  35.00  NaN      NaN   \n",
      "\n",
      "                                                group  ... p_value  \\\n",
      "2691898                              Bipolar Disorder  ...     NaN   \n",
      "2691899  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "2691900                            Bipolar Disorder 2  ...     NaN   \n",
      "2691998                              Bipolar Disorder  ...     NaN   \n",
      "2691999  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "...                                               ...  ...     ...   \n",
      "3125871  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "3125872                            Bipolar Disorder 1  ...     NaN   \n",
      "3125873                            Bipolar Disorder 2  ...     NaN   \n",
      "3125874               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "3125875            Bipolar Disorder without Psychosis  ...     NaN   \n",
      "\n",
      "         in_analysis  in_gnomad_non_neuro   chr        pos  gene_name  \\\n",
      "2691898         True                False  chr5     140464   PLEKHG4B   \n",
      "2691899         True                False  chr5     140464   PLEKHG4B   \n",
      "2691900         True                False  chr5     140464   PLEKHG4B   \n",
      "2691998         True                 True  chr5     140676   PLEKHG4B   \n",
      "2691999         True                 True  chr5     140676   PLEKHG4B   \n",
      "...              ...                  ...   ...        ...        ...   \n",
      "3125871         True                False  chr5  181260671     TRIM52   \n",
      "3125872         True                False  chr5  181260671     TRIM52   \n",
      "3125873         True                False  chr5  181260671     TRIM52   \n",
      "3125874         True                False  chr5  181260671     TRIM52   \n",
      "3125875         True                False  chr5  181260671     TRIM52   \n",
      "\n",
      "         total_ac_case  total_an_case  total_ac_ctrl  total_an_ctrl  \n",
      "2691898             20          27866             20          28844  \n",
      "2691899             20          28420             20          28844  \n",
      "2691900              3           6892             20          28844  \n",
      "2691998             20          27866             20          28844  \n",
      "2691999             20          28420             20          28844  \n",
      "...                ...            ...            ...            ...  \n",
      "3125871              2          28420              6          28844  \n",
      "3125872              2          16476              6          28844  \n",
      "3125873              0           6892              6          28844  \n",
      "3125874              2           8810              6          28844  \n",
      "3125875              0           7624              6          28844  \n",
      "\n",
      "[28699 rows x 27 columns]\n",
      "                0         1\n",
      "2691898  1.035122  1.000000\n",
      "2691899  1.014930  1.000000\n",
      "2691900  0.627609  0.601076\n",
      "2692000  1.050437  0.856758\n",
      "2692002  0.916656  1.000000\n",
      "...           ...       ...\n",
      "3125626  0.338260  0.289136\n",
      "3125627  0.583505  0.718978\n",
      "3125628  0.000000  0.603567\n",
      "3125629  1.091356  1.000000\n",
      "3125630  0.000000  0.355110\n",
      "\n",
      "[4790 rows x 2 columns]\n",
      "Empty DataFrame\n",
      "Columns: [gene_id, gene_name, group, total_ac_case, total_an_case, total_ac_ctrl, total_an_ctrl, odds_ratio, p_value]\n",
      "Index: []\n",
      "----------------------------------------------------------------------------------------------------\n",
      "PROCESSING CHROMOSOME 6\n",
      "                  locus    alleles          gene_id consequence       hgvsc  \\\n",
      "3125991     chr6:335112  [\"A\",\"G\"]  ENSG00000112679         ptv  c.139-2A>G   \n",
      "3125992     chr6:335112  [\"A\",\"G\"]  ENSG00000112679         ptv  c.139-2A>G   \n",
      "3125993     chr6:335112  [\"A\",\"G\"]  ENSG00000112679         ptv  c.139-2A>G   \n",
      "3125994     chr6:335112  [\"A\",\"G\"]  ENSG00000112679         ptv  c.139-2A>G   \n",
      "3125995     chr6:335112  [\"A\",\"G\"]  ENSG00000112679         ptv  c.139-2A>G   \n",
      "...                 ...        ...              ...         ...         ...   \n",
      "3616133  chr6:170583708  [\"G\",\"C\"]  ENSG00000071994         ptv    c.323C>G   \n",
      "3616134  chr6:170583708  [\"G\",\"C\"]  ENSG00000071994         ptv    c.323C>G   \n",
      "3616135  chr6:170583708  [\"G\",\"C\"]  ENSG00000071994         ptv    c.323C>G   \n",
      "3616136  chr6:170583708  [\"G\",\"C\"]  ENSG00000071994         ptv    c.323C>G   \n",
      "3616137  chr6:170583708  [\"G\",\"C\"]  ENSG00000071994         ptv    c.323C>G   \n",
      "\n",
      "               hgvsp  cadd  mpc polyphen  \\\n",
      "3125991          NaN  26.6  NaN      NaN   \n",
      "3125992          NaN  26.6  NaN      NaN   \n",
      "3125993          NaN  26.6  NaN      NaN   \n",
      "3125994          NaN  26.6  NaN      NaN   \n",
      "3125995          NaN  26.6  NaN      NaN   \n",
      "...              ...   ...  ...      ...   \n",
      "3616133  p.Ser108Ter  41.0  NaN      NaN   \n",
      "3616134  p.Ser108Ter  41.0  NaN      NaN   \n",
      "3616135  p.Ser108Ter  41.0  NaN      NaN   \n",
      "3616136  p.Ser108Ter  41.0  NaN      NaN   \n",
      "3616137  p.Ser108Ter  41.0  NaN      NaN   \n",
      "\n",
      "                                                group  ... p_value  \\\n",
      "3125991                              Bipolar Disorder  ...     NaN   \n",
      "3125992  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "3125993                            Bipolar Disorder 1  ...     NaN   \n",
      "3125994                            Bipolar Disorder 2  ...     NaN   \n",
      "3125995               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "...                                               ...  ...     ...   \n",
      "3616133  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "3616134                            Bipolar Disorder 1  ...     NaN   \n",
      "3616135                            Bipolar Disorder 2  ...     NaN   \n",
      "3616136               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "3616137            Bipolar Disorder without Psychosis  ...     NaN   \n",
      "\n",
      "         in_analysis  in_gnomad_non_neuro   chr        pos  gene_name  \\\n",
      "3125991         True                 True  chr6     335112     DUSP22   \n",
      "3125992         True                 True  chr6     335112     DUSP22   \n",
      "3125993         True                 True  chr6     335112     DUSP22   \n",
      "3125994         True                 True  chr6     335112     DUSP22   \n",
      "3125995         True                 True  chr6     335112     DUSP22   \n",
      "...              ...                  ...   ...        ...        ...   \n",
      "3616133         True                False  chr6  170583708      PDCD2   \n",
      "3616134         True                False  chr6  170583708      PDCD2   \n",
      "3616135         True                False  chr6  170583708      PDCD2   \n",
      "3616136         True                False  chr6  170583708      PDCD2   \n",
      "3616137         True                False  chr6  170583708      PDCD2   \n",
      "\n",
      "         total_ac_case  total_an_case  total_ac_ctrl  total_an_ctrl  \n",
      "3125991              5          27866              8          28844  \n",
      "3125992              5          28420              8          28844  \n",
      "3125993              3          16476              8          28844  \n",
      "3125994              2           6892              8          28844  \n",
      "3125995              2           8810              8          28844  \n",
      "...                ...            ...            ...            ...  \n",
      "3616133              2          28420              1          28844  \n",
      "3616134              2          16476              2          28844  \n",
      "3616135              4           6892              2          28844  \n",
      "3616136              3           8810              2          28844  \n",
      "3616137              4           7624              2          28844  \n",
      "\n",
      "[34638 rows x 27 columns]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                0         1\n",
      "3125991  0.646872  0.581758\n",
      "3125992  0.634260  0.581149\n",
      "3125993  0.656438  0.756122\n",
      "3125994  1.046299  1.000000\n",
      "3125995  0.818460  1.000000\n",
      "...           ...       ...\n",
      "3615939  2.029911  0.622237\n",
      "3615940  8.374564  0.014855\n",
      "3615941  7.570079  0.019895\n",
      "3616006  1.750759  0.625087\n",
      "3616008  4.912342  0.087325\n",
      "\n",
      "[5648 rows x 2 columns]\n",
      "                 gene_id gene_name  \\\n",
      "3210271  ENSG00000234127    TRIM26   \n",
      "3417042  ENSG00000083097     DOP1A   \n",
      "3417043  ENSG00000083097     DOP1A   \n",
      "\n",
      "                                                group  total_ac_case  \\\n",
      "3210271               Bipolar Disorder with Psychosis             12   \n",
      "3417042                              Bipolar Disorder             17   \n",
      "3417043  Bipolar Disorder (including Schizoaffective)             17   \n",
      "\n",
      "         total_an_case  total_ac_ctrl  total_an_ctrl  odds_ratio   p_value  \n",
      "3210271           8810              3          28844   13.112526  0.000006  \n",
      "3417042          27866              1          28844   17.606772  0.000060  \n",
      "3417043          28420              1          28844   17.263352  0.000069  \n",
      "----------------------------------------------------------------------------------------------------\n",
      "PROCESSING CHROMOSOME 7\n",
      "                  locus      alleles          gene_id consequence  \\\n",
      "3616295     chr7:246414    [\"G\",\"C\"]  ENSG00000177706         ptv   \n",
      "3616296     chr7:246414    [\"G\",\"C\"]  ENSG00000177706         ptv   \n",
      "3616297     chr7:246414    [\"G\",\"C\"]  ENSG00000177706         ptv   \n",
      "3616298     chr7:246414    [\"G\",\"C\"]  ENSG00000177706         ptv   \n",
      "3616320     chr7:246440  [\"CCT\",\"C\"]  ENSG00000177706         ptv   \n",
      "...                 ...          ...              ...         ...   \n",
      "4102773  chr7:159035985    [\"C\",\"T\"]  ENSG00000106018         ptv   \n",
      "4102774  chr7:159035985    [\"C\",\"T\"]  ENSG00000106018         ptv   \n",
      "4102775  chr7:159035985    [\"C\",\"T\"]  ENSG00000106018         ptv   \n",
      "4102865  chr7:159058482   [\"TG\",\"T\"]  ENSG00000106018         ptv   \n",
      "4102866  chr7:159058482   [\"TG\",\"T\"]  ENSG00000106018         ptv   \n",
      "\n",
      "                hgvsc               hgvsp  cadd  mpc polyphen  \\\n",
      "3616295    c.864-1G>C                 NaN  23.9  NaN      NaN   \n",
      "3616296    c.864-1G>C                 NaN  23.9  NaN      NaN   \n",
      "3616297    c.864-1G>C                 NaN  23.9  NaN      NaN   \n",
      "3616298    c.864-1G>C                 NaN  23.9  NaN      NaN   \n",
      "3616320  c.890_891del   p.Pro297ArgfsTer7  23.0  NaN      NaN   \n",
      "...               ...                 ...   ...  ...      ...   \n",
      "4102773      c.776G>A         p.Trp259Ter  37.0  NaN      NaN   \n",
      "4102774      c.776G>A         p.Trp259Ter  37.0  NaN      NaN   \n",
      "4102775      c.776G>A         p.Trp259Ter  37.0  NaN      NaN   \n",
      "4102865      c.453del  p.Phe151LeufsTer13  27.7  NaN      NaN   \n",
      "4102866      c.453del  p.Phe151LeufsTer13  27.7  NaN      NaN   \n",
      "\n",
      "                                                group  ... p_value  \\\n",
      "3616295                              Bipolar Disorder  ...     NaN   \n",
      "3616296  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "3616297                            Bipolar Disorder 1  ...     NaN   \n",
      "3616298               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "3616320                              Bipolar Disorder  ...     NaN   \n",
      "...                                               ...  ...     ...   \n",
      "4102773                            Bipolar Disorder 2  ...     NaN   \n",
      "4102774               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "4102775            Bipolar Disorder without Psychosis  ...     NaN   \n",
      "4102865                              Bipolar Disorder  ...     NaN   \n",
      "4102866  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "\n",
      "         in_analysis  in_gnomad_non_neuro   chr        pos  gene_name  \\\n",
      "3616295         True                False  chr7     246414     FAM20C   \n",
      "3616296         True                False  chr7     246414     FAM20C   \n",
      "3616297         True                False  chr7     246414     FAM20C   \n",
      "3616298         True                False  chr7     246414     FAM20C   \n",
      "3616320         True                False  chr7     246440     FAM20C   \n",
      "...              ...                  ...   ...        ...        ...   \n",
      "4102773         True                False  chr7  159035985      VIPR2   \n",
      "4102774         True                False  chr7  159035985      VIPR2   \n",
      "4102775         True                False  chr7  159035985      VIPR2   \n",
      "4102865         True                 True  chr7  159058482      VIPR2   \n",
      "4102866         True                 True  chr7  159058482      VIPR2   \n",
      "\n",
      "         total_ac_case  total_an_case  total_ac_ctrl  total_an_ctrl  \n",
      "3616295              4          27866              2          28844  \n",
      "3616296              4          28420              2          28844  \n",
      "3616297              4          16476              2          28844  \n",
      "3616298              3           8810              2          28844  \n",
      "3616320              4          27866              2          28844  \n",
      "...                ...            ...            ...            ...  \n",
      "4102773              0           6892              7          28844  \n",
      "4102774              1           8810              7          28844  \n",
      "4102775              1           7624              7          28844  \n",
      "4102865              5          27866              7          28844  \n",
      "4102866              5          28420              7          28844  \n",
      "\n",
      "[32537 rows x 27 columns]\n",
      "                0         1\n",
      "3616295  2.070347  0.445373\n",
      "3616296  2.029983  0.449698\n",
      "3616297  3.501943  0.199135\n",
      "3616298  4.912342  0.087325\n",
      "3616417  0.000000  1.000000\n",
      "...           ...       ...\n",
      "4102561  0.724894  0.774465\n",
      "4102562  0.750241  1.000000\n",
      "4102563  0.000000  0.359094\n",
      "4102564  0.467655  0.690257\n",
      "4102565  0.540413  1.000000\n",
      "\n",
      "[5148 rows x 2 columns]\n",
      "                 gene_id gene_name                               group  \\\n",
      "3810743  ENSG00000127951      FGL2  Bipolar Disorder without Psychosis   \n",
      "\n",
      "         total_ac_case  total_an_case  total_ac_ctrl  total_an_ctrl  \\\n",
      "3810743              6           7624              0          28844   \n",
      "\n",
      "         odds_ratio   p_value  \n",
      "3810743         inf  0.000083  \n",
      "----------------------------------------------------------------------------------------------------\n",
      "PROCESSING CHROMOSOME 8\n",
      "                  locus    alleles          gene_id consequence       hgvsc  \\\n",
      "4102980     chr8:240908  [\"G\",\"C\"]  ENSG00000172748         ptv   c.12+1G>C   \n",
      "4102981     chr8:240908  [\"G\",\"C\"]  ENSG00000172748         ptv   c.12+1G>C   \n",
      "4102982     chr8:240908  [\"G\",\"C\"]  ENSG00000172748         ptv   c.12+1G>C   \n",
      "4102983     chr8:240908  [\"G\",\"C\"]  ENSG00000172748         ptv   c.12+1G>C   \n",
      "4102984     chr8:240908  [\"G\",\"C\"]  ENSG00000172748         ptv   c.12+1G>C   \n",
      "...                 ...        ...              ...         ...         ...   \n",
      "4459573  chr8:145054017  [\"G\",\"A\"]  ENSG00000182307         ptv  c.551-1G>A   \n",
      "4459574  chr8:145054017  [\"G\",\"A\"]  ENSG00000182307         ptv  c.551-1G>A   \n",
      "4459575  chr8:145054017  [\"G\",\"A\"]  ENSG00000182307         ptv  c.551-1G>A   \n",
      "4459576  chr8:145054017  [\"G\",\"A\"]  ENSG00000182307         ptv  c.551-1G>A   \n",
      "4459577  chr8:145054017  [\"G\",\"A\"]  ENSG00000182307         ptv  c.551-1G>A   \n",
      "\n",
      "        hgvsp   cadd  mpc polyphen  \\\n",
      "4102980   NaN  21.40  NaN      NaN   \n",
      "4102981   NaN  21.40  NaN      NaN   \n",
      "4102982   NaN  21.40  NaN      NaN   \n",
      "4102983   NaN  21.40  NaN      NaN   \n",
      "4102984   NaN  21.40  NaN      NaN   \n",
      "...       ...    ...  ...      ...   \n",
      "4459573   NaN  14.99  NaN      NaN   \n",
      "4459574   NaN  14.99  NaN      NaN   \n",
      "4459575   NaN  14.99  NaN      NaN   \n",
      "4459576   NaN  14.99  NaN      NaN   \n",
      "4459577   NaN  14.99  NaN      NaN   \n",
      "\n",
      "                                                group  ... p_value  \\\n",
      "4102980                              Bipolar Disorder  ...     NaN   \n",
      "4102981  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "4102982                            Bipolar Disorder 1  ...     NaN   \n",
      "4102983                            Bipolar Disorder 2  ...     NaN   \n",
      "4102984               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "...                                               ...  ...     ...   \n",
      "4459573  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "4459574                            Bipolar Disorder 1  ...     NaN   \n",
      "4459575                            Bipolar Disorder 2  ...     NaN   \n",
      "4459576               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "4459577            Bipolar Disorder without Psychosis  ...     NaN   \n",
      "\n",
      "         in_analysis  in_gnomad_non_neuro   chr        pos  gene_name  \\\n",
      "4102980         True                 True  chr8     240908     ZNF596   \n",
      "4102981         True                 True  chr8     240908     ZNF596   \n",
      "4102982         True                 True  chr8     240908     ZNF596   \n",
      "4102983         True                 True  chr8     240908     ZNF596   \n",
      "4102984         True                 True  chr8     240908     ZNF596   \n",
      "...              ...                  ...   ...        ...        ...   \n",
      "4459573         True                False  chr8  145054017    C8orf33   \n",
      "4459574         True                False  chr8  145054017    C8orf33   \n",
      "4459575         True                False  chr8  145054017    C8orf33   \n",
      "4459576         True                False  chr8  145054017    C8orf33   \n",
      "4459577         True                False  chr8  145054017    C8orf33   \n",
      "\n",
      "         total_ac_case  total_an_case  total_ac_ctrl  total_an_ctrl  \n",
      "4102980              9          27866             14          28844  \n",
      "4102981              9          28420             14          28844  \n",
      "4102982              7          16476             14          28844  \n",
      "4102983              2           6892             19          28844  \n",
      "4102984              5           8810             14          28844  \n",
      "...                ...            ...            ...            ...  \n",
      "4459573              3          28420              1          28844  \n",
      "4459574              2          16476              1          28844  \n",
      "4459575              0           6892              1          28844  \n",
      "4459576              2           8810              1          28844  \n",
      "4459577              0           7624              1          28844  \n",
      "\n",
      "[22254 rows x 27 columns]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                0         1\n",
      "4102980  0.665311  0.406398\n",
      "4102981  0.652338  0.405077\n",
      "4102982  0.875281  1.000000\n",
      "4102983  0.440379  0.405161\n",
      "4102984  1.169384  0.786792\n",
      "...           ...       ...\n",
      "4459516  3.044973  0.371326\n",
      "4459552  3.501639  0.300399\n",
      "4459553  6.549273  0.138605\n",
      "4459575  0.000000  1.000000\n",
      "4459577  0.000000  1.000000\n",
      "\n",
      "[3610 rows x 2 columns]\n",
      "                 gene_id gene_name               group  total_ac_case  \\\n",
      "4335222  ENSG00000174417      TRHR  Bipolar Disorder 2             13   \n",
      "4309630  ENSG00000175305     CCNE2  Bipolar Disorder 2              7   \n",
      "\n",
      "         total_an_case  total_ac_ctrl  total_an_ctrl  odds_ratio   p_value  \n",
      "4335222           6892              6          28844    9.083055  0.000004  \n",
      "4309630           6892              0          28838         inf  0.000010  \n",
      "----------------------------------------------------------------------------------------------------\n",
      "PROCESSING CHROMOSOME 9\n",
      "                  locus                                            alleles  \\\n",
      "4459629      chr9:14817                                          [\"C\",\"T\"]   \n",
      "4459630      chr9:14817                                          [\"C\",\"T\"]   \n",
      "4459631      chr9:14817                                          [\"C\",\"T\"]   \n",
      "4459632      chr9:14817                                          [\"C\",\"T\"]   \n",
      "4459730      chr9:15092  [\"ATGACCAGCTTGTTGAAGAGATCCGACATCAAGTGCCCACCTTG...   \n",
      "...                 ...                                                ...   \n",
      "4881995  chr9:138176206                                          [\"G\",\"A\"]   \n",
      "4882000  chr9:138217231                                          [\"G\",\"C\"]   \n",
      "4882001  chr9:138217231                                          [\"G\",\"C\"]   \n",
      "4882002  chr9:138217231                                          [\"G\",\"C\"]   \n",
      "4882003  chr9:138217231                                          [\"G\",\"C\"]   \n",
      "\n",
      "                 gene_id consequence              hgvsc        hgvsp  cadd  \\\n",
      "4459629  ENSG00000181404         ptv          c.1388G>A  p.Trp463Ter  36.0   \n",
      "4459630  ENSG00000181404         ptv          c.1388G>A  p.Trp463Ter  36.0   \n",
      "4459631  ENSG00000181404         ptv          c.1388G>A  p.Trp463Ter  36.0   \n",
      "4459632  ENSG00000181404         ptv          c.1388G>A  p.Trp463Ter  36.0   \n",
      "4459730  ENSG00000181404         ptv  c.1196-14_1252del          NaN   NaN   \n",
      "...                  ...         ...                ...          ...   ...   \n",
      "4881995  ENSG00000159247         ptv        n.1434-1G>A          NaN  21.5   \n",
      "4882000  ENSG00000233013         ptv         n.163+1G>C          NaN  11.8   \n",
      "4882001  ENSG00000233013         ptv         n.163+1G>C          NaN  11.8   \n",
      "4882002  ENSG00000233013         ptv         n.163+1G>C          NaN  11.8   \n",
      "4882003  ENSG00000233013         ptv         n.163+1G>C          NaN  11.8   \n",
      "\n",
      "         mpc polyphen                                         group  ...  \\\n",
      "4459629  NaN      NaN                              Bipolar Disorder  ...   \n",
      "4459630  NaN      NaN  Bipolar Disorder (including Schizoaffective)  ...   \n",
      "4459631  NaN      NaN                            Bipolar Disorder 1  ...   \n",
      "4459632  NaN      NaN               Bipolar Disorder with Psychosis  ...   \n",
      "4459730  NaN      NaN                              Bipolar Disorder  ...   \n",
      "...      ...      ...                                           ...  ...   \n",
      "4881995  NaN      NaN            Bipolar Disorder without Psychosis  ...   \n",
      "4882000  NaN      NaN                              Bipolar Disorder  ...   \n",
      "4882001  NaN      NaN  Bipolar Disorder (including Schizoaffective)  ...   \n",
      "4882002  NaN      NaN                            Bipolar Disorder 2  ...   \n",
      "4882003  NaN      NaN            Bipolar Disorder without Psychosis  ...   \n",
      "\n",
      "        p_value  in_analysis  in_gnomad_non_neuro   chr        pos  gene_name  \\\n",
      "4459629     NaN         True                False  chr9      14817     WASHC1   \n",
      "4459630     NaN         True                False  chr9      14817     WASHC1   \n",
      "4459631     NaN         True                False  chr9      14817     WASHC1   \n",
      "4459632     NaN         True                False  chr9      14817     WASHC1   \n",
      "4459730     NaN         True                False  chr9      15092     WASHC1   \n",
      "...         ...          ...                  ...   ...        ...        ...   \n",
      "4881995     NaN         True                 True  chr9  138176206        NaN   \n",
      "4882000     NaN         True                False  chr9  138217231        NaN   \n",
      "4882001     NaN         True                False  chr9  138217231        NaN   \n",
      "4882002     NaN         True                False  chr9  138217231        NaN   \n",
      "4882003     NaN         True                False  chr9  138217231        NaN   \n",
      "\n",
      "         total_ac_case  total_an_case  total_ac_ctrl  total_an_ctrl  \n",
      "4459629              4          27864              9          28844  \n",
      "4459630              4          28418              9          28844  \n",
      "4459631              2          16476              9          28844  \n",
      "4459632              2           8810              9          28844  \n",
      "4459730              4          27864              9          28844  \n",
      "...                ...            ...            ...            ...  \n",
      "4881995              1           7624              2          28844  \n",
      "4882000              1          27866              0          28844  \n",
      "4882001              1          28420              0          28844  \n",
      "4882002              1           6892              0          28844  \n",
      "4882003              1           7624              0          28844  \n",
      "\n",
      "[26408 rows x 27 columns]\n",
      "                0         1\n",
      "4459629  0.459998  0.267728\n",
      "4459630  0.451030  0.266965\n",
      "4459631  0.388963  0.347638\n",
      "4459632  0.727495  1.000000\n",
      "4459733  0.929882  1.000000\n",
      "...           ...       ...\n",
      "4881938  1.891775  0.505210\n",
      "4882000       inf  0.491377\n",
      "4882001       inf  0.496298\n",
      "4882002       inf  0.192859\n",
      "4882003       inf  0.209060\n",
      "\n",
      "[4168 rows x 2 columns]\n",
      "Empty DataFrame\n",
      "Columns: [gene_id, gene_name, group, total_ac_case, total_an_case, total_ac_ctrl, total_an_ctrl, odds_ratio, p_value]\n",
      "Index: []\n",
      "----------------------------------------------------------------------------------------------------\n",
      "PROCESSING CHROMOSOME 10\n",
      "                   locus     alleles          gene_id consequence       hgvsc  \\\n",
      "4882027      chr10:47082  [\"T\",\"TC\"]  ENSG00000261456         ptv   c.1309dup   \n",
      "4882028      chr10:47082  [\"T\",\"TC\"]  ENSG00000261456         ptv   c.1309dup   \n",
      "4882029      chr10:47082  [\"T\",\"TC\"]  ENSG00000261456         ptv   c.1309dup   \n",
      "4882030      chr10:47082  [\"T\",\"TC\"]  ENSG00000261456         ptv   c.1309dup   \n",
      "4882031      chr10:47082  [\"T\",\"TC\"]  ENSG00000261456         ptv   c.1309dup   \n",
      "...                  ...         ...              ...         ...         ...   \n",
      "5264247  chr10:133625422   [\"G\",\"A\"]  ENSG00000225899         ptv    c.514C>T   \n",
      "5264248  chr10:133625422   [\"G\",\"A\"]  ENSG00000225899         ptv    c.514C>T   \n",
      "5264367  chr10:133626228   [\"C\",\"T\"]  ENSG00000225899         ptv  c.253+1G>A   \n",
      "5264368  chr10:133626228   [\"C\",\"T\"]  ENSG00000225899         ptv  c.253+1G>A   \n",
      "5264369  chr10:133626228   [\"C\",\"T\"]  ENSG00000225899         ptv  c.253+1G>A   \n",
      "\n",
      "                      hgvsp   cadd  mpc polyphen  \\\n",
      "4882027  p.Glu437GlyfsTer16  24.50  NaN      NaN   \n",
      "4882028  p.Glu437GlyfsTer16  24.50  NaN      NaN   \n",
      "4882029  p.Glu437GlyfsTer16  24.50  NaN      NaN   \n",
      "4882030  p.Glu437GlyfsTer16  24.50  NaN      NaN   \n",
      "4882031  p.Glu437GlyfsTer16  24.50  NaN      NaN   \n",
      "...                     ...    ...  ...      ...   \n",
      "5264247         p.Arg172Ter  22.30  NaN      NaN   \n",
      "5264248         p.Arg172Ter  22.30  NaN      NaN   \n",
      "5264367                 NaN  16.51  NaN      NaN   \n",
      "5264368                 NaN  16.51  NaN      NaN   \n",
      "5264369                 NaN  16.51  NaN      NaN   \n",
      "\n",
      "                                                group  ... p_value  \\\n",
      "4882027                              Bipolar Disorder  ...     NaN   \n",
      "4882028  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "4882029                            Bipolar Disorder 1  ...     NaN   \n",
      "4882030                            Bipolar Disorder 2  ...     NaN   \n",
      "4882031               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "...                                               ...  ...     ...   \n",
      "5264247               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "5264248            Bipolar Disorder without Psychosis  ...     NaN   \n",
      "5264367                            Bipolar Disorder 2  ...     NaN   \n",
      "5264368               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "5264369            Bipolar Disorder without Psychosis  ...     NaN   \n",
      "\n",
      "         in_analysis  in_gnomad_non_neuro    chr        pos  gene_name  \\\n",
      "4882027         True                 True  chr10      47082      TUBB8   \n",
      "4882028         True                 True  chr10      47082      TUBB8   \n",
      "4882029         True                 True  chr10      47082      TUBB8   \n",
      "4882030         True                 True  chr10      47082      TUBB8   \n",
      "4882031         True                 True  chr10      47082      TUBB8   \n",
      "...              ...                  ...    ...        ...        ...   \n",
      "5264247         True                 True  chr10  133625422      FRG2B   \n",
      "5264248         True                 True  chr10  133625422      FRG2B   \n",
      "5264367        False                False  chr10  133626228      FRG2B   \n",
      "5264368        False                False  chr10  133626228      FRG2B   \n",
      "5264369        False                False  chr10  133626228      FRG2B   \n",
      "\n",
      "         total_ac_case  total_an_case  total_ac_ctrl  total_an_ctrl  \n",
      "4882027              9          27866              5          28844  \n",
      "4882028             10          28420              5          28844  \n",
      "4882029              3          16476              5          28844  \n",
      "4882030              4           6892              5          28844  \n",
      "4882031              2           8810              5          28844  \n",
      "...                ...            ...            ...            ...  \n",
      "5264247              0           8810              6          28844  \n",
      "5264248              1           7624              6          28844  \n",
      "5264367              0           6892              6          28844  \n",
      "5264368              0           8810              6          28844  \n",
      "5264369              1           7624              6          28844  \n",
      "\n",
      "[25085 rows x 27 columns]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                0         1\n",
      "4882027  1.863453  0.294225\n",
      "4882028  2.030201  0.206841\n",
      "4882029  1.050410  1.000000\n",
      "4882030  3.349477  0.076477\n",
      "4882031  1.309673  0.669063\n",
      "...           ...       ...\n",
      "5264169  0.676590  1.000000\n",
      "5264170  0.583531  1.000000\n",
      "5264171  0.000000  0.603567\n",
      "5264172  0.000000  0.346925\n",
      "5264173  0.630504  1.000000\n",
      "\n",
      "[3979 rows x 2 columns]\n",
      "                 gene_id gene_name  \\\n",
      "4933872  ENSG00000223601     EBLN1   \n",
      "4933871  ENSG00000223601     EBLN1   \n",
      "4903026  ENSG00000165632      TAF3   \n",
      "\n",
      "                                                group  total_ac_case  \\\n",
      "4933872  Bipolar Disorder (including Schizoaffective)              3   \n",
      "4933871                              Bipolar Disorder              3   \n",
      "4903026                            Bipolar Disorder 2              8   \n",
      "\n",
      "         total_an_case  total_ac_ctrl  total_an_ctrl  odds_ratio   p_value  \n",
      "4933872          28420             24          28844    0.126773  0.000050  \n",
      "4933871          27866             24          28844    0.129293  0.000052  \n",
      "4903026           6892              2          28842   16.757699  0.000059  \n",
      "----------------------------------------------------------------------------------------------------\n",
      "PROCESSING CHROMOSOME 11\n",
      "                   locus     alleles          gene_id consequence      hgvsc  \\\n",
      "5264455     chr11:193155   [\"G\",\"A\"]  ENSG00000188076         ptv  c.55+1G>A   \n",
      "5264456     chr11:193155   [\"G\",\"A\"]  ENSG00000188076         ptv  c.55+1G>A   \n",
      "5264457     chr11:193155   [\"G\",\"A\"]  ENSG00000188076         ptv  c.55+1G>A   \n",
      "5264458     chr11:193155   [\"G\",\"A\"]  ENSG00000188076         ptv  c.55+1G>A   \n",
      "5264472     chr11:193754  [\"TC\",\"T\"]  ENSG00000188076         ptv   c.100del   \n",
      "...                  ...         ...              ...         ...        ...   \n",
      "5894555  chr11:134374697   [\"G\",\"A\"]  ENSG00000149328         ptv  c.1803G>A   \n",
      "5894559  chr11:134374716   [\"C\",\"T\"]  ENSG00000149328         ptv  c.1822C>T   \n",
      "5894560  chr11:134374716   [\"C\",\"T\"]  ENSG00000149328         ptv  c.1822C>T   \n",
      "5894590  chr11:134374997  [\"CG\",\"C\"]  ENSG00000149328         ptv  c.1853del   \n",
      "5894591  chr11:134374997  [\"CG\",\"C\"]  ENSG00000149328         ptv  c.1853del   \n",
      "\n",
      "                      hgvsp  cadd  mpc polyphen  \\\n",
      "5264455                 NaN  24.1  NaN      NaN   \n",
      "5264456                 NaN  24.1  NaN      NaN   \n",
      "5264457                 NaN  24.1  NaN      NaN   \n",
      "5264458                 NaN  24.1  NaN      NaN   \n",
      "5264472   p.Leu34CysfsTer30  23.3  NaN      NaN   \n",
      "...                     ...   ...  ...      ...   \n",
      "5894555         p.Trp601Ter  52.0  NaN      NaN   \n",
      "5894559         p.Gln608Ter  51.0  NaN      NaN   \n",
      "5894560         p.Gln608Ter  51.0  NaN      NaN   \n",
      "5894590  p.Gly618AlafsTer39  22.3  NaN      NaN   \n",
      "5894591  p.Gly618AlafsTer39  22.3  NaN      NaN   \n",
      "\n",
      "                                                group  ...  p_value  \\\n",
      "5264455                              Bipolar Disorder  ...      NaN   \n",
      "5264456  Bipolar Disorder (including Schizoaffective)  ...      NaN   \n",
      "5264457                            Bipolar Disorder 1  ...      NaN   \n",
      "5264458               Bipolar Disorder with Psychosis  ...      NaN   \n",
      "5264472                            Bipolar Disorder 2  ...  0.37457   \n",
      "...                                               ...  ...      ...   \n",
      "5894555  Bipolar Disorder (including Schizoaffective)  ...      NaN   \n",
      "5894559                              Bipolar Disorder  ...      NaN   \n",
      "5894560  Bipolar Disorder (including Schizoaffective)  ...      NaN   \n",
      "5894590                              Bipolar Disorder  ...      NaN   \n",
      "5894591  Bipolar Disorder (including Schizoaffective)  ...      NaN   \n",
      "\n",
      "         in_analysis  in_gnomad_non_neuro    chr        pos  gene_name  \\\n",
      "5264455         True                 True  chr11     193155    SCGB1C1   \n",
      "5264456         True                 True  chr11     193155    SCGB1C1   \n",
      "5264457         True                 True  chr11     193155    SCGB1C1   \n",
      "5264458         True                 True  chr11     193155    SCGB1C1   \n",
      "5264472        False                 True  chr11     193754    SCGB1C1   \n",
      "...              ...                  ...    ...        ...        ...   \n",
      "5894555         True                False  chr11  134374697     GLB1L2   \n",
      "5894559         True                 True  chr11  134374716     GLB1L2   \n",
      "5894560         True                 True  chr11  134374716     GLB1L2   \n",
      "5894590         True                 True  chr11  134374997     GLB1L2   \n",
      "5894591         True                 True  chr11  134374997     GLB1L2   \n",
      "\n",
      "         total_ac_case  total_an_case  total_ac_ctrl  total_an_ctrl  \n",
      "5264455              3          27802              0          28782  \n",
      "5264456              3          28356              0          28782  \n",
      "5264457              3          16448              0          28782  \n",
      "5264458              1           8748              0          28586  \n",
      "5264472              0           6892              5          28844  \n",
      "...                ...            ...            ...            ...  \n",
      "5894555             16          28420             16          28844  \n",
      "5894559             15          27866             16          28844  \n",
      "5894560             16          28420             16          28844  \n",
      "5894590             15          27866             16          28844  \n",
      "5894591             16          28420             16          28844  \n",
      "\n",
      "[40659 rows x 27 columns]\n",
      "                0         1\n",
      "5264455       inf  0.118611\n",
      "5264456       inf  0.122218\n",
      "5264457       inf  0.048085\n",
      "5264458       inf  0.234317\n",
      "5264472  0.000000  0.590703\n",
      "...           ...       ...\n",
      "5893850  1.014927  1.000000\n",
      "5893851  0.656375  0.507125\n",
      "5893852  0.613745  0.591421\n",
      "5893901  1.046312  1.000000\n",
      "5893903  1.182406  0.787822\n",
      "\n",
      "[7009 rows x 2 columns]\n",
      "                 gene_id gene_name  \\\n",
      "5344449  ENSG00000121236     TRIM6   \n",
      "5344448  ENSG00000121236     TRIM6   \n",
      "\n",
      "                                                group  total_ac_case  \\\n",
      "5344449  Bipolar Disorder (including Schizoaffective)              6   \n",
      "5344448                              Bipolar Disorder              6   \n",
      "\n",
      "         total_an_case  total_ac_ctrl  total_an_ctrl  odds_ratio   p_value  \n",
      "5344449          28420             36          28844    0.168978  0.000003  \n",
      "5344448          27866             36          28844    0.172338  0.000003  \n",
      "----------------------------------------------------------------------------------------------------\n",
      "PROCESSING CHROMOSOME 12\n",
      "                   locus            alleles          gene_id consequence  \\\n",
      "5894905      chr12:12824      [\"TGTGA\",\"T\"]  ENSG00000256263         ptv   \n",
      "5894906      chr12:12824      [\"TGTGA\",\"T\"]  ENSG00000256263         ptv   \n",
      "5894907      chr12:12824      [\"TGTGA\",\"T\"]  ENSG00000256263         ptv   \n",
      "5894908      chr12:12824      [\"TGTGA\",\"T\"]  ENSG00000256263         ptv   \n",
      "5894909      chr12:12824      [\"TGTGA\",\"T\"]  ENSG00000256263         ptv   \n",
      "...                  ...                ...              ...         ...   \n",
      "6398259  chr12:133234154  [\"ACACGGGCG\",\"A\"]  ENSG00000227059         ptv   \n",
      "6398260  chr12:133234154  [\"ACACGGGCG\",\"A\"]  ENSG00000227059         ptv   \n",
      "6398261  chr12:133234154  [\"ACACGGGCG\",\"A\"]  ENSG00000227059         ptv   \n",
      "6398262  chr12:133234154  [\"ACACGGGCG\",\"A\"]  ENSG00000227059         ptv   \n",
      "6398263  chr12:133234154  [\"ACACGGGCG\",\"A\"]  ENSG00000227059         ptv   \n",
      "\n",
      "                    hgvsc              hgvsp  cadd  mpc polyphen  \\\n",
      "5894905  n.134+4_134+7del                NaN  22.6  NaN      NaN   \n",
      "5894906  n.134+4_134+7del                NaN  22.6  NaN      NaN   \n",
      "5894907  n.134+4_134+7del                NaN  22.6  NaN      NaN   \n",
      "5894908  n.134+4_134+7del                NaN  22.6  NaN      NaN   \n",
      "5894909  n.134+4_134+7del                NaN  22.6  NaN      NaN   \n",
      "...                   ...                ...   ...  ...      ...   \n",
      "6398259      c.195_202del  p.Ala66ProfsTer81  22.5  NaN      NaN   \n",
      "6398260      c.195_202del  p.Ala66ProfsTer81  22.5  NaN      NaN   \n",
      "6398261      c.195_202del  p.Ala66ProfsTer81  22.5  NaN      NaN   \n",
      "6398262      c.195_202del  p.Ala66ProfsTer81  22.5  NaN      NaN   \n",
      "6398263      c.195_202del  p.Ala66ProfsTer81  22.5  NaN      NaN   \n",
      "\n",
      "                                                group  ... p_value  \\\n",
      "5894905                              Bipolar Disorder  ...     NaN   \n",
      "5894906  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "5894907                            Bipolar Disorder 1  ...     NaN   \n",
      "5894908                            Bipolar Disorder 2  ...     NaN   \n",
      "5894909               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "...                                               ...  ...     ...   \n",
      "6398259  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "6398260                            Bipolar Disorder 1  ...     NaN   \n",
      "6398261                            Bipolar Disorder 2  ...     NaN   \n",
      "6398262               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "6398263            Bipolar Disorder without Psychosis  ...     NaN   \n",
      "\n",
      "         in_analysis  in_gnomad_non_neuro    chr        pos  gene_name  \\\n",
      "5894905         True                False  chr12      12824        NaN   \n",
      "5894906         True                False  chr12      12824        NaN   \n",
      "5894907         True                False  chr12      12824        NaN   \n",
      "5894908         True                False  chr12      12824        NaN   \n",
      "5894909         True                False  chr12      12824        NaN   \n",
      "...              ...                  ...    ...        ...        ...   \n",
      "6398259         True                 True  chr12  133234154       ANHX   \n",
      "6398260         True                 True  chr12  133234154       ANHX   \n",
      "6398261         True                 True  chr12  133234154       ANHX   \n",
      "6398262         True                 True  chr12  133234154       ANHX   \n",
      "6398263         True                 True  chr12  133234154       ANHX   \n",
      "\n",
      "         total_ac_case  total_an_case  total_ac_ctrl  total_an_ctrl  \n",
      "5894905              0          27866              2          28844  \n",
      "5894906              0          28420              2          28844  \n",
      "5894907              0          16476              2          28844  \n",
      "5894908              0           6892              2          28844  \n",
      "5894909              0           8810              2          28844  \n",
      "...                ...            ...            ...            ...  \n",
      "6398259              1          28420              3          28844  \n",
      "6398260              1          16476              3          28844  \n",
      "6398261              0           6892              3          28844  \n",
      "6398262              0           8810              3          28844  \n",
      "6398263              0           7624              3          28844  \n",
      "\n",
      "[33158 rows x 27 columns]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                0         1\n",
      "5894905  0.000000  0.500140\n",
      "5894906  0.000000  0.500019\n",
      "5894907  0.000000  0.537228\n",
      "5894908  0.000000  1.000000\n",
      "5894909  0.000000  1.000000\n",
      "...           ...       ...\n",
      "6397978  0.338283  0.625028\n",
      "6397979  0.583531  1.000000\n",
      "6397980  0.000000  1.000000\n",
      "6397981  0.000000  1.000000\n",
      "6397982  0.000000  1.000000\n",
      "\n",
      "[5785 rows x 2 columns]\n",
      "Empty DataFrame\n",
      "Columns: [gene_id, gene_name, group, total_ac_case, total_an_case, total_ac_ctrl, total_an_ctrl, odds_ratio, p_value]\n",
      "Index: []\n",
      "----------------------------------------------------------------------------------------------------\n",
      "PROCESSING CHROMOSOME 13\n",
      "                   locus        alleles          gene_id consequence  \\\n",
      "6398413   chr13:19173992      [\"G\",\"C\"]  ENSG00000198033         ptv   \n",
      "6398414   chr13:19173992      [\"G\",\"C\"]  ENSG00000198033         ptv   \n",
      "6398415   chr13:19173992      [\"G\",\"C\"]  ENSG00000198033         ptv   \n",
      "6398416   chr13:19173992      [\"G\",\"C\"]  ENSG00000198033         ptv   \n",
      "6398417   chr13:19173992      [\"G\",\"C\"]  ENSG00000198033         ptv   \n",
      "...                  ...            ...              ...         ...   \n",
      "6562553  chr13:114304854     [\"CA\",\"C\"]  ENSG00000169062         ptv   \n",
      "6562582  chr13:114304911  [\"AGAGT\",\"A\"]  ENSG00000169062         ptv   \n",
      "6562583  chr13:114304911  [\"AGAGT\",\"A\"]  ENSG00000169062         ptv   \n",
      "6562584  chr13:114304911  [\"AGAGT\",\"A\"]  ENSG00000169062         ptv   \n",
      "6562585  chr13:114304911  [\"AGAGT\",\"A\"]  ENSG00000169062         ptv   \n",
      "\n",
      "                hgvsc               hgvsp  cadd  mpc polyphen  \\\n",
      "6398413     c.1224C>G         p.Tyr408Ter  42.0  NaN      NaN   \n",
      "6398414     c.1224C>G         p.Tyr408Ter  42.0  NaN      NaN   \n",
      "6398415     c.1224C>G         p.Tyr408Ter  42.0  NaN      NaN   \n",
      "6398416     c.1224C>G         p.Tyr408Ter  42.0  NaN      NaN   \n",
      "6398417     c.1224C>G         p.Tyr408Ter  42.0  NaN      NaN   \n",
      "...               ...                 ...   ...  ...      ...   \n",
      "6562553     c.1369del  p.Arg457GlyfsTer59   NaN  NaN      NaN   \n",
      "6562582  c.1430_*2del                 NaN  27.1  NaN      NaN   \n",
      "6562583  c.1430_*2del                 NaN  27.1  NaN      NaN   \n",
      "6562584  c.1430_*2del                 NaN  27.1  NaN      NaN   \n",
      "6562585  c.1430_*2del                 NaN  27.1  NaN      NaN   \n",
      "\n",
      "                                                group  ... p_value  \\\n",
      "6398413                              Bipolar Disorder  ...     NaN   \n",
      "6398414  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "6398415                            Bipolar Disorder 1  ...     NaN   \n",
      "6398416                            Bipolar Disorder 2  ...     NaN   \n",
      "6398417               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "...                                               ...  ...     ...   \n",
      "6562553            Bipolar Disorder without Psychosis  ...     NaN   \n",
      "6562582                              Bipolar Disorder  ...     NaN   \n",
      "6562583  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "6562584                            Bipolar Disorder 1  ...     NaN   \n",
      "6562585               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "\n",
      "         in_analysis  in_gnomad_non_neuro    chr        pos  gene_name  \\\n",
      "6398413         True                 True  chr13   19173992     TUBA3C   \n",
      "6398414         True                 True  chr13   19173992     TUBA3C   \n",
      "6398415         True                 True  chr13   19173992     TUBA3C   \n",
      "6398416         True                 True  chr13   19173992     TUBA3C   \n",
      "6398417         True                 True  chr13   19173992     TUBA3C   \n",
      "...              ...                  ...    ...        ...        ...   \n",
      "6562553         True                False  chr13  114304854      UPF3A   \n",
      "6562582         True                False  chr13  114304911      UPF3A   \n",
      "6562583         True                False  chr13  114304911      UPF3A   \n",
      "6562584         True                False  chr13  114304911      UPF3A   \n",
      "6562585         True                False  chr13  114304911      UPF3A   \n",
      "\n",
      "         total_ac_case  total_an_case  total_ac_ctrl  total_an_ctrl  \n",
      "6398413              1          27866              2          28844  \n",
      "6398414              1          28420              2          28844  \n",
      "6398415              0          16476              2          28844  \n",
      "6398416              2           6892              6          28844  \n",
      "6398417              0           8810              2          28844  \n",
      "...                ...            ...            ...            ...  \n",
      "6562553              1           7624              8          28844  \n",
      "6562582              7          27866              8          28844  \n",
      "6562583              7          28420              8          28844  \n",
      "6562584              6          16476              8          28844  \n",
      "6562585              3           8810              8          28844  \n",
      "\n",
      "[11667 rows x 27 columns]\n",
      "                0         1\n",
      "6398413  0.517531  1.000000\n",
      "6398414  0.507442  1.000000\n",
      "6398415  0.000000  0.537228\n",
      "6398416  1.395162  0.655650\n",
      "6398417  0.000000  1.000000\n",
      "...           ...       ...\n",
      "6562064  0.888027  1.000000\n",
      "6562065  1.313115  0.591403\n",
      "6562115  0.000000  0.367655\n",
      "6562116  1.227830  0.726514\n",
      "6562117  0.472845  0.695277\n",
      "\n",
      "[1707 rows x 2 columns]\n",
      "                 gene_id gene_name  \\\n",
      "6462450  ENSG00000023516    AKAP11   \n",
      "6462448  ENSG00000023516    AKAP11   \n",
      "6462449  ENSG00000023516    AKAP11   \n",
      "\n",
      "                                                group  total_ac_case  \\\n",
      "6462450                            Bipolar Disorder 1             12   \n",
      "6462448                              Bipolar Disorder             16   \n",
      "6462449  Bipolar Disorder (including Schizoaffective)             16   \n",
      "\n",
      "         total_an_case  total_ac_ctrl  total_an_ctrl  odds_ratio   p_value  \n",
      "6462450          16476              0          28844         inf  0.000005  \n",
      "6462448          27866              0          28844         inf  0.000012  \n",
      "6462449          28420              0          28844         inf  0.000014  \n",
      "----------------------------------------------------------------------------------------------------\n",
      "PROCESSING CHROMOSOME 14\n",
      "                   locus     alleles          gene_id consequence     hgvsc  \\\n",
      "6563321   chr14:18601321   [\"C\",\"T\"]  ENSG00000257115         ptv  c.205C>T   \n",
      "6563322   chr14:18601321   [\"C\",\"T\"]  ENSG00000257115         ptv  c.205C>T   \n",
      "6563323   chr14:18601321   [\"C\",\"T\"]  ENSG00000257115         ptv  c.205C>T   \n",
      "6563324   chr14:18601321   [\"C\",\"T\"]  ENSG00000257115         ptv  c.205C>T   \n",
      "6563376   chr14:18601377   [\"G\",\"A\"]  ENSG00000257115         ptv  c.261G>A   \n",
      "...                  ...         ...              ...         ...       ...   \n",
      "6901531  chr14:106810645  [\"TA\",\"T\"]  ENSG00000224650         ptv  c.147del   \n",
      "6901532  chr14:106810645  [\"TA\",\"T\"]  ENSG00000224650         ptv  c.147del   \n",
      "6901533  chr14:106810645  [\"TA\",\"T\"]  ENSG00000224650         ptv  c.147del   \n",
      "6901534  chr14:106810645  [\"TA\",\"T\"]  ENSG00000224650         ptv  c.147del   \n",
      "6901535  chr14:106810645  [\"TA\",\"T\"]  ENSG00000224650         ptv  c.147del   \n",
      "\n",
      "                     hgvsp   cadd  mpc polyphen  \\\n",
      "6563321         p.Arg69Ter  19.88  NaN      NaN   \n",
      "6563322         p.Arg69Ter  19.88  NaN      NaN   \n",
      "6563323         p.Arg69Ter  19.88  NaN      NaN   \n",
      "6563324         p.Arg69Ter  19.88  NaN      NaN   \n",
      "6563376         p.Trp87Ter  26.50  NaN      NaN   \n",
      "...                    ...    ...  ...      ...   \n",
      "6901531  p.Ser49ArgfsTer35    NaN  NaN      NaN   \n",
      "6901532  p.Ser49ArgfsTer35    NaN  NaN      NaN   \n",
      "6901533  p.Ser49ArgfsTer35    NaN  NaN      NaN   \n",
      "6901534  p.Ser49ArgfsTer35    NaN  NaN      NaN   \n",
      "6901535  p.Ser49ArgfsTer35    NaN  NaN      NaN   \n",
      "\n",
      "                                                group  ... p_value  \\\n",
      "6563321                              Bipolar Disorder  ...     NaN   \n",
      "6563322  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "6563323                            Bipolar Disorder 1  ...     NaN   \n",
      "6563324               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "6563376                              Bipolar Disorder  ...     NaN   \n",
      "...                                               ...  ...     ...   \n",
      "6901531  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "6901532                            Bipolar Disorder 1  ...     NaN   \n",
      "6901533                            Bipolar Disorder 2  ...     NaN   \n",
      "6901534               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "6901535            Bipolar Disorder without Psychosis  ...     NaN   \n",
      "\n",
      "         in_analysis  in_gnomad_non_neuro    chr        pos  gene_name  \\\n",
      "6563321         True                 True  chr14   18601321    OR11H12   \n",
      "6563322         True                 True  chr14   18601321    OR11H12   \n",
      "6563323         True                 True  chr14   18601321    OR11H12   \n",
      "6563324         True                 True  chr14   18601321    OR11H12   \n",
      "6563376         True                False  chr14   18601377    OR11H12   \n",
      "...              ...                  ...    ...        ...        ...   \n",
      "6901531         True                False  chr14  106810645        NaN   \n",
      "6901532         True                False  chr14  106810645        NaN   \n",
      "6901533         True                False  chr14  106810645        NaN   \n",
      "6901534         True                False  chr14  106810645        NaN   \n",
      "6901535         True                False  chr14  106810645        NaN   \n",
      "\n",
      "         total_ac_case  total_an_case  total_ac_ctrl  total_an_ctrl  \n",
      "6563321              7          27866              4          28844  \n",
      "6563322              7          28420              4          28844  \n",
      "6563323              5          16476              4          28844  \n",
      "6563324              4           8808              7          28844  \n",
      "6563376              7          27866              4          28844  \n",
      "...                ...            ...            ...            ...  \n",
      "6901531              0          28420              5          28844  \n",
      "6901532              0          16476              5          28844  \n",
      "6901533              0           6892              5          28844  \n",
      "6901534              0           8810              5          28844  \n",
      "6901535              0           7624              5          28844  \n",
      "\n",
      "[21085 rows x 27 columns]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                0         1\n",
      "6563321  1.811623  0.380042\n",
      "6563322  1.776300  0.384206\n",
      "6563323  2.188695  0.300570\n",
      "6563324  1.871682  0.296786\n",
      "6563379  1.621666  0.445473\n",
      "...           ...       ...\n",
      "6901410  0.000000  0.062523\n",
      "6901411  0.000000  0.166346\n",
      "6901412  0.000000  0.590703\n",
      "6901413  0.000000  0.597156\n",
      "6901414  0.000000  0.590886\n",
      "\n",
      "[3976 rows x 2 columns]\n",
      "Empty DataFrame\n",
      "Columns: [gene_id, gene_name, group, total_ac_case, total_an_case, total_ac_ctrl, total_an_ctrl, odds_ratio, p_value]\n",
      "Index: []\n",
      "----------------------------------------------------------------------------------------------------\n",
      "PROCESSING CHROMOSOME 15\n",
      "                   locus    alleles          gene_id consequence       hgvsc  \\\n",
      "6901777   chr15:19964670  [\"T\",\"A\"]  ENSG00000188403         ptv    c.349A>T   \n",
      "6901778   chr15:19964670  [\"T\",\"A\"]  ENSG00000188403         ptv    c.349A>T   \n",
      "6901779   chr15:19964670  [\"T\",\"A\"]  ENSG00000188403         ptv    c.349A>T   \n",
      "6901780   chr15:19964670  [\"T\",\"A\"]  ENSG00000188403         ptv    c.349A>T   \n",
      "6901781   chr15:19964670  [\"T\",\"A\"]  ENSG00000188403         ptv    c.349A>T   \n",
      "...                  ...        ...              ...         ...         ...   \n",
      "7255710  chr15:101972695  [\"G\",\"A\"]  ENSG00000185596         ptv  n.284+1G>A   \n",
      "7255711  chr15:101972695  [\"G\",\"T\"]  ENSG00000185596         ptv  n.284+1G>T   \n",
      "7255712  chr15:101972695  [\"G\",\"T\"]  ENSG00000185596         ptv  n.284+1G>T   \n",
      "7255713  chr15:101972695  [\"G\",\"T\"]  ENSG00000185596         ptv  n.284+1G>T   \n",
      "7255714  chr15:101972695  [\"G\",\"T\"]  ENSG00000185596         ptv  n.284+1G>T   \n",
      "\n",
      "               hgvsp  cadd  mpc polyphen  \\\n",
      "6901777  p.Arg117Ter  27.2  NaN      NaN   \n",
      "6901778  p.Arg117Ter  27.2  NaN      NaN   \n",
      "6901779  p.Arg117Ter  27.2  NaN      NaN   \n",
      "6901780  p.Arg117Ter  27.2  NaN      NaN   \n",
      "6901781  p.Arg117Ter  27.2  NaN      NaN   \n",
      "...              ...   ...  ...      ...   \n",
      "7255710          NaN  24.3  NaN      NaN   \n",
      "7255711          NaN  24.2  NaN      NaN   \n",
      "7255712          NaN  24.2  NaN      NaN   \n",
      "7255713          NaN  24.2  NaN      NaN   \n",
      "7255714          NaN  24.2  NaN      NaN   \n",
      "\n",
      "                                                group  ... p_value  \\\n",
      "6901777                              Bipolar Disorder  ...     NaN   \n",
      "6901778  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "6901779                            Bipolar Disorder 1  ...     NaN   \n",
      "6901780                            Bipolar Disorder 2  ...     NaN   \n",
      "6901781               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "...                                               ...  ...     ...   \n",
      "7255710               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "7255711                              Bipolar Disorder  ...     NaN   \n",
      "7255712  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "7255713                            Bipolar Disorder 1  ...     NaN   \n",
      "7255714               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "\n",
      "         in_analysis  in_gnomad_non_neuro    chr        pos  gene_name  \\\n",
      "6901777         True                False  chr15   19964670        NaN   \n",
      "6901778         True                False  chr15   19964670        NaN   \n",
      "6901779         True                False  chr15   19964670        NaN   \n",
      "6901780         True                False  chr15   19964670        NaN   \n",
      "6901781         True                False  chr15   19964670        NaN   \n",
      "...              ...                  ...    ...        ...        ...   \n",
      "7255710         True                 True  chr15  101972695        NaN   \n",
      "7255711         True                False  chr15  101972695        NaN   \n",
      "7255712         True                False  chr15  101972695        NaN   \n",
      "7255713         True                False  chr15  101972695        NaN   \n",
      "7255714         True                False  chr15  101972695        NaN   \n",
      "\n",
      "         total_ac_case  total_an_case  total_ac_ctrl  total_an_ctrl  \n",
      "6901777              4          27866              2          28844  \n",
      "6901778              4          28420              2          28844  \n",
      "6901779              2          16476              2          28844  \n",
      "6901780              2           6892              2          28844  \n",
      "6901781              0           8810              2          28844  \n",
      "...                ...            ...            ...            ...  \n",
      "7255710              2           8656              0          28482  \n",
      "7255711              2          27448              0          28482  \n",
      "7255712              2          28000              0          28482  \n",
      "7255713              2          16238              0          28482  \n",
      "7255714              2           8656              0          28482  \n",
      "\n",
      "[23639 rows x 27 columns]\n",
      "                0         1\n",
      "6901777  2.070347  0.445373\n",
      "6901778  2.029983  0.449698\n",
      "6901779  1.750759  0.625087\n",
      "6901780  4.186067  0.169922\n",
      "6901781  0.000000  1.000000\n",
      "...           ...       ...\n",
      "7255496  2.183284  0.333350\n",
      "7255707       inf  0.240837\n",
      "7255708       inf  0.245747\n",
      "7255709       inf  0.131839\n",
      "7255710       inf  0.054320\n",
      "\n",
      "[3391 rows x 2 columns]\n",
      "Empty DataFrame\n",
      "Columns: [gene_id, gene_name, group, total_ac_case, total_an_case, total_ac_ctrl, total_an_ctrl, odds_ratio, p_value]\n",
      "Index: []\n",
      "----------------------------------------------------------------------------------------------------\n",
      "PROCESSING CHROMOSOME 16\n",
      "                  locus      alleles          gene_id consequence       hgvsc  \\\n",
      "7255868     chr16:53512    [\"G\",\"T\"]  ENSG00000161980         ptv     c.75C>A   \n",
      "7255869     chr16:53512    [\"G\",\"T\"]  ENSG00000161980         ptv     c.75C>A   \n",
      "7255870     chr16:53512    [\"G\",\"T\"]  ENSG00000161980         ptv     c.75C>A   \n",
      "7255962     chr16:55519    [\"C\",\"T\"]  ENSG00000161981         ptv    c.130C>T   \n",
      "7255963     chr16:55519    [\"C\",\"T\"]  ENSG00000161981         ptv    c.130C>T   \n",
      "...                 ...          ...              ...         ...         ...   \n",
      "7760863  chr16:90075865  [\"CTG\",\"C\"]  ENSG00000126856         ptv  c.44_45del   \n",
      "7760900  chr16:90094341    [\"G\",\"A\"]  ENSG00000261812         ptv  n.701-1G>A   \n",
      "7760901  chr16:90094341    [\"G\",\"A\"]  ENSG00000261812         ptv  n.701-1G>A   \n",
      "7760902  chr16:90094341    [\"G\",\"A\"]  ENSG00000261812         ptv  n.701-1G>A   \n",
      "7760903  chr16:90094341    [\"G\",\"A\"]  ENSG00000261812         ptv  n.701-1G>A   \n",
      "\n",
      "                     hgvsp   cadd  mpc polyphen  \\\n",
      "7255868         p.Cys25Ter  39.00  NaN      NaN   \n",
      "7255869         p.Cys25Ter  39.00  NaN      NaN   \n",
      "7255870         p.Cys25Ter  39.00  NaN      NaN   \n",
      "7255962         p.Arg44Ter  38.00  NaN      NaN   \n",
      "7255963         p.Arg44Ter  38.00  NaN      NaN   \n",
      "...                    ...    ...  ...      ...   \n",
      "7760863  p.Thr15ArgfsTer37  15.90  NaN      NaN   \n",
      "7760900                NaN  19.28  NaN      NaN   \n",
      "7760901                NaN  19.28  NaN      NaN   \n",
      "7760902                NaN  19.28  NaN      NaN   \n",
      "7760903                NaN  19.28  NaN      NaN   \n",
      "\n",
      "                                                group  ... p_value  \\\n",
      "7255868                              Bipolar Disorder  ...     NaN   \n",
      "7255869  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "7255870               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "7255962                              Bipolar Disorder  ...     NaN   \n",
      "7255963  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "...                                               ...  ...     ...   \n",
      "7760863            Bipolar Disorder without Psychosis  ...     NaN   \n",
      "7760900                              Bipolar Disorder  ...     NaN   \n",
      "7760901  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "7760902                            Bipolar Disorder 1  ...     NaN   \n",
      "7760903               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "\n",
      "         in_analysis  in_gnomad_non_neuro    chr       pos  gene_name  \\\n",
      "7255868         True                False  chr16     53512     POLR3K   \n",
      "7255869         True                False  chr16     53512     POLR3K   \n",
      "7255870         True                False  chr16     53512     POLR3K   \n",
      "7255962         True                 True  chr16     55519    SNRNP25   \n",
      "7255963         True                 True  chr16     55519    SNRNP25   \n",
      "...              ...                  ...    ...       ...        ...   \n",
      "7760863         True                 True  chr16  90075865      PRDM7   \n",
      "7760900         True                False  chr16  90094341        NaN   \n",
      "7760901         True                False  chr16  90094341        NaN   \n",
      "7760902         True                False  chr16  90094341        NaN   \n",
      "7760903         True                False  chr16  90094341        NaN   \n",
      "\n",
      "         total_ac_case  total_an_case  total_ac_ctrl  total_an_ctrl  \n",
      "7255868              1          27862              0          28842  \n",
      "7255869              1          28416              0          28842  \n",
      "7255870              1           8808              0          28842  \n",
      "7255962              3          27866              4          28844  \n",
      "7255963              3          28420              4          28844  \n",
      "...                ...            ...            ...            ...  \n",
      "7760863              2           7624              7          28844  \n",
      "7760900              1          27866              0          28844  \n",
      "7760901              1          28420              0          28844  \n",
      "7760902              1          16476              0          28844  \n",
      "7760903              1           8810              0          28844  \n",
      "\n",
      "[29979 rows x 27 columns]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                0         1\n",
      "7255868       inf  0.491359\n",
      "7255869       inf  0.496280\n",
      "7255870       inf  0.233944\n",
      "7255962  0.776298  1.000000\n",
      "7255963  0.761164  1.000000\n",
      "...           ...       ...\n",
      "7760404  1.080969  1.000000\n",
      "7760900       inf  0.491377\n",
      "7760901       inf  0.496298\n",
      "7760902       inf  0.363548\n",
      "7760903       inf  0.233972\n",
      "\n",
      "[4768 rows x 2 columns]\n",
      "Empty DataFrame\n",
      "Columns: [gene_id, gene_name, group, total_ac_case, total_an_case, total_ac_ctrl, total_an_ctrl, odds_ratio, p_value]\n",
      "Index: []\n",
      "----------------------------------------------------------------------------------------------------\n",
      "PROCESSING CHROMOSOME 17\n",
      "                  locus        alleles          gene_id consequence  \\\n",
      "7761050    chr17:162120  [\"TAGTA\",\"T\"]  ENSG00000272636         ptv   \n",
      "7761051    chr17:162120  [\"TAGTA\",\"T\"]  ENSG00000272636         ptv   \n",
      "7761052    chr17:162120  [\"TAGTA\",\"T\"]  ENSG00000272636         ptv   \n",
      "7761053    chr17:162120  [\"TAGTA\",\"T\"]  ENSG00000272636         ptv   \n",
      "7761111    chr17:213889  [\"T\",\"TCAGC\"]  ENSG00000181031         ptv   \n",
      "...                 ...            ...              ...         ...   \n",
      "8361803  chr17:83094261      [\"C\",\"T\"]  ENSG00000176845         ptv   \n",
      "8361804  chr17:83094261      [\"C\",\"T\"]  ENSG00000176845         ptv   \n",
      "8361805  chr17:83094261      [\"C\",\"T\"]  ENSG00000176845         ptv   \n",
      "8361806  chr17:83094261      [\"C\",\"T\"]  ENSG00000176845         ptv   \n",
      "8361807  chr17:83094261      [\"C\",\"T\"]  ENSG00000176845         ptv   \n",
      "\n",
      "                hgvsc              hgvsp  cadd  mpc polyphen  \\\n",
      "7761050  c.595_598del  p.Tyr199ThrfsTer8   NaN  NaN      NaN   \n",
      "7761051  c.595_598del  p.Tyr199ThrfsTer8   NaN  NaN      NaN   \n",
      "7761052  c.595_598del  p.Tyr199ThrfsTer8   NaN  NaN      NaN   \n",
      "7761053  c.595_598del  p.Tyr199ThrfsTer8   NaN  NaN      NaN   \n",
      "7761111  c.907_910dup  p.Asp304GlyfsTer2   NaN  NaN      NaN   \n",
      "...               ...                ...   ...  ...      ...   \n",
      "8361803      c.622C>T        p.Arg208Ter  41.0  NaN      NaN   \n",
      "8361804      c.622C>T        p.Arg208Ter  41.0  NaN      NaN   \n",
      "8361805      c.622C>T        p.Arg208Ter  41.0  NaN      NaN   \n",
      "8361806      c.622C>T        p.Arg208Ter  41.0  NaN      NaN   \n",
      "8361807      c.622C>T        p.Arg208Ter  41.0  NaN      NaN   \n",
      "\n",
      "                                                group  ... p_value  \\\n",
      "7761050                              Bipolar Disorder  ...     NaN   \n",
      "7761051  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "7761052                            Bipolar Disorder 1  ...     NaN   \n",
      "7761053               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "7761111                              Bipolar Disorder  ...     NaN   \n",
      "...                                               ...  ...     ...   \n",
      "8361803  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "8361804                            Bipolar Disorder 1  ...     NaN   \n",
      "8361805                            Bipolar Disorder 2  ...     NaN   \n",
      "8361806               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "8361807            Bipolar Disorder without Psychosis  ...     NaN   \n",
      "\n",
      "         in_analysis  in_gnomad_non_neuro    chr       pos  gene_name  \\\n",
      "7761050         True                 True  chr17    162120      DOC2B   \n",
      "7761051         True                 True  chr17    162120      DOC2B   \n",
      "7761052         True                 True  chr17    162120      DOC2B   \n",
      "7761053         True                 True  chr17    162120      DOC2B   \n",
      "7761111         True                False  chr17    213889     RPH3AL   \n",
      "...              ...                  ...    ...       ...        ...   \n",
      "8361803         True                False  chr17  83094261     METRNL   \n",
      "8361804         True                False  chr17  83094261     METRNL   \n",
      "8361805         True                False  chr17  83094261     METRNL   \n",
      "8361806         True                False  chr17  83094261     METRNL   \n",
      "8361807         True                False  chr17  83094261     METRNL   \n",
      "\n",
      "         total_ac_case  total_an_case  total_ac_ctrl  total_an_ctrl  \n",
      "7761050              1          27862              0          28844  \n",
      "7761051              1          28416              0          28844  \n",
      "7761052              1          16476              0          28844  \n",
      "7761053              1           8810              0          28844  \n",
      "7761111             11          27866              6          28844  \n",
      "...                ...            ...            ...            ...  \n",
      "8361803              5          28404              1          28844  \n",
      "8361804              1          16028              1          28268  \n",
      "8361805              2           6890              1          28844  \n",
      "8361806              1           8804              1          28844  \n",
      "8361807              2           7618              1          28844  \n",
      "\n",
      "[38819 rows x 27 columns]\n",
      "                0         1\n",
      "7761050       inf  0.491341\n",
      "7761051       inf  0.496263\n",
      "7761052       inf  0.363548\n",
      "7761053       inf  0.233972\n",
      "7761111  1.898031  0.230807\n",
      "...           ...       ...\n",
      "8361664  5.078172  0.122157\n",
      "8361706  8.374855  0.097187\n",
      "8361707  3.276497  0.413020\n",
      "8361708  7.574317  0.112707\n",
      "8361804  1.763711  1.000000\n",
      "\n",
      "[6272 rows x 2 columns]\n",
      "Empty DataFrame\n",
      "Columns: [gene_id, gene_name, group, total_ac_case, total_an_case, total_ac_ctrl, total_an_ctrl, odds_ratio, p_value]\n",
      "Index: []\n",
      "----------------------------------------------------------------------------------------------------\n",
      "PROCESSING CHROMOSOME 18\n",
      "                  locus     alleles          gene_id consequence      hgvsc  \\\n",
      "8361982     chr18:47425   [\"C\",\"A\"]  ENSG00000173213         ptv  c.1300G>T   \n",
      "8361983     chr18:47425   [\"C\",\"A\"]  ENSG00000173213         ptv  c.1300G>T   \n",
      "8361984     chr18:47425   [\"C\",\"A\"]  ENSG00000173213         ptv  c.1300G>T   \n",
      "8361985     chr18:47425   [\"C\",\"A\"]  ENSG00000173213         ptv  c.1300G>T   \n",
      "8361986     chr18:47425   [\"C\",\"A\"]  ENSG00000173213         ptv  c.1300G>T   \n",
      "...                 ...         ...              ...         ...        ...   \n",
      "8509296  chr18:80159961  [\"TG\",\"T\"]  ENSG00000178184         ptv   c.940del   \n",
      "8509482  chr18:80160430   [\"G\",\"A\"]  ENSG00000178184         ptv   c.472C>T   \n",
      "8509483  chr18:80160430   [\"G\",\"A\"]  ENSG00000178184         ptv   c.472C>T   \n",
      "8509484  chr18:80160430   [\"G\",\"A\"]  ENSG00000178184         ptv   c.472C>T   \n",
      "8509485  chr18:80160430   [\"G\",\"A\"]  ENSG00000178184         ptv   c.472C>T   \n",
      "\n",
      "                      hgvsp  cadd  mpc polyphen  \\\n",
      "8361982         p.Glu434Ter  35.0  NaN      NaN   \n",
      "8361983         p.Glu434Ter  35.0  NaN      NaN   \n",
      "8361984         p.Glu434Ter  35.0  NaN      NaN   \n",
      "8361985         p.Glu434Ter  35.0  NaN      NaN   \n",
      "8361986         p.Glu434Ter  35.0  NaN      NaN   \n",
      "...                     ...   ...  ...      ...   \n",
      "8509296  p.Gln314ArgfsTer78  22.2  NaN      NaN   \n",
      "8509482         p.Arg158Ter  40.0  NaN      NaN   \n",
      "8509483         p.Arg158Ter  40.0  NaN      NaN   \n",
      "8509484         p.Arg158Ter  40.0  NaN      NaN   \n",
      "8509485         p.Arg158Ter  40.0  NaN      NaN   \n",
      "\n",
      "                                                group  ... p_value  \\\n",
      "8361982                              Bipolar Disorder  ...     NaN   \n",
      "8361983  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "8361984                            Bipolar Disorder 1  ...     NaN   \n",
      "8361985                            Bipolar Disorder 2  ...     NaN   \n",
      "8361986               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "...                                               ...  ...     ...   \n",
      "8509296            Bipolar Disorder without Psychosis  ...     NaN   \n",
      "8509482                              Bipolar Disorder  ...     NaN   \n",
      "8509483  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "8509484                            Bipolar Disorder 2  ...     NaN   \n",
      "8509485               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "\n",
      "         in_analysis  in_gnomad_non_neuro    chr       pos  gene_name  \\\n",
      "8361982         True                False  chr18     47425     TUBB8B   \n",
      "8361983         True                False  chr18     47425     TUBB8B   \n",
      "8361984         True                False  chr18     47425     TUBB8B   \n",
      "8361985         True                False  chr18     47425     TUBB8B   \n",
      "8361986         True                False  chr18     47425     TUBB8B   \n",
      "...              ...                  ...    ...       ...        ...   \n",
      "8509296         True                 True  chr18  80159961     PARD6G   \n",
      "8509482         True                 True  chr18  80160430     PARD6G   \n",
      "8509483         True                 True  chr18  80160430     PARD6G   \n",
      "8509484         True                 True  chr18  80160430     PARD6G   \n",
      "8509485         True                 True  chr18  80160430     PARD6G   \n",
      "\n",
      "         total_ac_case  total_an_case  total_ac_ctrl  total_an_ctrl  \n",
      "8361982             14          27866             21          28844  \n",
      "8361983             14          28420             21          28844  \n",
      "8361984              8          16476             21          28844  \n",
      "8361985              3           6892             33          28844  \n",
      "8361986              7           8810             28          28844  \n",
      "...                ...            ...            ...            ...  \n",
      "8509296              0           7534              3          28526  \n",
      "8509482              5          27866              3          28844  \n",
      "8509483              5          28420              3          28844  \n",
      "8509484              1           6892              3          28844  \n",
      "8509485              1           8810              3          28844  \n",
      "\n",
      "[9081 rows x 27 columns]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                0         1\n",
      "8361982  0.689909  0.312882\n",
      "8361983  0.676453  0.310823\n",
      "8361984  0.666759  0.440214\n",
      "8361985  0.380198  0.135562\n",
      "8361986  0.818357  0.841568\n",
      "...           ...       ...\n",
      "8509260  1.691653  0.504630\n",
      "8509261  1.751303  0.674668\n",
      "8509288  1.395105  0.575595\n",
      "8509289  1.091346  1.000000\n",
      "8509290  0.000000  1.000000\n",
      "\n",
      "[1532 rows x 2 columns]\n",
      "Empty DataFrame\n",
      "Columns: [gene_id, gene_name, group, total_ac_case, total_an_case, total_ac_ctrl, total_an_ctrl, odds_ratio, p_value]\n",
      "Index: []\n",
      "----------------------------------------------------------------------------------------------------\n",
      "PROCESSING CHROMOSOME 19\n",
      "                  locus     alleles          gene_id consequence       hgvsc  \\\n",
      "8509591    chr19:107557   [\"T\",\"C\"]  ENSG00000176695         ptv  c.-55+2T>C   \n",
      "8509592    chr19:107557   [\"T\",\"C\"]  ENSG00000176695         ptv  c.-55+2T>C   \n",
      "8509593    chr19:107557   [\"T\",\"C\"]  ENSG00000176695         ptv  c.-55+2T>C   \n",
      "8509594    chr19:107557   [\"T\",\"C\"]  ENSG00000176695         ptv  c.-55+2T>C   \n",
      "8509595    chr19:107557   [\"T\",\"C\"]  ENSG00000176695         ptv  c.-55+2T>C   \n",
      "...                 ...         ...              ...         ...         ...   \n",
      "9267806  chr19:58571430  [\"CT\",\"C\"]  ENSG00000099326         ptv  c.-40-2del   \n",
      "9267807  chr19:58571430  [\"CT\",\"C\"]  ENSG00000099326         ptv  c.-40-2del   \n",
      "9267808  chr19:58571430  [\"CT\",\"C\"]  ENSG00000099326         ptv  c.-40-2del   \n",
      "9267809  chr19:58571430  [\"CT\",\"C\"]  ENSG00000099326         ptv  c.-40-2del   \n",
      "9267810  chr19:58571430  [\"CT\",\"C\"]  ENSG00000099326         ptv  c.-40-2del   \n",
      "\n",
      "        hgvsp   cadd  mpc polyphen  \\\n",
      "8509591   NaN  14.09  NaN      NaN   \n",
      "8509592   NaN  14.09  NaN      NaN   \n",
      "8509593   NaN  14.09  NaN      NaN   \n",
      "8509594   NaN  14.09  NaN      NaN   \n",
      "8509595   NaN  14.09  NaN      NaN   \n",
      "...       ...    ...  ...      ...   \n",
      "9267806   NaN  19.40  NaN      NaN   \n",
      "9267807   NaN  19.40  NaN      NaN   \n",
      "9267808   NaN  19.40  NaN      NaN   \n",
      "9267809   NaN  19.40  NaN      NaN   \n",
      "9267810   NaN  19.40  NaN      NaN   \n",
      "\n",
      "                                                group  ... p_value  \\\n",
      "8509591                              Bipolar Disorder  ...     NaN   \n",
      "8509592  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "8509593                            Bipolar Disorder 1  ...     NaN   \n",
      "8509594                            Bipolar Disorder 2  ...     NaN   \n",
      "8509595            Bipolar Disorder without Psychosis  ...     NaN   \n",
      "...                                               ...  ...     ...   \n",
      "9267806  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "9267807                            Bipolar Disorder 1  ...     NaN   \n",
      "9267808                            Bipolar Disorder 2  ...     NaN   \n",
      "9267809               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "9267810            Bipolar Disorder without Psychosis  ...     NaN   \n",
      "\n",
      "         in_analysis  in_gnomad_non_neuro    chr       pos  gene_name  \\\n",
      "8509591         True                 True  chr19    107557     OR4F17   \n",
      "8509592         True                 True  chr19    107557     OR4F17   \n",
      "8509593         True                 True  chr19    107557     OR4F17   \n",
      "8509594         True                 True  chr19    107557     OR4F17   \n",
      "8509595         True                 True  chr19    107557     OR4F17   \n",
      "...              ...                  ...    ...       ...        ...   \n",
      "9267806         True                 True  chr19  58571430       MZF1   \n",
      "9267807         True                 True  chr19  58571430       MZF1   \n",
      "9267808         True                 True  chr19  58571430       MZF1   \n",
      "9267809         True                 True  chr19  58571430       MZF1   \n",
      "9267810         True                 True  chr19  58571430       MZF1   \n",
      "\n",
      "         total_ac_case  total_an_case  total_ac_ctrl  total_an_ctrl  \n",
      "8509591              6          27866              7          28844  \n",
      "8509592              6          28420              7          28844  \n",
      "8509593              2          16476              7          28844  \n",
      "8509594              1           6892              7          28844  \n",
      "8509595              3           7624              7          28844  \n",
      "...                ...            ...            ...            ...  \n",
      "9267806             14          28420              7          28844  \n",
      "9267807              9          16476              7          28844  \n",
      "9267808              4           6892              7          28844  \n",
      "9267809              2           8810              7          28844  \n",
      "9267810              4           7624              7          28844  \n",
      "\n",
      "[51848 rows x 27 columns]\n",
      "                0         1\n",
      "8509591  0.887201  1.000000\n",
      "8509592  0.869903  1.000000\n",
      "8509593  0.500130  0.501844\n",
      "8509594  0.597819  1.000000\n",
      "8509595  1.621666  0.445473\n",
      "...           ...       ...\n",
      "9266981  2.030346  0.131072\n",
      "9266982  2.251542  0.119414\n",
      "9266983  2.392318  0.240338\n",
      "9266984  0.935416  1.000000\n",
      "9266985  2.162505  0.257483\n",
      "\n",
      "[7973 rows x 2 columns]\n",
      "                 gene_id gene_name               group  total_ac_case  \\\n",
      "9129186  ENSG00000161681    SHANK1  Bipolar Disorder 1             12   \n",
      "\n",
      "         total_an_case  total_ac_ctrl  total_an_ctrl  odds_ratio   p_value  \n",
      "9129186          16476              1          28844   21.022595  0.000046  \n",
      "----------------------------------------------------------------------------------------------------\n",
      "PROCESSING CHROMOSOME 20\n",
      "                  locus     alleles          gene_id consequence      hgvsc  \\\n",
      "9267829     chr20:96004   [\"G\",\"A\"]  ENSG00000178591         ptv  c.59-1G>A   \n",
      "9267830     chr20:96004   [\"G\",\"A\"]  ENSG00000178591         ptv  c.59-1G>A   \n",
      "9267831     chr20:96004   [\"G\",\"A\"]  ENSG00000178591         ptv  c.59-1G>A   \n",
      "9267832     chr20:96013   [\"G\",\"T\"]  ENSG00000178591         ptv    c.67G>T   \n",
      "9267833     chr20:96013   [\"G\",\"T\"]  ENSG00000178591         ptv    c.67G>T   \n",
      "...                 ...         ...              ...         ...        ...   \n",
      "9519391  chr20:64273586  [\"T\",\"TA\"]  ENSG00000203880         ptv  c.1073dup   \n",
      "9519392  chr20:64273586  [\"T\",\"TA\"]  ENSG00000203880         ptv  c.1073dup   \n",
      "9519393  chr20:64273586  [\"T\",\"TA\"]  ENSG00000203880         ptv  c.1073dup   \n",
      "9519394  chr20:64273586  [\"T\",\"TA\"]  ENSG00000203880         ptv  c.1073dup   \n",
      "9519395  chr20:64273586  [\"T\",\"TA\"]  ENSG00000203880         ptv  c.1073dup   \n",
      "\n",
      "               hgvsp  cadd  mpc polyphen  \\\n",
      "9267829          NaN  20.1  NaN      NaN   \n",
      "9267830          NaN  20.1  NaN      NaN   \n",
      "9267831          NaN  20.1  NaN      NaN   \n",
      "9267832   p.Glu23Ter  27.3  NaN      NaN   \n",
      "9267833   p.Glu23Ter  27.3  NaN      NaN   \n",
      "...              ...   ...  ...      ...   \n",
      "9519391  p.Tyr358Ter  35.0  NaN      NaN   \n",
      "9519392  p.Tyr358Ter  35.0  NaN      NaN   \n",
      "9519393  p.Tyr358Ter  35.0  NaN      NaN   \n",
      "9519394  p.Tyr358Ter  35.0  NaN      NaN   \n",
      "9519395  p.Tyr358Ter  35.0  NaN      NaN   \n",
      "\n",
      "                                                group  ... p_value  \\\n",
      "9267829                              Bipolar Disorder  ...     NaN   \n",
      "9267830  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "9267831                            Bipolar Disorder 1  ...     NaN   \n",
      "9267832                              Bipolar Disorder  ...     NaN   \n",
      "9267833  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "...                                               ...  ...     ...   \n",
      "9519391  Bipolar Disorder (including Schizoaffective)  ...     NaN   \n",
      "9519392                            Bipolar Disorder 1  ...     NaN   \n",
      "9519393                            Bipolar Disorder 2  ...     NaN   \n",
      "9519394               Bipolar Disorder with Psychosis  ...     NaN   \n",
      "9519395            Bipolar Disorder without Psychosis  ...     NaN   \n",
      "\n",
      "         in_analysis  in_gnomad_non_neuro    chr       pos  gene_name  \\\n",
      "9267829         True                False  chr20     96004    DEFB125   \n",
      "9267830         True                False  chr20     96004    DEFB125   \n",
      "9267831         True                False  chr20     96004    DEFB125   \n",
      "9267832         True                 True  chr20     96013    DEFB125   \n",
      "9267833         True                 True  chr20     96013    DEFB125   \n",
      "...              ...                  ...    ...       ...        ...   \n",
      "9519391         True                False  chr20  64273586     PCMTD2   \n",
      "9519392         True                False  chr20  64273586     PCMTD2   \n",
      "9519393         True                False  chr20  64273586     PCMTD2   \n",
      "9519394         True                False  chr20  64273586     PCMTD2   \n",
      "9519395         True                False  chr20  64273586     PCMTD2   \n",
      "\n",
      "         total_ac_case  total_an_case  total_ac_ctrl  total_an_ctrl  \n",
      "9267829              5          27866              5          28844  \n",
      "9267830              5          28420              5          28844  \n",
      "9267831              3          16476              5          28844  \n",
      "9267832              5          27866              5          28844  \n",
      "9267833              5          28420              5          28844  \n",
      "...                ...            ...            ...            ...  \n",
      "9519391              1          28420              5          28844  \n",
      "9519392              1          16476              5          28844  \n",
      "9519393              0           6892              5          28844  \n",
      "9519394              1           8810              5          28844  \n",
      "9519395              0           7624              5          28844  \n",
      "\n",
      "[15178 rows x 27 columns]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                0         1\n",
      "9267829  1.035103  1.000000\n",
      "9267830  1.014922  1.000000\n",
      "9267831  1.050410  1.000000\n",
      "9267835  0.837005  1.000000\n",
      "9267836  0.654762  1.000000\n",
      "...           ...       ...\n",
      "9519128  0.202956  0.218803\n",
      "9519129  0.350094  0.426924\n",
      "9519130  0.000000  0.590703\n",
      "9519131  0.654762  1.000000\n",
      "9519132  0.000000  0.590886\n",
      "\n",
      "[2826 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "# Parameters for analysis\n",
    "prob_thr = 0.6\n",
    "cnt_thr = 5\n",
    "significance_thr = 1\n",
    "variant_results_df = bipex_variant_results_df\n",
    "\n",
    "# Initialize empty dataframes\n",
    "significant_genes_df = pd.DataFrame(columns=['gene_id', 'gene_name', 'group', 'total_ac_case', 'total_an_case', 'total_ac_ctrl', 'total_an_ctrl', 'odds_ratio', 'p_value'])\n",
    "\n",
    "for chromnum in range(1,23):\n",
    "    print('-'*100)\n",
    "    print(\"PROCESSING CHROMOSOME\", str(chromnum))\n",
    "    \n",
    "    chr_variant_results_df = variant_results_df[(variant_results_df['chr'] == 'chr' + str(chromnum)) & \n",
    "                                              (variant_results_df['ac_ctrl'] + variant_results_df['ac_case'] <= cnt_thr) & \n",
    "                                              (variant_results_df['an_case'] > 0) & \n",
    "                                              (variant_results_df['an_ctrl'] > 0)]\n",
    "    \n",
    "    # Subset analysis to pLoFs\n",
    "    cases_df = chr_variant_results_df[chr_variant_results_df['consequence'] == 'ptv']\n",
    "    \n",
    "    # Apply fishers exact test to get signicance scores from cases data\n",
    "    tested_df = apply_fishers_exact_test(cases_df)\n",
    "        \n",
    "    # Filter rows where p_value < significance threshold and append these rows to the DataFrame\n",
    "    significant_genes_df = pd.concat([significant_genes_df, tested_df[tested_df['p_value'] <= significance_thr]], ignore_index=True)\n",
    "    \n",
    "    # Display the sorted DataFrame\n",
    "    print(tested_df[tested_df['p_value'] < .0001])\n",
    "    \n",
    "# Write significant_genes_df to a csv to avoid recomputing\n",
    "significant_genes_df.to_csv(results_path + 'bipex_rgc_constraint_pvalues.tsv.gz', index=False, compression='gzip', sep='\\t')\n",
    "\n",
    "significant_genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "02df8a35",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Parameters for analysis\n",
    "# prob_thr = 0.6\n",
    "# cnt_thr = 5\n",
    "# significance_thr = 1\n",
    "# variant_results_df = bipex_variant_results_df\n",
    "# variant_results_df = variant_results_df[(variant_results_df['ac_ctrl'] + variant_results_df['ac_case'] <= cnt_thr) & \n",
    "#                                         (variant_results_df['an_case'] > 0) & \n",
    "#                                         (variant_results_df['an_ctrl'] > 0)]\n",
    "\n",
    "# # Subset analysis to pLoFs\n",
    "# cases_df = variant_results_df[variant_results_df['consequence'] == 'ptv']\n",
    "\n",
    "# # Apply fishers exact test to get signicance scores from cases data\n",
    "# tested_df = apply_fishers_exact_test(cases_df)\n",
    "\n",
    "# # Filter rows where p_value < significance threshold and append these rows to the DataFrame\n",
    "# significant_genes_df = tested_df[tested_df['p_value'] <= significance_thr]\n",
    "\n",
    "# # Display the sorted DataFrame\n",
    "# # print(tested_df[tested_df['p_value'] < .0001])\n",
    "    \n",
    "# # Write significant_genes_df to a csv to avoid recomputing\n",
    "# # significant_genes_df.to_csv(results_path + 'bipex_rgc_constraint_pvalues.tsv.gz', index=False, compression='gzip', sep='\\t')\n",
    "\n",
    "# significant_genes_df #[significant_genes_df['p_value'] < .0001]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4ea65f3",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "plot_group_QQ(significant_genes_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49c8014d",
   "metadata": {},
   "source": [
    "## Combine RGC constraint and Alpha Missense pathogenicity into meta regression model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a4b15a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnt_thr = 5\n",
    "\n",
    "# Group by 'chr' and 'pos' and find the maximum 'am_pathogenicity' value\n",
    "alpha_missense_grouped_df = alpha_missense_df.groupby(['chr', 'pos'])['am_pathogenicity'].max().reset_index()\n",
    "\n",
    "# Filter for case counts\n",
    "variants_filtered_df = variant_results_df[(variant_results_df['ac_ctrl'] + variant_results_df['ac_case'] <= cnt_thr) &\n",
    "                                          (variant_results_df['an_case'] > 0) &\n",
    "                                          (variant_results_df['an_ctrl'] > 0)\n",
    "                                         ].drop(columns=[\"v\", \"ref\", \"alt\", \"locus\", \"alleles\", \"hgvsc\", \"hgvsp\", \"mpc\", \"polyphen\", \"in_analysis\"])\n",
    "\n",
    "# Merge constraint predictions, am_pathogenicity predictions, and variant results based on chromosome and position\n",
    "merged_df = pd.merge(pd.merge(predictions_df, alpha_missense_grouped_df, on=['chr', 'pos'], how='inner'), variants_filtered_df, on=['chr', 'pos'], how='inner')\n",
    "\n",
    "merged_df['effect_size'] = np.log(((0.5 + merged_df['ac_case']) * (0.5 + merged_df['an_ctrl'] - merged_df['ac_ctrl'])) \n",
    "                                  / ((0.5 + merged_df['an_case'] - merged_df['ac_case']) * (0.5 + merged_df['ac_ctrl'])))\n",
    "\n",
    "merged_df['var_effect_size'] = (1 / (0.5 + merged_df['ac_case']) + 1 / (0.5 + merged_df['an_ctrl'] - merged_df['ac_ctrl']) \n",
    "                                + 1 / (0.5 + merged_df['an_case'] - merged_df['ac_case']) + 1 / (0.5 + merged_df['ac_ctrl']))\n",
    "\n",
    "# Write merged_df to a csv to repeat analysis in R\n",
    "merged_df.to_csv(results_path + \"constraint_am_bipex_results.tsv.gz\", index=False, compression='gzip', sep='\\t')\n",
    "\n",
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d7997d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Read the data from the file\n",
    "# merged_df = pd.read_csv(results_path + 'constraint_am_bipex_results.tsv.gz', sep='\\t')\n",
    "\n",
    "# Filter out pLoF\n",
    "filtered_df = merged_df[~merged_df['consequence'].isin(pLoF_consequence_list)]\n",
    "\n",
    "# Initialize lists to store results\n",
    "results = []\n",
    "\n",
    "# Group data by 'gene_id'\n",
    "grouped_data = filtered_df.groupby(['gene_id', 'gene_name', 'group'])\n",
    "\n",
    "# Loop over each gene group and build a meta-regression model\n",
    "for current_combination, gene_data in tqdm(grouped_data, desc=\"Processing genes\", unit=\"gene\"):\n",
    "    current_gene, current_name, current_group = current_combination\n",
    "    \n",
    "    # Meta-regression model for the current gene\n",
    "    X = sm.add_constant(gene_data[['prob_0', 'am_pathogenicity']])\n",
    "    y = gene_data['effect_size']\n",
    "    weights = 1 / gene_data['var_effect_size']\n",
    "\n",
    "    try:\n",
    "        model = sm.WLS(y, X, weights=weights, missing='drop').fit()\n",
    "\n",
    "        # Append results to the list\n",
    "        results.append({\n",
    "            'gene_id': current_gene,\n",
    "            'gene_name': current_name,\n",
    "            'group': current_group,\n",
    "            'p_value_prob_0': model.pvalues['prob_0'],\n",
    "            'p_value_am_pathogenicity': model.pvalues['am_pathogenicity'],\n",
    "            'p_value_const': model.pvalues['const'],\n",
    "            'p_value_overall': model.f_pvalue\n",
    "        })\n",
    "#         print(model.summary())\n",
    "#         break\n",
    "\n",
    "    except:\n",
    "#         print(f\"Skipping gene {current_gene} of group {current_group} due to an exception.\")\n",
    "        pass\n",
    "\n",
    "# Create a DataFrame from the results\n",
    "p_values_df = pd.DataFrame(results)\n",
    "\n",
    "# Save the DataFrame to a compressed CSV file\n",
    "p_values_df.to_csv(results_path + \"constraint_am_bipex_pvalues.tsv.gz\", index=False, compression='gzip', sep='\\t')\n",
    "\n",
    "# Display the DataFrame\n",
    "p_values_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54e7e38a",
   "metadata": {},
   "source": [
    "## Join pLoF and meta regression results into unified model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad3089b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# significant_genes_df = pd.read_csv(results_path + 'bipex_rgc_constraint_pvalues.tsv.gz', sep='\\t')\n",
    "# p_values_df = pd.read_csv(results_path + 'constraint_am_bipex_pvalues.tsv.gz', sep='\\t')\n",
    "\n",
    "joined_p_values_df = pd.merge(p_values_df, significant_genes_df, on=['gene_id', 'gene_name', 'group'], how='outer')\n",
    "\n",
    "joined_p_values_df['p_comb'] = joined_p_values_df.apply(lambda row: fisher_method(row['p_value_overall'], row['p_value']), axis=1)\n",
    "\n",
    "joined_p_values_df = joined_p_values_df.drop(columns=['total_ac_case', 'total_an_case', 'total_ac_ctrl', 'total_an_ctrl', 'odds_ratio'])\n",
    "\n",
    "joined_p_values_df.rename(columns={'p_value_prob_0': 'p_prob_0', 'p_value_am_pathogenicity': 'p_pathogenicity', 'p_value_const': 'p_const', 'p_value_overall': 'p_overall', 'p_value': 'p_pLoF'}, inplace=True)\n",
    "\n",
    "joined_p_values_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b19bc40",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# filter_mask = (joined_p_values_df['p_comb'] < 1e-4) | (joined_p_values_df['p_value_overall'] < 1e-4)\n",
    "filter_mask = (joined_p_values_df['p_comb'] < 1e-4) | (joined_p_values_df['p_prob_0'] < 1e-4)\n",
    "gene_name_mask = (joined_p_values_df['gene_name'] == 'SLC2A1')\n",
    "\n",
    "pd.set_option('display.max_rows', 200)\n",
    "joined_p_values_df[filter_mask].sort_values('p_comb')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c3fcb7d",
   "metadata": {},
   "source": [
    "## Compare against published results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6cc3fba",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "for group in significant_genes_df['group'].unique():\n",
    "    file_pattern = data_path + \"Autism_results_2023_12_30_14_36_32.csv\"\n",
    "    files = glob.glob(file_pattern)\n",
    "\n",
    "    if files:\n",
    "        results_pub_df = pd.read_csv(files[0], sep=\",\")\n",
    "        merged_pub_df = pd.merge(significant_genes_df[significant_genes_df['group'] == group], results_pub_df, left_on='gene_id', right_on='Gene')\n",
    "\n",
    "        merged_pub_df['minus_log_pval'] = -np.log10(merged_pub_df['p_value'])\n",
    "        merged_pub_df['minus_log_missense_pval'] = -np.log10(merged_pub_df['Q‑Val'])\n",
    "\n",
    "        # Set up the plot\n",
    "        plt.figure(figsize=(8, 6))\n",
    "\n",
    "        # Replace Inf values with NaN\n",
    "        merged_pub_df.replace([np.inf, -np.inf], np.nan, inplace=True)\n",
    "\n",
    "        # Calculate the minimum p-value, ignoring NaN values\n",
    "        min_pval = np.nanmin([merged_pub_df['minus_log_missense_pval'].min(), merged_pub_df['minus_log_pval'].min()])\n",
    "\n",
    "        # Calculate the maximum p-value, ignoring NaN values\n",
    "        max_pval = np.nanmax([merged_pub_df['minus_log_missense_pval'].max(), merged_pub_df['minus_log_pval'].max()])\n",
    "\n",
    "        # Ensure min_pval and max_pval are not NaN (in case all values are NaN or Inf)\n",
    "        if np.isnan(min_pval):\n",
    "            min_pval = 0  # or some other default value\n",
    "        if np.isnan(max_pval):\n",
    "            max_pval = 1  # or some other default value\n",
    "        # Create a hexbin (smoothscatter) plot\n",
    "        plt.hexbin(merged_pub_df['minus_log_missense_pval'], merged_pub_df['minus_log_pval'], gridsize=2, cmap='Blues')\n",
    "\n",
    "        # Add a diagonal line\n",
    "        plt.plot([min_pval, max_pval], [min_pval, max_pval], color='red', linestyle='--')\n",
    "\n",
    "        # Annotate points\n",
    "        for _, row in merged_pub_df.iterrows():\n",
    "            if (abs((row['minus_log_missense_pval'] - row['minus_log_pval'])) > 2) & (max(row['minus_log_missense_pval'], row['minus_log_pval']) > 4):\n",
    "                plt.text(row['minus_log_missense_pval'], row['minus_log_pval'], row['gene_name'], fontsize=6)\n",
    "\n",
    "        # Set axis limits\n",
    "        plt.xlim(min_pval, max_pval)\n",
    "        plt.ylim(min_pval, max_pval)\n",
    "\n",
    "        # Axis labels and title\n",
    "        plt.xlabel('-log10(Missense p-val)')\n",
    "        plt.ylabel('-log10(p-value) from merged_empty_df')\n",
    "        plt.title(f'Plot for Group: {group}')\n",
    "      #  plt.colorbar()  # Add a color bar to the hexbin plot\n",
    "\n",
    "        plt.show()\n",
    "\n",
    "    else:\n",
    "        print(f\"No files found for group {group}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
