{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b67e8655",
   "metadata": {
    "id": "vECLNfPmhyDb"
   },
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a1268ef9",
   "metadata": {
    "id": "bsBOtyuTAJsC",
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: numpy in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (1.24.3)\n",
      "Requirement already satisfied: pandas in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (1.5.3)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from pandas) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from pandas) (2022.7)\n",
      "Requirement already satisfied: numpy>=1.21.0 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from pandas) (1.24.3)\n",
      "Requirement already satisfied: six>=1.5 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from python-dateutil>=2.8.1->pandas) (1.16.0)\n",
      "Requirement already satisfied: scipy in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (1.11.4)\n",
      "Requirement already satisfied: numpy<1.28.0,>=1.21.6 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from scipy) (1.24.3)\n",
      "Requirement already satisfied: hmmlearn in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (0.3.0)\n",
      "Requirement already satisfied: numpy>=1.10 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from hmmlearn) (1.24.3)\n",
      "Requirement already satisfied: scikit-learn!=0.22.0,>=0.16 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from hmmlearn) (1.3.0)\n",
      "Requirement already satisfied: scipy>=0.19 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from hmmlearn) (1.11.4)\n",
      "Requirement already satisfied: joblib>=1.1.1 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from scikit-learn!=0.22.0,>=0.16->hmmlearn) (1.2.0)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from scikit-learn!=0.22.0,>=0.16->hmmlearn) (2.2.0)\n",
      "Requirement already satisfied: statsmodels in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (0.14.0)\n",
      "Requirement already satisfied: numpy>=1.18 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from statsmodels) (1.24.3)\n",
      "Requirement already satisfied: scipy!=1.9.2,>=1.4 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from statsmodels) (1.11.4)\n",
      "Requirement already satisfied: pandas>=1.0 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from statsmodels) (1.5.3)\n",
      "Requirement already satisfied: patsy>=0.5.2 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from statsmodels) (0.5.3)\n",
      "Requirement already satisfied: packaging>=21.3 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from statsmodels) (23.0)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from pandas>=1.0->statsmodels) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from pandas>=1.0->statsmodels) (2022.7)\n",
      "Requirement already satisfied: six in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from patsy>=0.5.2->statsmodels) (1.16.0)\n",
      "Requirement already satisfied: mpl_scatter_density in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (0.7)\n",
      "Requirement already satisfied: numpy in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from mpl_scatter_density) (1.24.3)\n",
      "Requirement already satisfied: matplotlib>=3.0 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from mpl_scatter_density) (3.7.1)\n",
      "Requirement already satisfied: fast-histogram>=0.3 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from mpl_scatter_density) (0.12)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from matplotlib>=3.0->mpl_scatter_density) (1.0.5)\n",
      "Requirement already satisfied: cycler>=0.10 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from matplotlib>=3.0->mpl_scatter_density) (0.11.0)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from matplotlib>=3.0->mpl_scatter_density) (4.25.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from matplotlib>=3.0->mpl_scatter_density) (1.4.4)\n",
      "Requirement already satisfied: packaging>=20.0 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from matplotlib>=3.0->mpl_scatter_density) (23.0)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from matplotlib>=3.0->mpl_scatter_density) (9.4.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from matplotlib>=3.0->mpl_scatter_density) (3.0.9)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from matplotlib>=3.0->mpl_scatter_density) (2.8.2)\n",
      "Requirement already satisfied: six>=1.5 in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (from python-dateutil>=2.7->matplotlib>=3.0->mpl_scatter_density) (1.16.0)\n",
      "Requirement already satisfied: tqdm in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (4.65.0)\n",
      "Requirement already satisfied: colorama in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (0.4.6)\n",
      "Requirement already satisfied: pyliftover in /oak/stanford/groups/mrivas/users/osthoag/anaconda3/lib/python3.11/site-packages (0.4)\n"
     ]
    }
   ],
   "source": [
    "!pip install numpy\n",
    "!pip install pandas\n",
    "!pip install scipy\n",
    "!pip install hmmlearn\n",
    "!pip install statsmodels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "399ed27c",
   "metadata": {
    "executionInfo": {
     "elapsed": 8,
     "status": "ok",
     "timestamp": 1703779362738,
     "user": {
      "displayName": "Oscar Thomas Aguilar",
      "userId": "15151607230073393698"
     },
     "user_tz": 300
    },
    "id": "BZfHQLuPhyDl"
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from hmmlearn import hmm\n",
    "import gzip\n",
    "from tqdm import tqdm\n",
    "import statsmodels.api as sm\n",
    "import scipy.stats as stats\n",
    "from scipy.stats import fisher_exact\n",
    "import warnings\n",
    "\n",
    "# Suppress all warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e5621862",
   "metadata": {
    "executionInfo": {
     "elapsed": 173,
     "status": "ok",
     "timestamp": 1703790405518,
     "user": {
      "displayName": "Oscar Thomas Aguilar",
      "userId": "15151607230073393698"
     },
     "user_tz": 300
    },
    "id": "9IWHWPQ_hyDm"
   },
   "outputs": [],
   "source": [
    "# Specify project directories in Sherlock\n",
    "data_path = '/oak/stanford/groups/mrivas/projects/wgs-constraint-llm/data/'\n",
    "results_path = '/oak/stanford/groups/mrivas/projects/wgs-constraint-llm/osthoag/wgs-constraint-llm/results/'\n",
    "\n",
    "# Specify the file paths\n",
    "ukb_variants_file_path = data_path + 'ukb24310_c19.qced_nonrel.vmiss.gz'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1f17326",
   "metadata": {
    "id": "vECLNfPmhyDb"
   },
   "source": [
    "## Define helper methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "84d6f6a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sequence(coverage_df, variants_df):\n",
    "    # Get the length of the genetic sequence\n",
    "    sequence_length = max(coverage_df['pos'].max(), variants_df['pos'].max())\n",
    "\n",
    "    # Create boolean mask for exome coverage\n",
    "    coverage_mask = np.zeros(sequence_length + 1, dtype=bool)\n",
    "\n",
    "    # Use NumPy boolean indexing to get mask for positions with over 80% coverage\n",
    "    coverage_mask[coverage_df['pos'].to_numpy()] = 1\n",
    "\n",
    "    # Initialize values to zero for all positions\n",
    "    sequence = np.zeros(sequence_length + 1)\n",
    "\n",
    "    # Set positions to 1 where a variant exists\n",
    "    sequence[variants_df['pos'].to_numpy()] = 1\n",
    "\n",
    "    # Filter for only the protein-coding regions with over 80% exome coverage\n",
    "    observations = np.array(sequence[coverage_mask])\n",
    "    \n",
    "    positions = np.where(coverage_mask)[0]\n",
    "    \n",
    "    return observations, positions\n",
    "\n",
    "def get_HMM_predictions(observations, model, order=2):\n",
    "    # Flatten the higher-order structure\n",
    "    X = np.stack([observations[i:i-order] for i in range(order)], axis=1)\n",
    "\n",
    "    # Convert observations to counts\n",
    "    X_counts = np.column_stack([(X == i).sum(axis=1) for i in range(2)])\n",
    "\n",
    "    # Predict probabilities for each position\n",
    "    probabilities = model.predict_proba(X_counts)\n",
    "\n",
    "    return probabilities\n",
    "\n",
    "def fit_HMM(observations, order=2):\n",
    "    # Flatten the higher-order structure\n",
    "    X = np.stack([observations[i:i-order] for i in range(order)], axis=1)\n",
    "\n",
    "    # Convert observations to counts\n",
    "    X_counts = np.column_stack([(X == i).sum(axis=1) for i in range(2)])\n",
    "\n",
    "    # Create and fit a first-order HMM\n",
    "    model = hmm.MultinomialHMM(n_components=2, random_state=10)\n",
    "    model.fit(X_counts)\n",
    "\n",
    "    return model\n",
    "\n",
    "def ols_regression(predictions_df):\n",
    "    # Add a constant term to the independent variable for the intercept\n",
    "    X = sm.add_constant(predictions_df['prob_0'])\n",
    "\n",
    "    # Fit the linear regression model\n",
    "    model4 = sm.OLS(predictions_df['observation'],X).fit()\n",
    "\n",
    "    model4.summary().tables[1].pvalues_precision = 100  # Adjust the number of significant digits\n",
    "\n",
    "    # Get the summary of the regression\n",
    "    print(model4.summary())\n",
    "\n",
    "    # Extract the F-statistic and its associated p-value\n",
    "    f_statistic = model4.fvalue\n",
    "    p_value_f_statistic = model4.f_pvalue\n",
    "    \n",
    "    return f_statistic, p_value_f_statistic\n",
    "\n",
    "def plot_hist_from_predictions(predictions_df):\n",
    "    # Create a figure with two subplots in one row and two columns\n",
    "    fig, axs = plt.subplots(1, 2, figsize=(12, 5))\n",
    "\n",
    "    # Plot the first histogram in the first subplot\n",
    "    axs[0].hist(predictions_df['observation'], edgecolor=\"blue\")\n",
    "    axs[0].set_title('Histogram of Observations')\n",
    "    axs[0].set_xlabel('Has a variant')\n",
    "    axs[0].set_ylabel('Frequency')\n",
    "\n",
    "    # Plot the second histogram in the second subplot\n",
    "    axs[1].hist(predictions_df['prob_0'], edgecolor=\"blue\")\n",
    "    axs[1].set_title('Histogram of Predictions')\n",
    "    axs[1].set_xlabel('Probability of 0')\n",
    "    axs[1].set_ylabel('Frequency')\n",
    "\n",
    "    # Adjust layout to prevent clipping of titles\n",
    "    plt.tight_layout()\n",
    "\n",
    "    # Show the plots\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6b15354",
   "metadata": {
    "id": "yAEgmgmSGY7d"
   },
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2a1e542e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>ID</th>\n",
       "      <th>MISSING_CT</th>\n",
       "      <th>OBS_CT</th>\n",
       "      <th>F_MISS</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>19</td>\n",
       "      <td>chr19:61721:C:G</td>\n",
       "      <td>38465</td>\n",
       "      <td>393852</td>\n",
       "      <td>0.097664</td>\n",
       "      <td>chr19</td>\n",
       "      <td>61721</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>19</td>\n",
       "      <td>chr19:61722:A:C</td>\n",
       "      <td>36449</td>\n",
       "      <td>393852</td>\n",
       "      <td>0.092545</td>\n",
       "      <td>chr19</td>\n",
       "      <td>61722</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>19</td>\n",
       "      <td>chr19:61722:A:G</td>\n",
       "      <td>36449</td>\n",
       "      <td>393852</td>\n",
       "      <td>0.092545</td>\n",
       "      <td>chr19</td>\n",
       "      <td>61722</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>19</td>\n",
       "      <td>chr19:61724:T:A</td>\n",
       "      <td>33923</td>\n",
       "      <td>393852</td>\n",
       "      <td>0.086131</td>\n",
       "      <td>chr19</td>\n",
       "      <td>61724</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>19</td>\n",
       "      <td>chr19:61727:G:A</td>\n",
       "      <td>30490</td>\n",
       "      <td>393852</td>\n",
       "      <td>0.077415</td>\n",
       "      <td>chr19</td>\n",
       "      <td>61727</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28664428</th>\n",
       "      <td>19</td>\n",
       "      <td>chr19:58607407:G:T</td>\n",
       "      <td>38549</td>\n",
       "      <td>393852</td>\n",
       "      <td>0.097877</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58607407</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28664429</th>\n",
       "      <td>19</td>\n",
       "      <td>chr19:58607408:G:A</td>\n",
       "      <td>38520</td>\n",
       "      <td>393852</td>\n",
       "      <td>0.097803</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58607408</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28664430</th>\n",
       "      <td>19</td>\n",
       "      <td>chr19:58607408:G:T</td>\n",
       "      <td>38520</td>\n",
       "      <td>393852</td>\n",
       "      <td>0.097803</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58607408</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28664431</th>\n",
       "      <td>19</td>\n",
       "      <td>chr19:58607409:G:T</td>\n",
       "      <td>38675</td>\n",
       "      <td>393852</td>\n",
       "      <td>0.098197</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58607409</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28664438</th>\n",
       "      <td>19</td>\n",
       "      <td>chr19:58607414:GCC:G</td>\n",
       "      <td>39051</td>\n",
       "      <td>393852</td>\n",
       "      <td>0.099152</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58607414</td>\n",
       "      <td>GCC</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26993675 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          #CHROM                    ID  MISSING_CT  OBS_CT    F_MISS    chr  \\\n",
       "167           19       chr19:61721:C:G       38465  393852  0.097664  chr19   \n",
       "168           19       chr19:61722:A:C       36449  393852  0.092545  chr19   \n",
       "169           19       chr19:61722:A:G       36449  393852  0.092545  chr19   \n",
       "170           19       chr19:61724:T:A       33923  393852  0.086131  chr19   \n",
       "171           19       chr19:61727:G:A       30490  393852  0.077415  chr19   \n",
       "...          ...                   ...         ...     ...       ...    ...   \n",
       "28664428      19    chr19:58607407:G:T       38549  393852  0.097877  chr19   \n",
       "28664429      19    chr19:58607408:G:A       38520  393852  0.097803  chr19   \n",
       "28664430      19    chr19:58607408:G:T       38520  393852  0.097803  chr19   \n",
       "28664431      19    chr19:58607409:G:T       38675  393852  0.098197  chr19   \n",
       "28664438      19  chr19:58607414:GCC:G       39051  393852  0.099152  chr19   \n",
       "\n",
       "               pos  ref alt  \n",
       "167          61721    C   G  \n",
       "168          61722    A   C  \n",
       "169          61722    A   G  \n",
       "170          61724    T   A  \n",
       "171          61727    G   A  \n",
       "...            ...  ...  ..  \n",
       "28664428  58607407    G   T  \n",
       "28664429  58607408    G   A  \n",
       "28664430  58607408    G   T  \n",
       "28664431  58607409    G   T  \n",
       "28664438  58607414  GCC   G  \n",
       "\n",
       "[26993675 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the compressed file into a pandas DataFrame\n",
    "ukb_variants_df = pd.read_csv(ukb_variants_file_path, compression='gzip', sep='\\t')\n",
    "\n",
    "# Filter locations for F_MISS < 0.1\n",
    "ukb_variants_df = ukb_variants_df[ukb_variants_df['F_MISS'] < 0.1]\n",
    "\n",
    "# Expand ID column to get pos\n",
    "ukb_variants_df[['chr', 'pos', 'ref', 'alt']] = ukb_variants_df['ID'].str.split(':', expand=True)\n",
    "\n",
    "# Conver pos to int\n",
    "ukb_variants_df['pos'] = ukb_variants_df['pos'].astype(int)\n",
    "\n",
    "ukb_variants_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8fa4b6b",
   "metadata": {},
   "source": [
    "## Get predictions for UKB chromosome 19 from the HMM trained on AoU chromosome 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "da1fa608",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Define pattern for coverage file paths\n",
    "coverage_file_prefix = data_path + 'gnomad.genomes.r3.0.1.coverage_0.5_over10_chr'\n",
    "coverage_file_suffix = '.summary.tsv.gz'\n",
    "\n",
    "# Parameters for analysis\n",
    "order = 2\n",
    "\n",
    "# Define path to trained HMM\n",
    "hmm_file_path = results_path + 'HMM_aou_wgs_chr2_model.joblib'\n",
    "\n",
    "# Load the pretrained model\n",
    "model = joblib.load(hmm_file_path)\n",
    "\n",
    "# Get coverage file for chromosome 19\n",
    "chr19_coverage_df = pd.read_csv(coverage_file_prefix + '19' + coverage_file_suffix, sep='\\t', names=['pos'])\n",
    "\n",
    "chr19_coverage_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04ed599e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get coverage file for chromosome 19\n",
    "chr_coverage_df = pd.read_csv(coverage_file_prefix + '19' + coverage_file_suffix, sep='\\t', names=['pos'])\n",
    "\n",
    "chr_variants_df = ukb_variants_df\n",
    "\n",
    "# Get training data for the chromosome\n",
    "observations, positions = get_sequence(chr_coverage_df, chr_variants_df)\n",
    "\n",
    "# Fit HMM and retrieve probabilites\n",
    "probabilities = get_HMM_predictions(observations, model, order=order)\n",
    "\n",
    "# Create a DataFrame with 'pos' reflecting the index of the original sequence and 'prob_0/1' as the predictions\n",
    "chr_predictions_df = pd.DataFrame({'chr': 'chr19',\n",
    "                                   'pos': positions[0:-order],\n",
    "                                   'prob_0': probabilities[:, 0], \n",
    "                                   'prob_1': probabilities[:, 1], \n",
    "                                   'observation': observations[0:-order]\n",
    "                                  })\n",
    "\n",
    "# Checkpoint chr_predictions_df to a csv to avoid memory bottleneck\n",
    "chr_predictions_df.to_csv(results_path + \"HMM_ukb_constraint_predictions_chr19.tsv.gz\", index=False, compression='gzip', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cda63593",
   "metadata": {},
   "source": [
    "## Train HMM on UKB chromosome 19 and get predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a560ebc2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MultinomialHMM has undergone major changes. The previous version was implementing a CategoricalHMM (a special case of MultinomialHMM). This new implementation follows the standard definition for a Multinomial distribution (e.g. as in https://en.wikipedia.org/wiki/Multinomial_distribution). See these issues for details:\n",
      "https://github.com/hmmlearn/hmmlearn/issues/335\n",
      "https://github.com/hmmlearn/hmmlearn/issues/340\n"
     ]
    }
   ],
   "source": [
    "# Define pattern for coverage file paths\n",
    "coverage_file_prefix = data_path + 'gnomad.genomes.r3.0.1.coverage_0.5_over10_chr'\n",
    "coverage_file_suffix = '.summary.tsv.gz'\n",
    "\n",
    "# Parameters for analysis\n",
    "order = 2\n",
    "\n",
    "# Define path to trained HMM\n",
    "hmm_file_path = results_path + 'HMM_ukb_wgs_chr19_model.joblib'\n",
    "\n",
    "# Filter rows for chromosome 19\n",
    "chr_variants_df = ukb_variants_df\n",
    "chr_coverage_df = pd.read_csv(coverage_file_prefix + '19' + coverage_file_suffix, sep='\\t', names=['pos'])\n",
    "\n",
    "# Get training data for the chromosome\n",
    "observations, positions = get_sequence(chr_coverage_df, chr_variants_df)\n",
    "\n",
    "# Fit HMM to Chromosome 2\n",
    "model = fit_HMM(observations, order=order)\n",
    "\n",
    "# Save the HMM model to a file to avoid retraining\n",
    "joblib.dump(model, hmm_file_path)\n",
    "\n",
    "# Fit HMM and retrieve probabilites\n",
    "probabilities = get_HMM_predictions(observations, model, order=order)\n",
    "\n",
    "# Create a DataFrame with 'pos' reflecting the index of the original sequence and 'prob_0/1' as the predictions\n",
    "chr_predictions_df = pd.DataFrame({'chr': 'chr19',\n",
    "                                   'pos': positions[0:-order],\n",
    "                                   'prob_0': probabilities[:, 0], \n",
    "                                   'prob_1': probabilities[:, 1], \n",
    "                                   'observation': observations[0:-order]\n",
    "                                  })\n",
    "\n",
    "# Checkpoint chr_predictions_df to a csv to avoid memory bottleneck\n",
    "chr_predictions_df.to_csv(results_path + \"HMM_ukb19_constraint_predictions_chr19.tsv.gz\", index=False, compression='gzip', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cab790bc",
   "metadata": {},
   "source": [
    "## Compare results to AoU Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e30a29e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>prob_0_aou</th>\n",
       "      <th>prob_1_aou</th>\n",
       "      <th>observation_aou</th>\n",
       "      <th>prob_0_ukb</th>\n",
       "      <th>prob_1_ukb</th>\n",
       "      <th>observation_ukb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr19</td>\n",
       "      <td>60645</td>\n",
       "      <td>2.621388e-35</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.210649e-16</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr19</td>\n",
       "      <td>60646</td>\n",
       "      <td>6.635288e-01</td>\n",
       "      <td>0.336471</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.981195</td>\n",
       "      <td>1.880520e-02</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr19</td>\n",
       "      <td>60647</td>\n",
       "      <td>8.721584e-01</td>\n",
       "      <td>0.127842</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.975906</td>\n",
       "      <td>2.409405e-02</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr19</td>\n",
       "      <td>60648</td>\n",
       "      <td>9.377551e-01</td>\n",
       "      <td>0.062245</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.974418</td>\n",
       "      <td>2.558151e-02</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr19</td>\n",
       "      <td>60649</td>\n",
       "      <td>9.583747e-01</td>\n",
       "      <td>0.041625</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.974000</td>\n",
       "      <td>2.599985e-02</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55042752</th>\n",
       "      <td>chr19</td>\n",
       "      <td>58607370</td>\n",
       "      <td>5.260989e-04</td>\n",
       "      <td>0.999474</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001104</td>\n",
       "      <td>9.988959e-01</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55042753</th>\n",
       "      <td>chr19</td>\n",
       "      <td>58607371</td>\n",
       "      <td>2.120975e-06</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>9.999916e-01</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55042754</th>\n",
       "      <td>chr19</td>\n",
       "      <td>58607372</td>\n",
       "      <td>2.271861e-06</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.001104</td>\n",
       "      <td>9.988959e-01</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55042755</th>\n",
       "      <td>chr19</td>\n",
       "      <td>58607373</td>\n",
       "      <td>3.873230e-03</td>\n",
       "      <td>0.996127</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.001105</td>\n",
       "      <td>9.988955e-01</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55042756</th>\n",
       "      <td>chr19</td>\n",
       "      <td>58607374</td>\n",
       "      <td>3.033034e-01</td>\n",
       "      <td>0.696697</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>9.999742e-01</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>55042757 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            chr       pos    prob_0_aou  prob_1_aou  observation_aou  \\\n",
       "0         chr19     60645  2.621388e-35    1.000000              0.0   \n",
       "1         chr19     60646  6.635288e-01    0.336471              0.0   \n",
       "2         chr19     60647  8.721584e-01    0.127842              0.0   \n",
       "3         chr19     60648  9.377551e-01    0.062245              0.0   \n",
       "4         chr19     60649  9.583747e-01    0.041625              0.0   \n",
       "...         ...       ...           ...         ...              ...   \n",
       "55042752  chr19  58607370  5.260989e-04    0.999474              0.0   \n",
       "55042753  chr19  58607371  2.120975e-06    0.999998              1.0   \n",
       "55042754  chr19  58607372  2.271861e-06    0.999998              1.0   \n",
       "55042755  chr19  58607373  3.873230e-03    0.996127              1.0   \n",
       "55042756  chr19  58607374  3.033034e-01    0.696697              0.0   \n",
       "\n",
       "          prob_0_ukb    prob_1_ukb  observation_ukb  \n",
       "0           1.000000  3.210649e-16              0.0  \n",
       "1           0.981195  1.880520e-02              0.0  \n",
       "2           0.975906  2.409405e-02              0.0  \n",
       "3           0.974418  2.558151e-02              0.0  \n",
       "4           0.974000  2.599985e-02              0.0  \n",
       "...              ...           ...              ...  \n",
       "55042752    0.001104  9.988959e-01              0.0  \n",
       "55042753    0.000008  9.999916e-01              1.0  \n",
       "55042754    0.001104  9.988959e-01              1.0  \n",
       "55042755    0.001105  9.988955e-01              0.0  \n",
       "55042756    0.000026  9.999742e-01              1.0  \n",
       "\n",
       "[55042757 rows x 8 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aou_chr19_predictions_df = pd.read_csv(results_path + \"HMM_aou_constraint_predictions_chr19.tsv.gz\", compression='gzip', sep='\\t')\n",
    "ukb_chr19_predictions_df = pd.read_csv(results_path + \"HMM_ukb19_constraint_predictions_chr19.tsv.gz\", compression='gzip', sep='\\t')\n",
    "\n",
    "# Merge the DataFrames on the columns 'chr' and 'pos'\n",
    "merged_df = pd.merge(aou_chr19_predictions_df, ukb_chr19_predictions_df, on=['chr', 'pos'], suffixes=('_aou', '_ukb'))\n",
    "\n",
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1b3b2771",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.4297016502484097\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:        observation_ukb   No. Observations:             55042757\n",
      "Model:                            GLM   Df Residuals:                 55042755\n",
      "Model Family:                Binomial   Df Model:                            1\n",
      "Link Function:                  Logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:            -2.0827e+07\n",
      "Date:                Sun, 18 Feb 2024   Deviance:                   4.1654e+07\n",
      "Time:                        17:35:52   Pearson chi2:                 3.27e+07\n",
      "No. Iterations:                    13   Pseudo R-squ. (CS):             0.4346\n",
      "Covariance Type:            nonrobust                                         \n",
      "==============================================================================\n",
      "                 coef    std err          z      P>|z|      [0.025      0.975]\n",
      "------------------------------------------------------------------------------\n",
      "const          0.9509      0.000   1929.704      0.000       0.950       0.952\n",
      "prob_0_ukb   -71.6727      0.055  -1296.297      0.000     -71.781     -71.564\n",
      "==============================================================================\n"
     ]
    }
   ],
   "source": [
    "X = merged_df['prob_0_ukb']\n",
    "X = sm.add_constant(X)\n",
    "y = merged_df['observation_ukb']\n",
    "\n",
    "# Fit the logistic regression model\n",
    "model = sm.GLM(y, X, family=sm.families.Binomial()).fit()\n",
    "\n",
    "print(model.pseudo_rsquared('McFadden'))\n",
    "print(model.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bf6c9187",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.03993771719471406\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:        observation_aou   No. Observations:             55042757\n",
      "Model:                            GLM   Df Residuals:                 55042755\n",
      "Model Family:                Binomial   Df Model:                            1\n",
      "Link Function:                  Logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:            -3.4186e+07\n",
      "Date:                Sun, 18 Feb 2024   Deviance:                   6.8371e+07\n",
      "Time:                        17:31:12   Pearson chi2:                 5.54e+07\n",
      "No. Iterations:                     5   Pseudo R-squ. (CS):            0.05036\n",
      "Covariance Type:            nonrobust                                         \n",
      "==============================================================================\n",
      "                 coef    std err          z      P>|z|      [0.025      0.975]\n",
      "------------------------------------------------------------------------------\n",
      "const         -0.2899      0.000   -852.626      0.000      -0.291      -0.289\n",
      "prob_0_ukb    -1.4549      0.001  -1583.754      0.000      -1.457      -1.453\n",
      "==============================================================================\n"
     ]
    }
   ],
   "source": [
    "X = merged_df['prob_0_ukb']\n",
    "X = sm.add_constant(X)\n",
    "y = merged_df['observation_aou']\n",
    "\n",
    "# Fit the logistic regression model\n",
    "model = sm.GLM(y, X, family=sm.families.Binomial()).fit()\n",
    "\n",
    "print(model.pseudo_rsquared('McFadden'))\n",
    "print(model.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "48d2970c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                            OLS Regression Results                            \n",
      "==============================================================================\n",
      "Dep. Variable:             prob_0_aou   R-squared:                       0.095\n",
      "Model:                            OLS   Adj. R-squared:                  0.095\n",
      "Method:                 Least Squares   F-statistic:                 5.767e+06\n",
      "Date:                Sun, 18 Feb 2024   Prob (F-statistic):               0.00\n",
      "Time:                        17:38:12   Log-Likelihood:            -1.8981e+07\n",
      "No. Observations:            55042757   AIC:                         3.796e+07\n",
      "Df Residuals:                55042755   BIC:                         3.796e+07\n",
      "Df Model:                           1                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "==============================================================================\n",
      "                 coef    std err          t      P>|t|      [0.025      0.975]\n",
      "------------------------------------------------------------------------------\n",
      "const          0.1957   5.71e-05   3430.245      0.000       0.196       0.196\n",
      "prob_0_ukb     0.3101      0.000   2401.443      0.000       0.310       0.310\n",
      "==============================================================================\n",
      "Omnibus:                  6694851.171   Durbin-Watson:                   0.447\n",
      "Prob(Omnibus):                  0.000   Jarque-Bera (JB):          6191686.260\n",
      "Skew:                           0.749   Prob(JB):                         0.00\n",
      "Kurtosis:                       2.327   Cond. No.                         3.02\n",
      "==============================================================================\n",
      "\n",
      "Notes:\n",
      "[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.\n"
     ]
    }
   ],
   "source": [
    "X = merged_df['prob_0_ukb']\n",
    "X = sm.add_constant(X)\n",
    "y = merged_df['prob_0_aou']\n",
    "\n",
    "# Fit the logistic regression model\n",
    "model = sm.OLS(y, X).fit()\n",
    "\n",
    "# print(model.pseudo_rsquared('McFadden'))\n",
    "print(model.summary())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
