{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b7610e27",
   "metadata": {
    "id": "vECLNfPmhyDb"
   },
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9599826",
   "metadata": {
    "id": "bsBOtyuTAJsC",
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "!pip install numpy\n",
    "!pip install pandas\n",
    "!pip install scipy\n",
    "!pip install hmmlearn\n",
    "!pip install statsmodels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9f00b492",
   "metadata": {
    "executionInfo": {
     "elapsed": 8,
     "status": "ok",
     "timestamp": 1703779362738,
     "user": {
      "displayName": "Oscar Thomas Aguilar",
      "userId": "15151607230073393698"
     },
     "user_tz": 300
    },
    "id": "BZfHQLuPhyDl"
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from hmmlearn import hmm\n",
    "import gzip\n",
    "import os\n",
    "from tqdm import tqdm, trange\n",
    "import statsmodels.api as sm\n",
    "import scipy.stats as stats\n",
    "from scipy.stats import fisher_exact\n",
    "import math\n",
    "import joblib\n",
    "import warnings\n",
    "\n",
    "# Suppress all warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "feae3e50",
   "metadata": {
    "executionInfo": {
     "elapsed": 173,
     "status": "ok",
     "timestamp": 1703790405518,
     "user": {
      "displayName": "Oscar Thomas Aguilar",
      "userId": "15151607230073393698"
     },
     "user_tz": 300
    },
    "id": "9IWHWPQ_hyDm"
   },
   "outputs": [],
   "source": [
    "# Specify project directories in Sherlock\n",
    "data_path = '/oak/stanford/groups/mrivas/projects/wgs-constraint-llm/data/'\n",
    "results_path = '/oak/stanford/groups/mrivas/projects/wgs-constraint-llm/osthoag/wgs-constraint-llm/results/'\n",
    "\n",
    "# Specify the file paths\n",
    "variants_file_path = data_path + 'aou_variants.tsv.gz'\n",
    "gene_annotation_file_path = data_path + 'gencode.v44.basic.annotation.gtf.gz'\n",
    "genome_coverage_file_path = data_path + 'gnomad.genomes.r3.0.1.coverage.summary.tsv.bgz'\n",
    "coverage_file_path = data_path + 'gnomad.exomes.v4.0.coverage.summary.tsv.bgz'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a232aea",
   "metadata": {
    "id": "vECLNfPmhyDb"
   },
   "source": [
    "## Define helper methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "55b70a71",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to extract the observation sequence for the HMM\n",
    "def get_sequence(chromnum):\n",
    "    # Filter gene df for given chromosome\n",
    "    chr_gene_df = gene_df[gene_df['chr'] == 'chr' + str(chromnum)]\n",
    "    \n",
    "    # Filter variant df for given chromosome\n",
    "    chr_variants_df = variants_df[variants_df['chr'] == chromnum]\n",
    "    \n",
    "    # Filter coverage data for chromosome 2 based on the coverage threshold\n",
    "    chr_coverage_df = coverage_df[(coverage_df['chr'] == 'chr' + str(chromnum))]\n",
    "    \n",
    "    # Determine the length of the entire sequence based on gene, coverage, and variant data\n",
    "    sequence_length = max(chr_gene_df['end'].max(), chr_coverage_df['pos'].max(), chr_variants_df['pos'].max())\n",
    "\n",
    "    # Initialize a boolean mask for protein-coding regions\n",
    "    protein_coding_mask = np.zeros(sequence_length + 1, dtype=bool)\n",
    "\n",
    "    # Mark positions that are within protein-coding regions\n",
    "    for start, end in zip(chr_gene_df['start'].to_numpy(), chr_gene_df['end'].to_numpy()):\n",
    "        protein_coding_mask[start:end+1] = 1\n",
    "\n",
    "    # Initialize a boolean mask for regions with sufficient exome coverage\n",
    "    coverage_mask = np.zeros(sequence_length + 1, dtype=bool)\n",
    "\n",
    "    # Mark positions with coverage greater than 80%\n",
    "    coverage_mask[chr_coverage_df['pos'].to_numpy()] = 1\n",
    "\n",
    "    # Initialize the sequence array (observations) to zeros\n",
    "    sequence = np.zeros(sequence_length + 1)\n",
    "\n",
    "    # Mark positions where variants are observed\n",
    "    sequence[chr_variants_df['pos'].to_numpy()] = 1\n",
    "\n",
    "    # Filter the sequence for protein-coding regions with sufficient coverage\n",
    "    observations = np.array(sequence[protein_coding_mask & coverage_mask])\n",
    "    \n",
    "    # Identify the positions that correspond to these observations\n",
    "    positions = np.where(protein_coding_mask & coverage_mask)[0]\n",
    "    \n",
    "    return observations, positions\n",
    "\n",
    "# Function to fit an HMM to the observations\n",
    "def fit_HMM(observations, order=2):\n",
    "    # Prepare the observations for HMM fitting by creating higher-order observations\n",
    "    X = np.stack([observations[i:i-order] for i in range(order)], axis=1)\n",
    "\n",
    "    # Convert observations to counts\n",
    "    X_counts = np.column_stack([(X == i).sum(axis=1) for i in range(2)])\n",
    "\n",
    "    # Instantiate and fit a first-order HMM\n",
    "    model = hmm.MultinomialHMM(n_components=2, random_state=10)\n",
    "    model.fit(X_counts)\n",
    "\n",
    "    return model\n",
    "\n",
    "# Function to predict probabilities using a fitted HMM\n",
    "def get_HMM_predictions(observations, model, order=2):\n",
    "    # Create higher-order observations (context of order-1)\n",
    "    X = np.stack([observations[i:i-order] for i in range(order)], axis=1)\n",
    "\n",
    "    # Convert observations to counts for the HMM\n",
    "    X_counts = np.column_stack([(X == i).sum(axis=1) for i in range(2)])\n",
    "\n",
    "    # Predict probabilities for each position in the sequence\n",
    "    probabilities = model.predict_proba(X_counts)\n",
    "\n",
    "    return probabilities\n",
    "\n",
    "# Function to perform Ordinary Least Squares (OLS) regression on prediction data\n",
    "def ols_regression(predictions_df):\n",
    "    # Add a constant term to the predictor variable\n",
    "    X = sm.add_constant(predictions_df['prob_0'])\n",
    "\n",
    "    # Fit the OLS model and print the summary\n",
    "    model4 = sm.OLS(predictions_df['observation'],X).fit()\n",
    "\n",
    "    model4.summary().tables[1].pvalues_precision = 100  # Adjust the number of significant digits\n",
    "\n",
    "    # Print the regression summary\n",
    "    print(model4.summary())\n",
    "\n",
    "    # Extract the F-statistic and associated p-value\n",
    "    f_statistic = model4.fvalue\n",
    "    p_value_f_statistic = model4.f_pvalue\n",
    "    \n",
    "    return f_statistic, p_value_f_statistic\n",
    "\n",
    "# Function to plot histograms of observed variants and HMM predictions\n",
    "def plot_hist_from_predictions(predictions_df):\n",
    "    # Create a figure with two subplots in one row and two columns\n",
    "    fig, axs = plt.subplots(1, 2, figsize=(12, 5))\n",
    "\n",
    "    # Plot the first histogram in the first subplot\n",
    "    axs[0].hist(predictions_df['observation'], edgecolor=\"blue\")\n",
    "    axs[0].set_title('Histogram of Observations')\n",
    "    axs[0].set_xlabel('Has a variant')\n",
    "    axs[0].set_ylabel('Frequency')\n",
    "\n",
    "    # Plot the second histogram in the second subplot\n",
    "    axs[1].hist(predictions_df['prob_0'], edgecolor=\"blue\")\n",
    "    axs[1].set_title('Histogram of Predictions')\n",
    "    axs[1].set_xlabel('Probability of 0')\n",
    "    axs[1].set_ylabel('Frequency')\n",
    "\n",
    "    # Adjust layout to prevent clipping of titles\n",
    "    plt.tight_layout()\n",
    "\n",
    "    # Show the plots\n",
    "    plt.show()\n",
    "    \n",
    "# Function to plot observed mutations for a subsequence\n",
    "def plot_subsequence_observations(observations, start_idx, end_idx, name='Target Region'): \n",
    "    plt.bar(range(start_idx, end_idx), observations[start_idx:end_idx], color='black', width=1.0)#, label='Original Sequence')\n",
    "    plt.title(\"Observed mutations for part of \" + name)\n",
    "    plt.xlabel('Position on Chromosome 2')\n",
    "    plt.yticks([0,1]) \n",
    "    plt.ylabel('Observation')\n",
    "#     plt.legend()\n",
    "    plt.margins(x=0)\n",
    "    plt.tight_layout(rect=[0, 0, 2, 0.5])\n",
    "    plt.show()\n",
    "    \n",
    "# Function to plot HMM predictions for a subsequence\n",
    "def plot_subsequence_predictions(predictions, start_idx, end_idx, name='Target Region'):    \n",
    "    plt.bar(range(start_idx, end_idx), predictions[start_idx:end_idx, 0], color='#8C1515', width=1.0, label='Probability of 0')\n",
    "#     plt.bar(range(start_idx, end_idx), predictions[start_idx:end_idx, 1], bottom=predictions[start_idx:end_idx, 0], color='#D2C295', label='Probability of 1')\n",
    "    plt.title(\"HMM Predictions for part of \" + name)\n",
    "    plt.xlabel('Position on Chromosome 2')\n",
    "    plt.ylabel('Probability of 0')\n",
    "#     plt.legend()\n",
    "    plt.margins(x=0)\n",
    "    plt.tight_layout(rect=[0, 0, 2, 0.5])\n",
    "    plt.show()\n",
    "    \n",
    "# Function to plot both observed mutations and HMM predictions for a subsequence\n",
    "def plot_subsequence(observations, predictions, start_idx, end_idx, gene_name='Target Region', save_name=None):\n",
    "    fig, axes = plt.subplots(nrows=2, ncols=1, figsize=(10, 4), sharex=True)\n",
    "    plt.suptitle(\"Observed Mutations vs HMM Predictions for \" + gene_name, y=0.9)\n",
    "\n",
    "    # Plot the original sequence\n",
    "    axes[0].bar(range(start_idx, end_idx), observations[start_idx:end_idx], width=1.0, color='black', label='Original Sequence')\n",
    "    axes[0].set_ylabel('Observation')\n",
    "#     axes[0].legend()\n",
    "    axes[0].margins(x=0)\n",
    "    \n",
    "    # Plot the predicted probabilities as stacked barplots\n",
    "    axes[1].bar(range(start_idx, end_idx), predictions[start_idx:end_idx, 0], width=1.0, color='#8C1515')\n",
    "#     axes[1].bar(range(start_idx, end_idx), predictions[start_idx:end_idx, 1], bottom=predictions[start_idx:end_idx, 0], color='#D2C295', label='Probability of 1')\n",
    "    axes[1].set_xlabel('Position')\n",
    "    axes[1].set_ylabel('Probability of 0')\n",
    "#     axes[1].legend()\n",
    "    axes[1].margins(x=0)\n",
    "\n",
    "    plt.tight_layout(rect=[0, 0, 1, 0.96])\n",
    "    \n",
    "    if save_name is not None:\n",
    "        plt.savefig(results_path + save_name)\n",
    "        \n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e3e9090",
   "metadata": {
    "id": "yAEgmgmSGY7d"
   },
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ba74ed58",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 424
    },
    "executionInfo": {
     "elapsed": 87986,
     "status": "ok",
     "timestamp": 1703321915656,
     "user": {
      "displayName": "Oscar Thomas Aguilar",
      "userId": "15151607230073393698"
     },
     "user_tz": 300
    },
    "id": "Bz601YCxD8vb",
    "outputId": "8d636ae6-142b-4681-ca77-89e68e1cc0d1"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>over_10</th>\n",
       "      <th>over_20</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170202922</th>\n",
       "      <td>0.29592</td>\n",
       "      <td>0.23756</td>\n",
       "      <td>chrM</td>\n",
       "      <td>16069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170202923</th>\n",
       "      <td>0.29582</td>\n",
       "      <td>0.23788</td>\n",
       "      <td>chrM</td>\n",
       "      <td>16070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170202924</th>\n",
       "      <td>0.29689</td>\n",
       "      <td>0.23808</td>\n",
       "      <td>chrM</td>\n",
       "      <td>16071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170202925</th>\n",
       "      <td>0.29674</td>\n",
       "      <td>0.23815</td>\n",
       "      <td>chrM</td>\n",
       "      <td>16072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170202926</th>\n",
       "      <td>0.29712</td>\n",
       "      <td>0.23825</td>\n",
       "      <td>chrM</td>\n",
       "      <td>16073</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>170202927 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           over_10  over_20   chr    pos\n",
       "0          0.00000  0.00000  chr1  11819\n",
       "1          0.00000  0.00000  chr1  11820\n",
       "2          0.00000  0.00000  chr1  11821\n",
       "3          0.00000  0.00000  chr1  11822\n",
       "4          0.00000  0.00000  chr1  11823\n",
       "...            ...      ...   ...    ...\n",
       "170202922  0.29592  0.23756  chrM  16069\n",
       "170202923  0.29582  0.23788  chrM  16070\n",
       "170202924  0.29689  0.23808  chrM  16071\n",
       "170202925  0.29674  0.23815  chrM  16072\n",
       "170202926  0.29712  0.23825  chrM  16073\n",
       "\n",
       "[170202927 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with gzip.open(coverage_file_path, 'rt') as coverage_file:\n",
    "    # Read the file into a pandas DataFrame\n",
    "    coverage_df = pd.read_csv(coverage_file, sep='\\t',\n",
    "                             usecols=['locus', 'over_10', 'over_20'])\n",
    "\n",
    "# Expand locus column\n",
    "coverage_df[['chr', 'pos']] = coverage_df['locus'].str.split(':', expand=True)\n",
    "coverage_df['pos'] = coverage_df['pos'].astype(int)\n",
    "\n",
    "# Drop the original locus column\n",
    "coverage_df = coverage_df.drop('locus', axis=1)\n",
    "\n",
    "coverage_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7e140df9",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 424
    },
    "executionInfo": {
     "elapsed": 54171,
     "status": "ok",
     "timestamp": 1703323440602,
     "user": {
      "displayName": "Oscar Thomas Aguilar",
      "userId": "15151607230073393698"
     },
     "user_tz": 300
    },
    "id": "-kMYvaQmGs44",
    "outputId": "70fd4749-819f-42a6-98c3-c7e4c6f6d22a",
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript</th>\n",
       "      <th>transcript_num</th>\n",
       "      <th>transcript_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>65565</td>\n",
       "      <td>65573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENST00000641515</td>\n",
       "      <td>2</td>\n",
       "      <td>OR4F5-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69037</td>\n",
       "      <td>70005</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>ENST00000641515</td>\n",
       "      <td>2</td>\n",
       "      <td>OR4F5-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>450743</td>\n",
       "      <td>451678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000284733.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>ENST00000426406.4</td>\n",
       "      <td>ENST00000426406</td>\n",
       "      <td>4</td>\n",
       "      <td>OR4F29-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>685719</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000284662.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>ENST00000332831.5</td>\n",
       "      <td>ENST00000332831</td>\n",
       "      <td>5</td>\n",
       "      <td>OR4F16-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>524</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>924432</td>\n",
       "      <td>924948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENST00000616016.5</td>\n",
       "      <td>ENST00000616016</td>\n",
       "      <td>5</td>\n",
       "      <td>SAMD11-209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1998488</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>10470</td>\n",
       "      <td>10763</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000212907.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND4L</td>\n",
       "      <td>ENST00000361335.1</td>\n",
       "      <td>ENST00000361335</td>\n",
       "      <td>1</td>\n",
       "      <td>MT-ND4L-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1998495</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>10760</td>\n",
       "      <td>12137</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000198886.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND4</td>\n",
       "      <td>ENST00000361381.2</td>\n",
       "      <td>ENST00000361381</td>\n",
       "      <td>2</td>\n",
       "      <td>MT-ND4-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1998509</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>12337</td>\n",
       "      <td>14145</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000198786.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND5</td>\n",
       "      <td>ENST00000361567.2</td>\n",
       "      <td>ENST00000361567</td>\n",
       "      <td>2</td>\n",
       "      <td>MT-ND5-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1998515</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>ENST00000361681.2</td>\n",
       "      <td>ENST00000361681</td>\n",
       "      <td>2</td>\n",
       "      <td>MT-ND6-201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1998523</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>ENST00000361789.2</td>\n",
       "      <td>ENST00000361789</td>\n",
       "      <td>2</td>\n",
       "      <td>MT-CYB-201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>654548 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr   source feature   start     end score strand frame  \\\n",
       "60       chr1   HAVANA     CDS   65565   65573     .      +     0   \n",
       "63       chr1   HAVANA     CDS   69037   70005     .      +     0   \n",
       "236      chr1   HAVANA     CDS  450743  451678     .      -     0   \n",
       "304      chr1   HAVANA     CDS  685719  686654     .      -     0   \n",
       "524      chr1   HAVANA     CDS  924432  924948     .      +     0   \n",
       "...       ...      ...     ...     ...     ...   ...    ...   ...   \n",
       "1998488  chrM  ENSEMBL     CDS   10470   10763     .      +     0   \n",
       "1998495  chrM  ENSEMBL     CDS   10760   12137     .      +     0   \n",
       "1998509  chrM  ENSEMBL     CDS   12337   14145     .      +     0   \n",
       "1998515  chrM  ENSEMBL     CDS   14149   14673     .      -     0   \n",
       "1998523  chrM  ENSEMBL     CDS   14747   15887     .      +     0   \n",
       "\n",
       "                    gene_id       gene_type gene_name      transcript_id  \\\n",
       "60        ENSG00000186092.7  protein_coding     OR4F5  ENST00000641515.2   \n",
       "63        ENSG00000186092.7  protein_coding     OR4F5  ENST00000641515.2   \n",
       "236       ENSG00000284733.2  protein_coding    OR4F29  ENST00000426406.4   \n",
       "304       ENSG00000284662.2  protein_coding    OR4F16  ENST00000332831.5   \n",
       "524      ENSG00000187634.13  protein_coding    SAMD11  ENST00000616016.5   \n",
       "...                     ...             ...       ...                ...   \n",
       "1998488   ENSG00000212907.2  protein_coding   MT-ND4L  ENST00000361335.1   \n",
       "1998495   ENSG00000198886.2  protein_coding    MT-ND4  ENST00000361381.2   \n",
       "1998509   ENSG00000198786.2  protein_coding    MT-ND5  ENST00000361567.2   \n",
       "1998515   ENSG00000198695.2  protein_coding    MT-ND6  ENST00000361681.2   \n",
       "1998523   ENSG00000198727.2  protein_coding    MT-CYB  ENST00000361789.2   \n",
       "\n",
       "              transcript transcript_num transcript_name  \n",
       "60       ENST00000641515              2       OR4F5-201  \n",
       "63       ENST00000641515              2       OR4F5-201  \n",
       "236      ENST00000426406              4      OR4F29-201  \n",
       "304      ENST00000332831              5      OR4F16-201  \n",
       "524      ENST00000616016              5      SAMD11-209  \n",
       "...                  ...            ...             ...  \n",
       "1998488  ENST00000361335              1     MT-ND4L-201  \n",
       "1998495  ENST00000361381              2      MT-ND4-201  \n",
       "1998509  ENST00000361567              2      MT-ND5-201  \n",
       "1998515  ENST00000361681              2      MT-ND6-201  \n",
       "1998523  ENST00000361789              2      MT-CYB-201  \n",
       "\n",
       "[654548 rows x 15 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the GTF file into a pandas DataFrame\n",
    "gene_df = pd.read_csv(gene_annotation_file_path, sep='\\t', comment='#', header=None, \n",
    "                      names=['chr', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute'], \n",
    "                      dtype={'start': int, 'end': int})\n",
    "\n",
    "# Extract 'gene_id' from attributes\n",
    "gene_df['gene_id'] = gene_df['attribute'].str.extract(r'gene_id \"(.*?)\"')\n",
    "\n",
    "# Extract 'gene_type' from attributes\n",
    "gene_df['gene_type'] = gene_df['attribute'].str.extract(r'gene_type \"(.*?)\"')\n",
    "\n",
    "# Extract 'gene_name' from attributes\n",
    "gene_df['gene_name'] = gene_df['attribute'].str.extract(r'gene_name \"(.*?)\"')\n",
    "\n",
    "# Extract 'transcript_id' from attributes\n",
    "gene_df['transcript_id'] = gene_df['attribute'].str.extract(r'transcript_id \"(.*?)\"')\n",
    "\n",
    "# Extract 'transcript' and 'num' from transcript_id\n",
    "gene_df[['transcript', 'transcript_num']] = gene_df['transcript_id'].str.split('.', expand=True)\n",
    "\n",
    "# Extract 'transcript_name' from attributes\n",
    "gene_df['transcript_name'] = gene_df['attribute'].str.extract(r'transcript_name \"(.*?)\"')\n",
    "\n",
    "# Drop the original attribute column\n",
    "gene_df = gene_df.drop('attribute', axis=1)\n",
    "\n",
    "# Filter rows for protein-coding regions\n",
    "gene_df = gene_df[(gene_df['gene_type'] == 'protein_coding') & (gene_df['feature'] == 'CDS')]\n",
    "\n",
    "gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ba4c787f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>934655699</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>934655700</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>934655701</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>934655702</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>934655703</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56887898</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>934655704 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            chr       pos\n",
       "0          chr1     10001\n",
       "1          chr1     10073\n",
       "2          chr1     10108\n",
       "3          chr1     10110\n",
       "4          chr1     10113\n",
       "...         ...       ...\n",
       "934655699  chrY  56887741\n",
       "934655700  chrY  56887744\n",
       "934655701  chrY  56887793\n",
       "934655702  chrY  56887802\n",
       "934655703  chrY  56887898\n",
       "\n",
       "[934655704 rows x 2 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Initialize an empty dataframe to store the concatenated data\n",
    "variants_df = pd.DataFrame()\n",
    "\n",
    "# Check if the joined file already exists\n",
    "if not os.path.exists(variants_file_path):\n",
    "    # Loop through chromosomes and append dataframes with tqdm\n",
    "    for filenum in tqdm(range(1, 9), desc='Processing Files'):\n",
    "        # Read the variants file into a pandas DataFrame\n",
    "        file_variants_df = pd.read_csv(\n",
    "            data_path + \"output_prefix_0\" + str(filenum) + \".tsv.gz\",\n",
    "            sep='-',\n",
    "            skiprows=1,\n",
    "            usecols=[0,1],\n",
    "            names=['chr', 'pos'],\n",
    "            dtype={'chr': str, 'pos': int},\n",
    "        )\n",
    "        \n",
    "        file_variants_df.drop_duplicates(inplace=True)\n",
    "\n",
    "        # Append the dataframe to the main dataframe\n",
    "        variants_df = pd.concat([variants_df, file_variants_df], ignore_index=True)\n",
    "    \n",
    "    # Add chr prefix for consistency\n",
    "    variants_df['chr'] = 'chr' + variants_df['chr'].astype(str)\n",
    "    \n",
    "    # Write variants_df to a csv to avoid recomputing\n",
    "    variants_df.to_csv(variants_file_path, index=False, compression='gzip', sep='\\t')\n",
    "    \n",
    "else:\n",
    "    # Read the file into a pandas DataFrame\n",
    "    variants_df = pd.read_csv(variants_file_path, sep='\\t')\n",
    "\n",
    "# variants_df = variants_df[variants_df['chr'] == 'chr1']\n",
    "    \n",
    "variants_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a2bac06",
   "metadata": {},
   "source": [
    "## Train HMM on AoU chromosome 2 and get predictions for full genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45af0bd4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Define coverage filter\n",
    "coverage_filter = '0.5_over10' # coverage['over10'] > 0.5\n",
    "# coverage_filter = '0.9_over20' # coverage['over20'] > 0.9\n",
    "\n",
    "# Define pattern for coverage file paths\n",
    "# coverage_file_prefix = data_path + 'gnomad.genomes.r3.0.1.coverage_0.5_over10_chr'\n",
    "coverage_file_prefix = data_path + 'gnomad_r3_coverage/' + 'gnomad.genomes.r3.0.1.coverage_' + coverage_filter + '_chr'\n",
    "coverage_file_suffix = '.summary.tsv.gz'\n",
    "\n",
    "# Parameters for analysis\n",
    "order = 2\n",
    "\n",
    "# Define model name (and absolute path)\n",
    "model_name = results_path + 'HMM_aou_' + coverage_filter + '_chr2'\n",
    "\n",
    "# Define path to trained HMM\n",
    "hmm_file_path = model_name + '_model.joblib'\n",
    "\n",
    "if os.path.exists(hmm_file_path):\n",
    "    # Load the pretrained model\n",
    "    model = joblib.load(hmm_file_path)\n",
    "    \n",
    "else:\n",
    "    # Filter rows for chromosome 2\n",
    "    chr_variants_df = aou_variants_df[aou_variants_df['chr'] == 'chr2']\n",
    "    chr_coverage_df = pd.read_csv(coverage_file_prefix + '2' + coverage_file_suffix, sep='\\t', names=['pos'])\n",
    "\n",
    "    # Get training data for the chromosome\n",
    "    observations, positions = get_sequence(chr_coverage_df, chr_variants_df)\n",
    "\n",
    "    # Fit HMM to Chromosome 2\n",
    "    model = fit_HMM(observations, order=order)\n",
    "\n",
    "    # Save the HMM model to a file to avoid retraining\n",
    "    joblib.dump(model, hmm_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d92f4bbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for chromnum in range(1,23):\n",
    "for chromnum in [1,]:\n",
    "    print('-'*100)\n",
    "    print(\"PROCESSING CHROMOSOME\", str(chromnum))\n",
    "\n",
    "    # Filter rows for given chromosome\n",
    "    chr_variants_df = aou_variants_df[aou_variants_df['chr'] == 'chr' + str(chromnum)]\n",
    "    chr_coverage_df = pd.read_csv(coverage_file_prefix + str(chromnum) + coverage_file_suffix, sep='\\t', names=['pos'])\n",
    "    \n",
    "    # Get training data for the chromosome\n",
    "    observations, positions = get_sequence(chr_coverage_df, chr_variants_df)\n",
    "\n",
    "    # Fit HMM and retrieve probabilites\n",
    "    probabilities = get_HMM_predictions(observations, model, order=order)\n",
    "    \n",
    "    # Create a DataFrame with 'pos' reflecting the index of the original sequence and 'prob_0/1' as the predictions\n",
    "    chr_predictions_df = pd.DataFrame({'chr': 'chr' + str(chromnum),\n",
    "                                       'pos': positions[0:-order],\n",
    "                                       'prob_0': probabilities[:, 0], \n",
    "                                       'prob_1': probabilities[:, 1], \n",
    "                                       'observation': observations[0:-order]\n",
    "                                      })\n",
    "    \n",
    "    predictions_file_prefix = model_name + \"_predictions_aou_chr\"\n",
    "    \n",
    "    # Checkpoint chr_predictions_df to a csv to avoid memory bottleneck\n",
    "    chr_predictions_df.to_csv(predictions_file_prefix + str(chromnum) + \".tsv.gz\", index=False, compression='gzip', sep='\\t')\n",
    "    \n",
    "    # Plot histograms for predicted vs observed variants\n",
    "    plot_hist_from_predictions(chr_predictions_df)\n",
    "    \n",
    "    # Run regression\n",
    "    f_statistic, p_value_f_statistic = ols_regression(chr_predictions_df)\n",
    "    \n",
    "chr_predictions_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2afa9677",
   "metadata": {},
   "source": [
    "## Get predictions for AoU WES from the HMM trained on RGC chromosome 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a21e31ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing Chromosomes: 100%|████████████████████████████████████████████████| 22/22 [25:51<00:00, 70.53s/it]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>prob_0</th>\n",
       "      <th>prob_1</th>\n",
       "      <th>observation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>925922</td>\n",
       "      <td>9.596546e-31</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>925923</td>\n",
       "      <td>7.247857e-01</td>\n",
       "      <td>0.275214</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>925924</td>\n",
       "      <td>9.135695e-01</td>\n",
       "      <td>0.086431</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>925925</td>\n",
       "      <td>9.627417e-01</td>\n",
       "      <td>0.037258</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>925926</td>\n",
       "      <td>9.755495e-01</td>\n",
       "      <td>0.024450</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28933587</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50777975</td>\n",
       "      <td>9.796181e-01</td>\n",
       "      <td>0.020382</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28933588</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50777976</td>\n",
       "      <td>9.783619e-01</td>\n",
       "      <td>0.021638</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28933589</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50777977</td>\n",
       "      <td>9.735393e-01</td>\n",
       "      <td>0.026461</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28933590</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50777978</td>\n",
       "      <td>9.550240e-01</td>\n",
       "      <td>0.044976</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28933591</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50777979</td>\n",
       "      <td>8.839394e-01</td>\n",
       "      <td>0.116061</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28933592 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            chr       pos        prob_0    prob_1  observation\n",
       "0          chr1    925922  9.596546e-31  1.000000          0.0\n",
       "1          chr1    925923  7.247857e-01  0.275214          0.0\n",
       "2          chr1    925924  9.135695e-01  0.086431          0.0\n",
       "3          chr1    925925  9.627417e-01  0.037258          0.0\n",
       "4          chr1    925926  9.755495e-01  0.024450          0.0\n",
       "...         ...       ...           ...       ...          ...\n",
       "28933587  chr22  50777975  9.796181e-01  0.020382          0.0\n",
       "28933588  chr22  50777976  9.783619e-01  0.021638          0.0\n",
       "28933589  chr22  50777977  9.735393e-01  0.026461          0.0\n",
       "28933590  chr22  50777978  9.550240e-01  0.044976          0.0\n",
       "28933591  chr22  50777979  8.839394e-01  0.116061          0.0\n",
       "\n",
       "[28933592 rows x 5 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Parameters for analysis\n",
    "# Order of the HMM (defines how many previous states influence the current state)\n",
    "order = 2\n",
    "\n",
    "# Coverage threshold for filtering regions with sufficient sequencing depth\n",
    "coverage_thr = 0.9\n",
    "\n",
    "# Coverage category to use for filtering (e.g., 'over_20' means regions with coverage > 20)\n",
    "coverage_category = 'over_20'\n",
    "\n",
    "# Define the coverage filter based on the selected threshold and category\n",
    "# Example: '0.9_over20' indicates we are interested in regions where over 90% of positions have coverage > 20\n",
    "coverage_df = coverage_df[coverage_df[coverage_category] > coverage_thr]\n",
    "\n",
    "# Define the model name based on the selected coverage filter and chromosome 2\n",
    "model_name = results_path + 'HMM_rgc_0.9_over20_chr2'\n",
    "\n",
    "# Load the pretrained model\n",
    "model = joblib.load(model_name + '_model.joblib')\n",
    "\n",
    "# Initialize an empty DataFrame to store predictions for all chromosomes\n",
    "predictions_df = pd.DataFrame(columns=['chr', 'pos','prob_0', 'prob_1', 'observation'])\n",
    "\n",
    "# Loop over each autosome (chromosomes 1-22) using trange for a progress bar\n",
    "for chromnum in trange(1, 23, desc=\"Processing Chromosomes\"):\n",
    "    # Extract the sequence of observations and positions for the current chromosome\n",
    "    observations, positions = get_sequence(chromnum)\n",
    "\n",
    "    # Use the trained HMM to predict probabilities for each state (0 and 1) in the sequence\n",
    "    probabilities = get_HMM_predictions(observations, model, order=order)\n",
    "    \n",
    "    # Create a DataFrame with the predictions for the current chromosome\n",
    "    chr_predictions_df = pd.DataFrame({'chr': 'chr' + str(chromnum),\n",
    "                                       'pos': positions[0:-order],\n",
    "                                       'prob_0': probabilities[:, 0], \n",
    "                                       'prob_1': probabilities[:, 1], \n",
    "                                       'observation': observations[0:-order]\n",
    "                                      })\n",
    "    \n",
    "    # Plot histograms comparing predicted vs observed variants\n",
    "#     plot_hist_from_predictions(chr_predictions_df)\n",
    "    \n",
    "    # Perform regression analysis on the predictions\n",
    "#     f_statistic, p_value_f_statistic = ols_regression(chr_predictions_df)\n",
    "    \n",
    "    # Append the predictions for the current chromosome to the overall DataFrame\n",
    "    predictions_df = pd.concat([predictions_df, chr_predictions_df], ignore_index=True)\n",
    "\n",
    "# Save the complete predictions DataFrame to a compressed CSV file to avoid recomputing\n",
    "predictions_df.to_csv(model_name + \"_predictions_aou_wes.tsv.gz\", index=False, compression='gzip', sep='\\t')\n",
    "\n",
    "# Display the predictions DataFrame\n",
    "predictions_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
